
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
      <link rel="icon" href="../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.2.4, mkdocs-material-7.3.0">
    
    
      
        <title>Comparator - Ediacara</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.8b42a75e.min.css">
      
        
        <link rel="stylesheet" href="../../../assets/stylesheets/palette.3f5d1f46.min.css">
        
          
          
          <meta name="theme-color" content="#4cae4f">
        
      
    
    
    
      
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700%7CRoboto+Mono&display=fallback">
        <style>:root{--md-text-font-family:"Roboto";--md-code-font-family:"Roboto Mono"}</style>
      
    
    
    
    
      


    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="" data-md-color-primary="green" data-md-color-accent="lightgreen">
  
    
    <script>function __prefix(e){return new URL("../../..",location).pathname+"."+e}function __get(e,t=localStorage){return JSON.parse(t.getItem(__prefix(e)))}</script>
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#module-ediacaracomparator" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../.." title="Ediacara" class="md-header__button md-logo" aria-label="Ediacara" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Ediacara
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Comparator
            
          </span>
        </div>
      </div>
    </div>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
      </label>
      
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" aria-label="Clear" tabindex="-1">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        
<a href="https://github.com/Edinburgh-Genome-Foundry/Ediacara/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    Ediacara
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../.." title="Ediacara" class="md-nav__button md-logo" aria-label="Ediacara" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54z"/></svg>

    </a>
    Ediacara
  </label>
  
    <div class="md-nav__source">
      
<a href="https://github.com/Edinburgh-Genome-Foundry/Ediacara/" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    Ediacara
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  
  
  
    <li class="md-nav__item">
      <a href="../../.." class="md-nav__link">
        Home
      </a>
    </li>
  

    
      
      
      

  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_2" type="checkbox" id="__nav_2" checked>
      
      
      
      
        <label class="md-nav__link" for="__nav_2">
          Reference
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Reference" data-md-level="1">
        <label class="md-nav__title" for="__nav_2">
          <span class="md-nav__icon md-icon"></span>
          Reference
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
    
  
  
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
      
        <input class="md-nav__toggle md-toggle" data-md-toggle="__nav_2_1" type="checkbox" id="__nav_2_1" checked>
      
      
      
      
        <label class="md-nav__link" for="__nav_2_1">
          Ediacara
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <nav class="md-nav" aria-label="Ediacara" data-md-level="2">
        <label class="md-nav__title" for="__nav_2_1">
          <span class="md-nav__icon md-icon"></span>
          Ediacara
        </label>
        <ul class="md-nav__list" data-md-scrollfix>
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../Assembly/" class="md-nav__link">
        Assembly
      </a>
    </li>
  

            
          
            
              
  
  
    
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" data-md-toggle="toc" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          Comparator
          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        Comparator
      </a>
      
        
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#variables" class="md-nav__link">
    Variables
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#classes" class="md-nav__link">
    Classes
  </a>
  
    <nav class="md-nav" aria-label="Classes">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#comparator" class="md-nav__link">
    Comparator
  </a>
  
    <nav class="md-nav" aria-label="Comparator">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#methods" class="md-nav__link">
    Methods
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#calculate_stats" class="md-nav__link">
    calculate_stats
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compare_with_assembly" class="md-nav__link">
    compare_with_assembly
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#create_vcf_table" class="md-nav__link">
    create_vcf_table
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#filter_fastq" class="md-nav__link">
    filter_fastq
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#find_big_inserts" class="md-nav__link">
    find_big_inserts
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#get_weak_read_histogram_data" class="md-nav__link">
    get_weak_read_histogram_data
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#perform_comparison" class="md-nav__link">
    perform_comparison
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plot_coverage" class="md-nav__link">
    plot_coverage
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#comparatorgroup" class="md-nav__link">
    ComparatorGroup
  </a>
  
    <nav class="md-nav" aria-label="ComparatorGroup">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#static-methods" class="md-nav__link">
    Static methods
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#load_paf" class="md-nav__link">
    load_paf
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#load_tsv" class="md-nav__link">
    load_tsv
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#methods_1" class="md-nav__link">
    Methods
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#add_comparator" class="md-nav__link">
    add_comparator
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#create_comparator" class="md-nav__link">
    create_comparator
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#perform_all_comparisons" class="md-nav__link">
    perform_all_comparisons
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plot_fastq_histogram" class="md-nav__link">
    plot_fastq_histogram
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#subset_paf" class="md-nav__link">
    subset_paf
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#summarise_analysis" class="md-nav__link">
    summarise_analysis
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#customtranslator" class="md-nav__link">
    CustomTranslator
  </a>
  
    <nav class="md-nav" aria-label="CustomTranslator">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ancestors-in-mro" class="md-nav__link">
    Ancestors (in MRO)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#class-variables" class="md-nav__link">
    Class variables
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#static-methods_1" class="md-nav__link">
    Static methods
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#quick_class_plot" class="md-nav__link">
    quick_class_plot
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#methods_2" class="md-nav__link">
    Methods
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_box_color" class="md-nav__link">
    compute_feature_box_color
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_box_linewidth" class="md-nav__link">
    compute_feature_box_linewidth
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_color" class="md-nav__link">
    compute_feature_color
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_fontdict" class="md-nav__link">
    compute_feature_fontdict
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_html" class="md-nav__link">
    compute_feature_html
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_label" class="md-nav__link">
    compute_feature_label
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_label_link_color" class="md-nav__link">
    compute_feature_label_link_color
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_legend_text" class="md-nav__link">
    compute_feature_legend_text
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_linewidth" class="md-nav__link">
    compute_feature_linewidth
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_filtered_features" class="md-nav__link">
    compute_filtered_features
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#quick_plot" class="md-nav__link">
    quick_plot
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#translate_feature" class="md-nav__link">
    translate_feature
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#translate_record" class="md-nav__link">
    translate_record
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sequencinggroup" class="md-nav__link">
    SequencingGroup
  </a>
  
    <nav class="md-nav" aria-label="SequencingGroup">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#methods_3" class="md-nav__link">
    Methods
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#perform_all_comparisons_in_sequencinggroup" class="md-nav__link">
    perform_all_comparisons_in_sequencinggroup
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../" class="md-nav__link">
        Index
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../reports/" class="md-nav__link">
        Reports
      </a>
    </li>
  

            
          
            
              
  
  
  
    <li class="md-nav__item">
      <a href="../version/" class="md-nav__link">
        Version
      </a>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

            
          
        </ul>
      </nav>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#variables" class="md-nav__link">
    Variables
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#classes" class="md-nav__link">
    Classes
  </a>
  
    <nav class="md-nav" aria-label="Classes">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#comparator" class="md-nav__link">
    Comparator
  </a>
  
    <nav class="md-nav" aria-label="Comparator">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#methods" class="md-nav__link">
    Methods
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#calculate_stats" class="md-nav__link">
    calculate_stats
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compare_with_assembly" class="md-nav__link">
    compare_with_assembly
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#create_vcf_table" class="md-nav__link">
    create_vcf_table
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#filter_fastq" class="md-nav__link">
    filter_fastq
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#find_big_inserts" class="md-nav__link">
    find_big_inserts
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#get_weak_read_histogram_data" class="md-nav__link">
    get_weak_read_histogram_data
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#perform_comparison" class="md-nav__link">
    perform_comparison
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plot_coverage" class="md-nav__link">
    plot_coverage
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#comparatorgroup" class="md-nav__link">
    ComparatorGroup
  </a>
  
    <nav class="md-nav" aria-label="ComparatorGroup">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#static-methods" class="md-nav__link">
    Static methods
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#load_paf" class="md-nav__link">
    load_paf
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#load_tsv" class="md-nav__link">
    load_tsv
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#methods_1" class="md-nav__link">
    Methods
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#add_comparator" class="md-nav__link">
    add_comparator
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#create_comparator" class="md-nav__link">
    create_comparator
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#perform_all_comparisons" class="md-nav__link">
    perform_all_comparisons
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#plot_fastq_histogram" class="md-nav__link">
    plot_fastq_histogram
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#subset_paf" class="md-nav__link">
    subset_paf
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#summarise_analysis" class="md-nav__link">
    summarise_analysis
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#customtranslator" class="md-nav__link">
    CustomTranslator
  </a>
  
    <nav class="md-nav" aria-label="CustomTranslator">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#ancestors-in-mro" class="md-nav__link">
    Ancestors (in MRO)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#class-variables" class="md-nav__link">
    Class variables
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#static-methods_1" class="md-nav__link">
    Static methods
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#quick_class_plot" class="md-nav__link">
    quick_class_plot
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#methods_2" class="md-nav__link">
    Methods
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_box_color" class="md-nav__link">
    compute_feature_box_color
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_box_linewidth" class="md-nav__link">
    compute_feature_box_linewidth
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_color" class="md-nav__link">
    compute_feature_color
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_fontdict" class="md-nav__link">
    compute_feature_fontdict
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_html" class="md-nav__link">
    compute_feature_html
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_label" class="md-nav__link">
    compute_feature_label
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_label_link_color" class="md-nav__link">
    compute_feature_label_link_color
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_legend_text" class="md-nav__link">
    compute_feature_legend_text
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_feature_linewidth" class="md-nav__link">
    compute_feature_linewidth
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#compute_filtered_features" class="md-nav__link">
    compute_filtered_features
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#quick_plot" class="md-nav__link">
    quick_plot
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#translate_feature" class="md-nav__link">
    translate_feature
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#translate_record" class="md-nav__link">
    translate_record
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#sequencinggroup" class="md-nav__link">
    SequencingGroup
  </a>
  
    <nav class="md-nav" aria-label="SequencingGroup">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#methods_3" class="md-nav__link">
    Methods
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#perform_all_comparisons_in_sequencinggroup" class="md-nav__link">
    perform_all_comparisons_in_sequencinggroup
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content" data-md-component="content">
            <article class="md-content__inner md-typeset">
              
                
                  <a href="https://github.com/Edinburgh-Genome-Foundry/Ediacara/edit/main/reference/ediacara/Comparator.md" title="Edit this page" class="md-content__button md-icon">
                    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25z"/></svg>
                  </a>
                
                
                <h1 id="module-ediacaracomparator">Module ediacara.Comparator</h1>
<p>None</p>
<p>None</p>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code><span class="kn">import</span> <span class="nn">itertools</span>

<span class="kn">import</span> <span class="nn">re</span>

<span class="kn">import</span> <span class="nn">statistics</span>

<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>

<span class="kn">import</span> <span class="nn">numpy</span>

<span class="kn">import</span> <span class="nn">pandas</span>

<span class="kn">from</span> <span class="nn">weighted_levenshtein</span> <span class="kn">import</span> <span class="n">lev</span>

<span class="kn">from</span> <span class="nn">Bio</span> <span class="kn">import</span> <span class="n">SeqIO</span>

<span class="kn">import</span> <span class="nn">cyvcf2</span>

<span class="kn">import</span> <span class="nn">dna_features_viewer</span>

<span class="kn">import</span> <span class="nn">geneblocks</span>

<span class="n">vcf_selected_info_keytable</span> <span class="o">=</span> <span class="p">{</span>

    <span class="s2">&quot;DP&quot;</span><span class="p">:</span> <span class="s2">&quot;Total read depth at the locus&quot;</span><span class="p">,</span>

    <span class="s2">&quot;RO&quot;</span><span class="p">:</span> <span class="s2">&quot;Reference allele observation count&quot;</span><span class="p">,</span>

    <span class="s2">&quot;AO&quot;</span><span class="p">:</span> <span class="s2">&quot;Alternate allele observations&quot;</span><span class="p">,</span>

    <span class="s2">&quot;TYPE&quot;</span><span class="p">:</span> <span class="s2">&quot;The type of allele (either snp, mnp, ins, del or complex)&quot;</span><span class="p">,</span>

<span class="p">}</span>

<span class="n">result_keywords</span> <span class="o">=</span> <span class="p">{</span>

    <span class="s2">&quot;good&quot;</span><span class="p">:</span> <span class="s2">&quot;PASS&quot;</span><span class="p">,</span>

    <span class="s2">&quot;error&quot;</span><span class="p">:</span> <span class="s2">&quot;FAIL&quot;</span><span class="p">,</span>

    <span class="s2">&quot;warning&quot;</span><span class="p">:</span> <span class="s2">&quot;WARNING&quot;</span><span class="p">,</span>

    <span class="s2">&quot;uncertain&quot;</span><span class="p">:</span> <span class="s2">&quot;LOW_COVERAGE&quot;</span><span class="p">,</span>

<span class="p">}</span>

<span class="k">class</span> <span class="nc">CustomTranslator</span><span class="p">(</span><span class="n">dna_features_viewer</span><span class="o">.</span><span class="n">BiopythonTranslator</span><span class="p">):</span>

    <span class="sd">&quot;&quot;&quot;Custom translator.&quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="nf">compute_filtered_features</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">features</span><span class="p">):</span>

        <span class="sd">&quot;&quot;&quot;Display only &quot;From &quot; features and overhangs.&quot;&quot;&quot;</span>

        <span class="n">filtered_features</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">feature</span> <span class="ow">in</span> <span class="n">features</span><span class="p">:</span>

            <span class="k">try</span><span class="p">:</span>  <span class="c1"># may not have a &#39;label&#39;, count 4 to include overhang annotations</span>

                <span class="k">if</span> <span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">feature</span><span class="o">.</span><span class="n">qualifiers</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;label&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">])</span> <span class="o">==</span> <span class="mi">4</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span>

                    <span class="s2">&quot;From &quot;</span> <span class="ow">in</span> <span class="nb">str</span><span class="p">(</span><span class="n">feature</span><span class="o">.</span><span class="n">qualifiers</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;label&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">))</span>

                <span class="p">):</span>

                    <span class="n">filtered_features</span> <span class="o">+=</span> <span class="p">[</span><span class="n">feature</span><span class="p">]</span>

            <span class="k">except</span><span class="p">:</span>

                <span class="k">pass</span>

        <span class="k">return</span> <span class="n">filtered_features</span>

<span class="k">class</span> <span class="nc">SequencingGroup</span><span class="p">:</span>

    <span class="sd">&quot;&quot;&quot;Analyse alignments of a sequencing run.</span>

<span class="sd">    **Parameters**</span>

<span class="sd">    **comparatorgroups**</span>

<span class="sd">    &gt; List of `ComparatorGroup` instances.</span>

<span class="sd">    **name**</span>

<span class="sd">    &gt; Name of the sequencing project (`str`).</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">comparatorgroups</span><span class="p">,</span> <span class="n">name</span><span class="o">=</span><span class="s2">&quot;Unnamed&quot;</span><span class="p">):</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">comparatorgroups</span> <span class="o">=</span> <span class="n">comparatorgroups</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">name</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">vcf_cutoff</span> <span class="o">=</span> <span class="mi">20</span>  <span class="c1"># max number of VCF entries to display in report</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">result_keywords</span> <span class="o">=</span> <span class="n">result_keywords</span>

    <span class="k">def</span> <span class="nf">perform_all_comparisons_in_sequencinggroup</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">number_of_reads</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="k">for</span> <span class="n">comparatorgroup</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">comparatorgroups</span><span class="p">:</span>

            <span class="n">comparatorgroup</span><span class="o">.</span><span class="n">perform_all_comparisons</span><span class="p">()</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">number_of_reads</span> <span class="o">+=</span> <span class="n">comparatorgroup</span><span class="o">.</span><span class="n">n_fastq_reads</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">number_of_barcodes</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">comparatorgroups</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">comparisons_performed</span> <span class="o">=</span> <span class="kc">True</span>

<span class="k">class</span> <span class="nc">ComparatorGroup</span><span class="p">:</span>

    <span class="sd">&quot;&quot;&quot;Analyse alignments to a set of references.</span>

<span class="sd">    **Parameters**</span>

<span class="sd">    **references**</span>

<span class="sd">    &gt; Dictionary of name: Biopython record.</span>

<span class="sd">    **alignments**</span>

<span class="sd">    &gt; Dictionary of pandas dataframes: {&quot;paf&quot;: df_paf, &quot;tsv&quot;: df_tsv}. The PAF dataframe</span>

<span class="sd">    was created from the output of `paftools.js sam2paf` and the TSV dataframe was</span>

<span class="sd">    created from the output of `samtools depth -aa`.</span>

<span class="sd">    **barcode**</span>

<span class="sd">    &gt; The barcode number or ID (`str`).</span>

<span class="sd">    **assembly_paths**</span>

<span class="sd">    &gt; Dictionary of sample: consensus sequence path (`dict`).</span>

<span class="sd">    **vcf_paths**</span>

<span class="sd">    &gt; Dictionary of sample: filtered VCF filepath (`dict`).</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span>

        <span class="bp">self</span><span class="p">,</span>

        <span class="n">references</span><span class="p">,</span>

        <span class="n">alignments</span><span class="p">,</span>

        <span class="n">barcode</span><span class="o">=</span><span class="s2">&quot;barcode&quot;</span><span class="p">,</span>

        <span class="n">assembly_paths</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>

        <span class="n">vcf_paths</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>

    <span class="p">):</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">references</span> <span class="o">=</span> <span class="n">references</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">paf</span> <span class="o">=</span> <span class="n">alignments</span><span class="p">[</span><span class="s2">&quot;paf&quot;</span><span class="p">]</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">tsv</span> <span class="o">=</span> <span class="n">alignments</span><span class="p">[</span><span class="s2">&quot;tsv&quot;</span><span class="p">]</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">comparators</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">barcode</span> <span class="o">=</span> <span class="n">barcode</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">assembly_paths</span> <span class="o">=</span> <span class="n">assembly_paths</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">vcf_paths</span> <span class="o">=</span> <span class="n">vcf_paths</span>

    <span class="k">def</span> <span class="nf">create_comparator</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">reference_name</span><span class="p">):</span>

        <span class="sd">&quot;&quot;&quot;Create a Comparator instance from one of the loaded references.</span>

<span class="sd">        **Parameters**</span>

<span class="sd">        **reference_name**</span>

<span class="sd">        &gt; The name of the reference (`str`).</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">subset_paf</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">subset_paf</span><span class="p">(</span><span class="n">reference_name</span><span class="p">)</span>

        <span class="n">subset_tsv</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">tsv</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">tsv</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">reference_name</span><span class="p">]</span>

        <span class="n">alignment</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;paf&quot;</span><span class="p">:</span> <span class="n">subset_paf</span><span class="p">,</span> <span class="s2">&quot;tsv&quot;</span><span class="p">:</span> <span class="n">subset_tsv</span><span class="p">}</span>

        <span class="n">comparator</span> <span class="o">=</span> <span class="n">Comparator</span><span class="p">(</span>

            <span class="p">{</span><span class="n">reference_name</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">references</span><span class="p">[</span><span class="n">reference_name</span><span class="p">]},</span> <span class="n">alignment</span>

        <span class="p">)</span>

        <span class="k">return</span> <span class="n">comparator</span>

    <span class="k">def</span> <span class="nf">add_comparator</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">reference_name</span><span class="p">):</span>

        <span class="sd">&quot;&quot;&quot;Create comparator and add to the group. See create_comparator().&quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">comparators</span> <span class="o">+=</span> <span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">create_comparator</span><span class="p">(</span><span class="n">reference_name</span><span class="p">)]</span>

    <span class="k">def</span> <span class="nf">subset_paf</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">reference_name</span><span class="p">):</span>

        <span class="sd">&quot;&quot;&quot;Subset PAF dataframe for the reference.</span>

<span class="sd">        Select reads that align to the reference and also align *best* to the reference.</span>

<span class="sd">        **Parameters**</span>

<span class="sd">        **reference_name**</span>

<span class="sd">        &gt; Name of the reference sequence to filter for (`str`).</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Work only with the values used for filtering:</span>

        <span class="n">paf3</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">paf</span><span class="p">[[</span><span class="s2">&quot;query_name&quot;</span><span class="p">,</span> <span class="s2">&quot;target_name&quot;</span><span class="p">,</span> <span class="s2">&quot;mapping_matches&quot;</span><span class="p">]]</span>

        <span class="n">grouped</span> <span class="o">=</span> <span class="n">paf3</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;query_name&quot;</span><span class="p">)</span>

        <span class="n">selected_reads</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">name</span><span class="p">,</span> <span class="n">group</span> <span class="ow">in</span> <span class="n">grouped</span><span class="p">:</span>

            <span class="c1"># not all reads align to the reference:</span>

            <span class="k">if</span> <span class="n">reference_name</span> <span class="ow">in</span> <span class="n">group</span><span class="p">[</span><span class="s2">&quot;target_name&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">:</span>

                <span class="n">max_matches</span> <span class="o">=</span> <span class="n">group</span><span class="p">[</span><span class="s2">&quot;mapping_matches&quot;</span><span class="p">]</span><span class="o">.</span><span class="kp">max</span><span class="p">()</span>

                <span class="c1"># filter that the read aligns best to this reference:</span>

                <span class="k">if</span> <span class="p">(</span>

                    <span class="nb">max</span><span class="p">(</span>

                        <span class="n">group</span><span class="p">[</span><span class="n">group</span><span class="p">[</span><span class="s2">&quot;target_name&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">reference_name</span><span class="p">][</span>

                            <span class="s2">&quot;mapping_matches&quot;</span>

                        <span class="p">]</span><span class="o">.</span><span class="n">values</span>

                    <span class="p">)</span>

                    <span class="o">==</span> <span class="n">max_matches</span>

                <span class="p">):</span>

                    <span class="n">selected_reads</span> <span class="o">+=</span> <span class="p">[</span><span class="n">name</span><span class="p">]</span>

        <span class="c1"># Filter for selected reads ..</span>

        <span class="n">subset_paf_filtered</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">paf</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">paf</span><span class="p">[</span><span class="s2">&quot;query_name&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">selected_reads</span><span class="p">)]</span>

        <span class="c1"># ..then get only the reads aligning to reference:</span>

        <span class="n">subset_paf_final</span> <span class="o">=</span> <span class="n">subset_paf_filtered</span><span class="p">[</span>

            <span class="n">subset_paf_filtered</span><span class="p">[</span><span class="s2">&quot;target_name&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">reference_name</span>

        <span class="p">]</span>

        <span class="k">return</span> <span class="n">subset_paf_final</span>

    <span class="k">def</span> <span class="nf">perform_all_comparisons</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

        <span class="sd">&quot;&quot;&quot;Call perform_comparison() on each Comparator.</span>

<span class="sd">        **Parameters**</span>

<span class="sd">        **assembly_paths**</span>

<span class="sd">        &gt; Dictionary of construct name: consensus sequence file.</span>

<span class="sd">        For example: `{&quot;Construct_1&quot;: &quot;/path/to/assembly.fa&quot;}`.</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">assembly_paths</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>

            <span class="n">assembly_paths</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="k">else</span><span class="p">:</span>

            <span class="n">assembly_paths</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">assembly_paths</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">vcf_paths</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>

            <span class="n">vcf_paths</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="k">else</span><span class="p">:</span>

            <span class="n">vcf_paths</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">vcf_paths</span>

        <span class="k">for</span> <span class="n">comparator</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">comparators</span><span class="p">:</span>

            <span class="k">try</span><span class="p">:</span>

                <span class="n">assembly_path</span> <span class="o">=</span> <span class="n">assembly_paths</span><span class="p">[</span><span class="n">comparator</span><span class="o">.</span><span class="n">name</span><span class="p">]</span>

            <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>

                <span class="n">assembly_path</span> <span class="o">=</span> <span class="kc">None</span>

            <span class="k">try</span><span class="p">:</span>

                <span class="n">vcf_path</span> <span class="o">=</span> <span class="n">vcf_paths</span><span class="p">[</span><span class="n">comparator</span><span class="o">.</span><span class="n">name</span><span class="p">]</span>

            <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>

                <span class="n">vcf_path</span> <span class="o">=</span> <span class="kc">None</span>

            <span class="n">comparator</span><span class="o">.</span><span class="n">perform_comparison</span><span class="p">(</span>

                <span class="n">assembly_path</span><span class="o">=</span><span class="n">assembly_path</span><span class="p">,</span> <span class="n">vcf_path</span><span class="o">=</span><span class="n">vcf_path</span>

            <span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">comparisons_performed</span> <span class="o">=</span> <span class="kc">True</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">summarise_analysis</span><span class="p">()</span>

    <span class="k">def</span> <span class="nf">summarise_analysis</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

        <span class="sd">&quot;&quot;&quot;Create variables for the PDF report summary.&quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">number_of_constructs</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">comparators</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">result_good</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">result_warning</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">result_error</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">result_uncertain</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="n">names</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="n">reference_lengths</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="n">results</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="n">number_of_reads_aligning</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="n">median_coverages</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">comparator</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">comparators</span><span class="p">:</span>

            <span class="n">names</span> <span class="o">+=</span> <span class="p">[</span><span class="n">comparator</span><span class="o">.</span><span class="n">name</span><span class="p">]</span>

            <span class="n">reference_lengths</span> <span class="o">+=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">comparator</span><span class="o">.</span><span class="n">reference_length</span><span class="p">)]</span>

            <span class="n">number_of_reads_aligning</span> <span class="o">+=</span> <span class="p">[</span>

                <span class="nb">str</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">comparator</span><span class="o">.</span><span class="n">paf</span><span class="p">[</span><span class="s2">&quot;query_name&quot;</span><span class="p">]</span><span class="o">.</span><span class="kp">unique</span><span class="p">()))</span>

            <span class="p">]</span>

            <span class="n">median_coverages</span> <span class="o">+=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">comparator</span><span class="o">.</span><span class="n">median_yy</span><span class="p">))]</span>

            <span class="k">if</span> <span class="n">comparator</span><span class="o">.</span><span class="n">is_uncertain</span><span class="p">:</span>

                <span class="bp">self</span><span class="o">.</span><span class="n">result_uncertain</span> <span class="o">+=</span> <span class="mi">1</span>

                <span class="n">comparator</span><span class="o">.</span><span class="n">result</span> <span class="o">=</span> <span class="n">result_keywords</span><span class="p">[</span><span class="s2">&quot;uncertain&quot;</span><span class="p">]</span>

            <span class="k">else</span><span class="p">:</span>

                <span class="k">if</span> <span class="n">comparator</span><span class="o">.</span><span class="n">has_errors</span><span class="p">:</span>

                    <span class="bp">self</span><span class="o">.</span><span class="n">result_error</span> <span class="o">+=</span> <span class="mi">1</span>

                    <span class="n">comparator</span><span class="o">.</span><span class="n">result</span> <span class="o">=</span> <span class="n">result_keywords</span><span class="p">[</span><span class="s2">&quot;error&quot;</span><span class="p">]</span>

                <span class="k">else</span><span class="p">:</span>

                    <span class="k">if</span> <span class="n">comparator</span><span class="o">.</span><span class="n">has_warnings</span><span class="p">:</span>

                        <span class="bp">self</span><span class="o">.</span><span class="n">result_warning</span> <span class="o">+=</span> <span class="mi">1</span>

                        <span class="n">comparator</span><span class="o">.</span><span class="n">result</span> <span class="o">=</span> <span class="n">result_keywords</span><span class="p">[</span><span class="s2">&quot;warning&quot;</span><span class="p">]</span>

                    <span class="k">else</span><span class="p">:</span>

                        <span class="bp">self</span><span class="o">.</span><span class="n">result_good</span> <span class="o">+=</span> <span class="mi">1</span>

                        <span class="n">comparator</span><span class="o">.</span><span class="n">result</span> <span class="o">=</span> <span class="n">result_keywords</span><span class="p">[</span><span class="s2">&quot;good&quot;</span><span class="p">]</span>

            <span class="n">results</span> <span class="o">+=</span> <span class="p">[</span><span class="n">comparator</span><span class="o">.</span><span class="n">result</span><span class="p">]</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">n_fastq_reads</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">paf</span><span class="o">.</span><span class="n">query_name</span><span class="p">))</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">fastq_plot</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">plot_fastq_histogram</span><span class="p">()</span>

        <span class="c1"># Summary table</span>

        <span class="n">d</span> <span class="o">=</span> <span class="p">{</span>

            <span class="s2">&quot;Name&quot;</span><span class="p">:</span> <span class="n">pandas</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">names</span><span class="p">),</span>

            <span class="s2">&quot;Result&quot;</span><span class="p">:</span> <span class="n">pandas</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">results</span><span class="p">),</span>

            <span class="s2">&quot;Length [bp]&quot;</span><span class="p">:</span> <span class="n">pandas</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">reference_lengths</span><span class="p">),</span>

            <span class="s2">&quot;FASTQ reads&quot;</span><span class="p">:</span> <span class="n">pandas</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">number_of_reads_aligning</span><span class="p">),</span>

            <span class="s2">&quot;Coverage [x]&quot;</span><span class="p">:</span> <span class="n">pandas</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">median_coverages</span><span class="p">),</span>

        <span class="p">}</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">summary_table</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">d</span><span class="p">)</span>

        <span class="c1"># plt.close(&quot;all&quot;)</span>

    <span class="k">def</span> <span class="nf">plot_fastq_histogram</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">n_bins</span><span class="o">=</span><span class="mi">50</span><span class="p">):</span>

        <span class="sd">&quot;&quot;&quot;Plot a histogram of the FASTQ reads.</span>

<span class="sd">        **Parameters**</span>

<span class="sd">        **n_bins**</span>

<span class="sd">        &gt; Number of bins in the histogram (`int`).</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">()</span>

        <span class="n">plt</span><span class="o">.</span><span class="n">hist</span><span class="p">(</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">paf</span><span class="p">[</span><span class="s2">&quot;query_length&quot;</span><span class="p">],</span>

            <span class="n">bins</span><span class="o">=</span><span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">paf</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;target_length&quot;</span><span class="p">]</span> <span class="o">/</span> <span class="n">n_bins</span><span class="p">),</span>

            <span class="n">alpha</span><span class="o">=</span><span class="mf">0.8</span><span class="p">,</span>

        <span class="p">)</span>

        <span class="n">ax</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Number of reads&quot;</span><span class="p">)</span>

        <span class="n">ax</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Read length [bp]&quot;</span><span class="p">)</span>

        <span class="c1"># Mark expected lengths for interpreting the plot:</span>

        <span class="k">for</span> <span class="n">comparator</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">comparators</span><span class="p">:</span>

            <span class="n">plt</span><span class="o">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">comparator</span><span class="o">.</span><span class="n">record</span><span class="p">),</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;#fd5a31&quot;</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">fig</span>

    <span class="nd">@staticmethod</span>

    <span class="k">def</span> <span class="nf">load_paf</span><span class="p">(</span><span class="n">paf_path</span><span class="p">):</span>

        <span class="sd">&quot;&quot;&quot;Create a dataframe from a PAF file of alignments.</span>

<span class="sd">        **Parameters**</span>

<span class="sd">        **paf_path**</span>

<span class="sd">        &gt; Path (`str`) to PAF file: `paftools.js sam2paf aln.sam &gt; aln.paf`.</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">try</span><span class="p">:</span>

            <span class="n">paf</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">paf_path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>

        <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>  <span class="c1"># unequal number of columns, the first 12 is assumed to be OK</span>

            <span class="c1"># The last column is CIGAR, so we need to read lines one by one:</span>

            <span class="n">rows</span> <span class="o">=</span> <span class="p">[]</span>

            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">paf_path</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f_input</span><span class="p">:</span>

                <span class="n">paf_lines</span> <span class="o">=</span> <span class="n">f_input</span><span class="o">.</span><span class="n">read</span><span class="p">()</span><span class="o">.</span><span class="n">splitlines</span><span class="p">()</span>

                <span class="k">for</span> <span class="n">paf_line</span> <span class="ow">in</span> <span class="n">paf_lines</span><span class="p">:</span>

                    <span class="n">paf_line</span> <span class="o">=</span> <span class="n">paf_line</span><span class="o">.</span><span class="kp">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>  <span class="c1"># PAF is tab separated</span>

                    <span class="n">row</span> <span class="o">=</span> <span class="n">paf_line</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">12</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="s2">&quot;;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">paf_line</span><span class="p">[</span><span class="mi">12</span><span class="p">:</span><span class="o">-</span><span class="mi">1</span><span class="p">])]</span> <span class="o">+</span> <span class="p">[</span><span class="n">paf_line</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]]</span>

                    <span class="n">rows</span> <span class="o">+=</span> <span class="p">[</span><span class="n">row</span><span class="p">]</span>

            <span class="n">paf</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">rows</span><span class="p">)</span>

            <span class="c1"># Set numeric columns:</span>

            <span class="n">numeric_columns</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">6</span><span class="p">,</span> <span class="mi">7</span><span class="p">,</span> <span class="mi">8</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">11</span><span class="p">]</span>  <span class="c1"># see list below</span>

            <span class="n">paf</span><span class="p">[</span><span class="n">numeric_columns</span><span class="p">]</span> <span class="o">=</span> <span class="n">paf</span><span class="p">[</span><span class="n">numeric_columns</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="n">pandas</span><span class="o">.</span><span class="n">to_numeric</span><span class="p">)</span>

        <span class="c1"># First 12 columns are defined by the format:</span>

        <span class="n">columns_12</span> <span class="o">=</span> <span class="p">[</span>

            <span class="s2">&quot;query_name&quot;</span><span class="p">,</span>

            <span class="s2">&quot;query_length&quot;</span><span class="p">,</span>

            <span class="s2">&quot;query_start&quot;</span><span class="p">,</span>

            <span class="s2">&quot;query_end&quot;</span><span class="p">,</span>

            <span class="s2">&quot;strand&quot;</span><span class="p">,</span>

            <span class="s2">&quot;target_name&quot;</span><span class="p">,</span>

            <span class="s2">&quot;target_length&quot;</span><span class="p">,</span>

            <span class="s2">&quot;target_start&quot;</span><span class="p">,</span>

            <span class="s2">&quot;target_end&quot;</span><span class="p">,</span>

            <span class="s2">&quot;mapping_matches&quot;</span><span class="p">,</span>

            <span class="s2">&quot;mapping_size&quot;</span><span class="p">,</span>

            <span class="s2">&quot;mapping_quality&quot;</span><span class="p">,</span>

        <span class="p">]</span>

        <span class="c1"># For additional columns optionally created during alignment:</span>

        <span class="n">columns_13plus</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">index</span><span class="p">)</span> <span class="k">for</span> <span class="n">index</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">paf</span><span class="o">.</span><span class="n">columns</span><span class="p">))]</span>

        <span class="n">paf</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="n">columns_12</span> <span class="o">+</span> <span class="n">columns_13plus</span>

        <span class="k">return</span> <span class="n">paf</span>

    <span class="nd">@staticmethod</span>

    <span class="k">def</span> <span class="nf">load_tsv</span><span class="p">(</span><span class="n">tsv_path</span><span class="p">):</span>

        <span class="sd">&quot;&quot;&quot;Create a dataframe from a TSV file of depth counts.</span>

<span class="sd">        **Parameters**</span>

<span class="sd">        **tsv_path**</span>

<span class="sd">        &gt; Path to output of `samtools depth -aa` (`str`).</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">tsv</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">tsv_path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>

        <span class="n">tsv</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">,</span> <span class="s2">&quot;position&quot;</span><span class="p">,</span> <span class="s2">&quot;depth&quot;</span><span class="p">]</span>

        <span class="k">return</span> <span class="n">tsv</span>

<span class="k">class</span> <span class="nc">Comparator</span><span class="p">:</span>

    <span class="sd">&quot;&quot;&quot;Analyse alignments to a reference.</span>

<span class="sd">    **Parameters**</span>

<span class="sd">    **reference**</span>

<span class="sd">    &gt; Dictionary of one name: one Biopython record.</span>

<span class="sd">    **alignment**</span>

<span class="sd">    &gt; Dictionary of pandas dataframes: {&quot;paf&quot;: df_paf, &quot;tsv&quot;: df_tsv}. The dataframes</span>

<span class="sd">    contain entries only for the reference.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">reference</span><span class="p">,</span> <span class="n">alignment</span><span class="p">):</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="nb">iter</span><span class="p">(</span><span class="n">reference</span><span class="p">))</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">record</span> <span class="o">=</span> <span class="nb">next</span><span class="p">(</span><span class="nb">iter</span><span class="p">(</span><span class="n">reference</span><span class="o">.</span><span class="n">values</span><span class="p">()))</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">paf</span> <span class="o">=</span> <span class="n">alignment</span><span class="p">[</span><span class="s2">&quot;paf&quot;</span><span class="p">]</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">tsv</span> <span class="o">=</span> <span class="n">alignment</span><span class="p">[</span><span class="s2">&quot;tsv&quot;</span><span class="p">]</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">coverage_cutoff</span> <span class="o">=</span> <span class="mf">0.5</span>  <span class="c1"># fraction of median coverage</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">coverage_cutoff_pct</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">coverage_cutoff</span> <span class="o">*</span> <span class="mi">100</span><span class="p">)</span>  <span class="c1"># for PDF report</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">reference_length</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">record</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">has_warnings</span> <span class="o">=</span> <span class="kc">False</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">has_errors</span> <span class="o">=</span> <span class="kc">False</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">geneblocks_outcome</span> <span class="o">=</span> <span class="s2">&quot;none&quot;</span>  <span class="c1"># stores outcome, used in PDF report making</span>

        <span class="c1"># Set True if references are DNA Cauldron-simulated files:</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">dnacauldron</span> <span class="o">=</span> <span class="kc">True</span>  <span class="c1"># for plotting</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">has_consensus</span> <span class="o">=</span> <span class="kc">False</span>  <span class="c1"># for consensus sequence comparison</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">is_uncertain</span> <span class="o">=</span> <span class="kc">True</span>  <span class="c1"># for sequencing data quality, used in report</span>

    <span class="k">def</span> <span class="nf">perform_comparison</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">assembly_path</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">vcf_path</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>

        <span class="sd">&quot;&quot;&quot;Plot coverage and compare reference with a consensus sequence.</span>

<span class="sd">        **Parameters**</span>

<span class="sd">        **assembly_path**</span>

<span class="sd">        &gt; Optional. Path to a consensus FASTA file (`str`).</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">fig</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">plot_coverage</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">assembly_path</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">comparison_figure</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">compare_with_assembly</span><span class="p">(</span>

                <span class="n">assembly_path</span><span class="o">=</span><span class="n">assembly_path</span>

            <span class="p">)</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">has_consensus</span> <span class="o">=</span> <span class="kc">True</span>

        <span class="k">if</span> <span class="n">vcf_path</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">vcf_table</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">create_vcf_table</span><span class="p">(</span><span class="n">vcf_path</span><span class="o">=</span><span class="n">vcf_path</span><span class="p">)</span>

        <span class="c1"># plt.close(&quot;all&quot;)</span>

    <span class="k">def</span> <span class="nf">create_vcf_table</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">vcf_path</span><span class="p">):</span>

        <span class="n">vcf_dict</span> <span class="o">=</span> <span class="p">{</span><span class="n">key</span><span class="p">:</span> <span class="p">[]</span> <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">vcf_selected_info_keytable</span><span class="o">.</span><span class="n">keys</span><span class="p">()}</span>

        <span class="n">vcf_dict</span><span class="p">[</span><span class="s2">&quot;REF&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="n">vcf_dict</span><span class="p">[</span><span class="s2">&quot;ALT&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="n">vcf_dict</span><span class="p">[</span><span class="s2">&quot;LOC&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">variant</span> <span class="ow">in</span> <span class="n">cyvcf2</span><span class="o">.</span><span class="n">VCF</span><span class="p">(</span><span class="n">vcf_path</span><span class="p">):</span>

            <span class="n">vcf_dict</span><span class="p">[</span><span class="s2">&quot;REF&quot;</span><span class="p">]</span> <span class="o">+=</span> <span class="p">[</span><span class="n">variant</span><span class="o">.</span><span class="n">REF</span><span class="p">]</span>

            <span class="n">vcf_dict</span><span class="p">[</span><span class="s2">&quot;ALT&quot;</span><span class="p">]</span> <span class="o">+=</span> <span class="p">[</span><span class="n">variant</span><span class="o">.</span><span class="n">ALT</span><span class="p">]</span>

            <span class="n">vcf_dict</span><span class="p">[</span><span class="s2">&quot;LOC&quot;</span><span class="p">]</span> <span class="o">+=</span> <span class="p">[</span><span class="n">variant</span><span class="o">.</span><span class="n">start</span><span class="p">]</span>

            <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">vcf_selected_info_keytable</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>

                <span class="n">vcf_dict</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">+=</span> <span class="p">[</span><span class="n">variant</span><span class="o">.</span><span class="n">INFO</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">key</span><span class="p">)]</span>

        <span class="n">vcf_table</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>

            <span class="n">vcf_dict</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;LOC&quot;</span><span class="p">,</span> <span class="s2">&quot;REF&quot;</span><span class="p">,</span> <span class="s2">&quot;ALT&quot;</span><span class="p">,</span> <span class="s2">&quot;TYPE&quot;</span><span class="p">,</span> <span class="s2">&quot;DP&quot;</span><span class="p">,</span> <span class="s2">&quot;RO&quot;</span><span class="p">,</span> <span class="s2">&quot;AO&quot;</span><span class="p">]</span>

        <span class="p">)</span>  <span class="c1"># also set the order of the columns</span>

        <span class="k">return</span> <span class="n">vcf_table</span>

    <span class="k">def</span> <span class="nf">find_big_inserts</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">threshold</span><span class="o">=</span><span class="mi">100</span><span class="p">):</span>

        <span class="sd">&quot;&quot;&quot;Calculate % of reads with big inserts.</span>

<span class="sd">        **Parameters**</span>

<span class="sd">        **threshold**</span>

<span class="sd">        &gt; Size of insert in bp to be considered big (`int`).&quot;&quot;&quot;</span>

        <span class="n">read_insert_dict</span> <span class="o">=</span> <span class="p">{</span>

            <span class="n">read_id</span><span class="p">:</span> <span class="kc">False</span> <span class="k">for</span> <span class="n">read_id</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">paf</span><span class="p">[</span><span class="s2">&quot;query_name&quot;</span><span class="p">]</span><span class="o">.</span><span class="kp">unique</span><span class="p">()</span>

        <span class="p">}</span>  <span class="c1"># Initialize. Value True when read has insert.</span>

        <span class="k">for</span> <span class="n">index</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">paf</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>

            <span class="n">read_id</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;query_name&quot;</span><span class="p">]</span>

            <span class="n">cigar</span> <span class="o">=</span> <span class="n">row</span><span class="o">.</span><span class="n">tail</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="kp">item</span><span class="p">()</span>  <span class="c1"># CIGAR string column is the last</span>

            <span class="k">for</span> <span class="kp">size</span><span class="p">,</span> <span class="n">mutation</span> <span class="ow">in</span> <span class="n">re</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;(\d+)([A-Z]</span><span class="si">{1}</span><span class="s2">)&quot;</span><span class="p">,</span> <span class="n">cigar</span><span class="p">):</span>

                <span class="k">if</span> <span class="nb">int</span><span class="p">(</span><span class="kp">size</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">threshold</span><span class="p">:</span>

                    <span class="k">if</span> <span class="n">mutation</span> <span class="o">==</span> <span class="s2">&quot;I&quot;</span><span class="p">:</span>  <span class="c1"># CIGAR code for insert</span>

                        <span class="n">read_insert_dict</span><span class="p">[</span><span class="n">read_id</span><span class="p">]</span> <span class="o">=</span> <span class="kc">True</span>

                        <span class="k">break</span>

        <span class="n">fraction_of_big_inserts</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="n">read_insert_dict</span><span class="o">.</span><span class="n">values</span><span class="p">())</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span>

            <span class="n">read_insert_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>

        <span class="p">)</span>

        <span class="c1"># imprecise rounding, but ok for our purposes:</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">pct_big_insert</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="n">fraction_of_big_inserts</span> <span class="o">*</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>  <span class="c1"># pct multiplier</span>

    <span class="k">def</span> <span class="nf">calculate_stats</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

        <span class="sd">&quot;&quot;&quot;Calculate statistics for the coverage plot, used in plot_coverage().&quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">xx</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="kp">arange</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">))</span>  <span class="c1"># for the plot x axis</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">yy</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">tsv</span><span class="p">[</span><span class="s2">&quot;depth&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">()</span>  <span class="c1"># for plotting coverage</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">median_yy</span> <span class="o">=</span> <span class="n">statistics</span><span class="o">.</span><span class="kp">median</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">yy</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">median_yy</span> <span class="o">&lt;</span> <span class="mi">30</span><span class="p">:</span>  <span class="c1"># a good cutoff for low-depth sequencing</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">has_low_coverage</span> <span class="o">=</span> <span class="kc">True</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">has_warnings</span> <span class="o">=</span> <span class="kc">True</span>

        <span class="k">else</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">is_uncertain</span> <span class="o">=</span> <span class="kc">False</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">has_low_coverage</span> <span class="o">=</span> <span class="kc">False</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">find_big_inserts</span><span class="p">()</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">pct_big_insert</span> <span class="o">&gt;=</span> <span class="mi">50</span><span class="p">:</span>  <span class="c1"># at least half of reads have big insert</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">has_errors</span> <span class="o">=</span> <span class="kc">True</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">has_big_insert</span> <span class="o">=</span> <span class="kc">True</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">pct_big_insert</span> <span class="o">&gt;=</span> <span class="mi">5</span><span class="p">:</span>  <span class="c1"># five pct has big insert</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">has_warnings</span> <span class="o">=</span> <span class="kc">True</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">has_reads_with_insert</span> <span class="o">=</span> <span class="kc">True</span>

        <span class="k">else</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">has_big_insert</span> <span class="o">=</span> <span class="kc">False</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">has_reads_with_insert</span> <span class="o">=</span> <span class="kc">False</span>

        <span class="c1"># This section creates a list of low coverage position to be reported</span>

        <span class="kp">indices</span> <span class="o">=</span> <span class="p">[</span>

            <span class="n">i</span>

            <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">yy</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">value</span> <span class="o">&lt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">median_yy</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">coverage_cutoff</span>

        <span class="p">]</span>

        <span class="n">G</span> <span class="o">=</span> <span class="p">(</span>

            <span class="nb">list</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>

            <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">itertools</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span>

                <span class="kp">indices</span><span class="p">,</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="n">itertools</span><span class="o">.</span><span class="n">count</span><span class="p">():</span> <span class="nb">next</span><span class="p">(</span><span class="n">c</span><span class="p">)</span> <span class="o">-</span> <span class="n">x</span>

            <span class="p">)</span>

        <span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">low_coverage_positions_string</span> <span class="o">=</span> <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>

            <span class="s2">&quot;-&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="nb">str</span><span class="p">,</span> <span class="p">(</span><span class="n">g</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">g</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])[:</span> <span class="nb">len</span><span class="p">(</span><span class="n">g</span><span class="p">)]))</span> <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">G</span>

        <span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">low_coverage_positions_string</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">low_coverage_positions_string</span> <span class="o">=</span> <span class="s2">&quot;-&quot;</span>  <span class="c1"># looks better in the pdf report</span>

        <span class="k">else</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">has_warnings</span> <span class="o">=</span> <span class="kc">True</span>

    <span class="k">def</span> <span class="nf">plot_coverage</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

        <span class="sd">&quot;&quot;&quot;Plot the reference with the coverage and weak reads.&quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="nb">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s2">&quot;xx&quot;</span><span class="p">):</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">calculate_stats</span><span class="p">()</span>  <span class="c1"># also calculates yy and zz</span>

        <span class="c1"># Plot</span>

        <span class="c1">######</span>

        <span class="n">fig</span><span class="p">,</span> <span class="p">(</span><span class="n">ax1</span><span class="p">,</span> <span class="n">ax2</span><span class="p">)</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span>

            <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">7</span><span class="p">,</span> <span class="mi">3</span><span class="p">),</span> <span class="n">sharex</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">gridspec_kw</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;height_ratios&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">4</span><span class="p">,</span> <span class="mi">1</span><span class="p">]}</span>

        <span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">dnacauldron</span><span class="p">:</span>

            <span class="n">graphic_record</span> <span class="o">=</span> <span class="n">CustomTranslator</span><span class="p">()</span><span class="o">.</span><span class="n">translate_record</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">record</span><span class="p">)</span>

        <span class="k">else</span><span class="p">:</span>

            <span class="n">graphic_record</span> <span class="o">=</span> <span class="n">dna_features_viewer</span><span class="o">.</span><span class="n">BiopythonTranslator</span><span class="p">()</span><span class="o">.</span><span class="n">translate_record</span><span class="p">(</span>

                <span class="bp">self</span><span class="o">.</span><span class="n">record</span>

            <span class="p">)</span>

        <span class="n">graphic_record</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">ax</span><span class="o">=</span><span class="n">ax1</span><span class="p">,</span> <span class="n">with_ruler</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">strand_in_label_threshold</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>

        <span class="c1"># Plot coverage</span>

        <span class="n">ax2</span><span class="o">.</span><span class="n">fill_between</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">xx</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">yy</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.8</span><span class="p">)</span>

        <span class="n">stdev_yy</span> <span class="o">=</span> <span class="n">statistics</span><span class="o">.</span><span class="n">stdev</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">yy</span><span class="p">)</span>

        <span class="n">ax2</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="n">bottom</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

        <span class="n">ax2</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Depth&quot;</span><span class="p">)</span>

        <span class="n">ax2</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Base [bp]&quot;</span><span class="p">)</span>

        <span class="n">ax2</span><span class="o">.</span><span class="n">axhline</span><span class="p">(</span>

            <span class="n">y</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">median_yy</span><span class="p">,</span> <span class="n">xmin</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">xmax</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;grey&quot;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mf">0.8</span>

        <span class="p">)</span>

        <span class="n">ax2</span><span class="o">.</span><span class="n">axhline</span><span class="p">(</span>

            <span class="n">y</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">median_yy</span> <span class="o">-</span> <span class="n">stdev_yy</span><span class="p">,</span>

            <span class="n">xmin</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>

            <span class="n">xmax</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>

            <span class="n">color</span><span class="o">=</span><span class="s2">&quot;grey&quot;</span><span class="p">,</span>

            <span class="n">linestyle</span><span class="o">=</span><span class="s2">&quot;--&quot;</span><span class="p">,</span>

            <span class="n">linewidth</span><span class="o">=</span><span class="mf">0.8</span><span class="p">,</span>

        <span class="p">)</span>

        <span class="n">ax2</span><span class="o">.</span><span class="n">axhline</span><span class="p">(</span>

            <span class="n">y</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">median_yy</span> <span class="o">+</span> <span class="n">stdev_yy</span><span class="p">,</span>

            <span class="n">xmin</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>

            <span class="n">xmax</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>

            <span class="n">color</span><span class="o">=</span><span class="s2">&quot;grey&quot;</span><span class="p">,</span>

            <span class="n">linestyle</span><span class="o">=</span><span class="s2">&quot;--&quot;</span><span class="p">,</span>

            <span class="n">linewidth</span><span class="o">=</span><span class="mf">0.8</span><span class="p">,</span>

        <span class="p">)</span>

        <span class="k">return</span> <span class="n">fig</span>

    <span class="k">def</span> <span class="nf">get_weak_read_histogram_data</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cutoff</span><span class="p">):</span>

        <span class="sd">&quot;&quot;&quot;Create data for the plotting function.</span>

<span class="sd">        **Parameters**</span>

<span class="sd">        **cutoff**</span>

<span class="sd">        &gt; Cutoff for proportion of bases that map (`float`). Reads below the cutoff are</span>

<span class="sd">        viewed as weak reads. Recommended value: 0.8 (80%).</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># For each read, calculate the proportion of bases that map, then create an array</span>

        <span class="c1"># for each read that is below the cutoff. The sum of these arrays will give the</span>

        <span class="c1"># depth of aligning bad reads for each base of the reference.</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">proportion_maps</span> <span class="o">=</span> <span class="p">(</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">paf</span><span class="p">[</span><span class="s2">&quot;mapping_matches&quot;</span><span class="p">]</span> <span class="o">/</span> <span class="bp">self</span><span class="o">.</span><span class="n">paf</span><span class="p">[</span><span class="s2">&quot;query_length&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">()</span>

        <span class="p">)</span>

        <span class="n">filtered_paf</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">paf</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">proportion_maps</span> <span class="o">&lt;</span> <span class="n">cutoff</span><span class="p">]</span>

        <span class="n">array_length</span> <span class="o">=</span> <span class="n">filtered_paf</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;target_length&quot;</span><span class="p">]</span>

        <span class="n">histogram_array</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">index</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">filtered_paf</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>

            <span class="n">read_array</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="n">array_length</span>

            <span class="n">start</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;target_start&quot;</span><span class="p">]</span>

            <span class="n">end</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;target_end&quot;</span><span class="p">]</span>

            <span class="n">value</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">/</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;multialign&quot;</span><span class="p">]</span>

            <span class="n">values</span> <span class="o">=</span> <span class="p">[</span><span class="n">value</span><span class="p">]</span> <span class="o">*</span> <span class="p">(</span><span class="n">end</span> <span class="o">-</span> <span class="n">start</span><span class="p">)</span>

            <span class="n">read_array</span><span class="p">[</span><span class="n">start</span><span class="p">:</span><span class="n">end</span><span class="p">]</span> <span class="o">=</span> <span class="n">values</span>

            <span class="n">histogram_array</span><span class="o">.</span><span class="kp">append</span><span class="p">(</span><span class="n">read_array</span><span class="p">)</span>

        <span class="n">histogram_numpy_array</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="kp">array</span><span class="p">(</span><span class="n">histogram_array</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">zz</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="kp">sum</span><span class="p">(</span><span class="n">histogram_numpy_array</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">compare_with_assembly</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">assembly_path</span><span class="p">):</span>

        <span class="sd">&quot;&quot;&quot;Compare the reference with a consensus sequence.</span>

<span class="sd">        Check length, orientation (sense or reverse complement), then use GeneBlocks to</span>

<span class="sd">        highlight differences between reference and the plasmid assembly.</span>

<span class="sd">        **Parameters**</span>

<span class="sd">        **assembly_path**</span>

<span class="sd">        &gt; Path to a consensus FASTA file (`str`).</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Note that in this context &#39;assembly&#39; means a consensus of variant calls</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">assembly</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="o">=</span><span class="n">assembly_path</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s2">&quot;fasta&quot;</span><span class="p">)</span>

        <span class="n">length_ratio</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">assembly</span><span class="p">)</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">record</span><span class="p">)</span>

        <span class="k">if</span> <span class="mf">0.95</span> <span class="o">&lt;</span> <span class="n">length_ratio</span> <span class="o">&lt;</span> <span class="mf">1.05</span><span class="p">:</span>  <span class="c1"># length within reason</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">is_length_ok</span> <span class="o">=</span> <span class="kc">True</span>  <span class="c1"># can compare with expected sequence</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">incorrect_length_msg</span> <span class="o">=</span> <span class="kc">None</span>

        <span class="k">else</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">geneblocks_outcome</span> <span class="o">=</span> <span class="s2">&quot;incorrect_length&quot;</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">is_length_ok</span> <span class="o">=</span> <span class="kc">False</span>  <span class="c1"># do not compare sequences</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">has_errors</span> <span class="o">=</span> <span class="kc">True</span>

            <span class="n">difference_bases</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">assembly</span><span class="p">)</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">record</span><span class="p">)</span>

            <span class="n">difference_percent</span> <span class="o">=</span> <span class="p">(</span><span class="n">length_ratio</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="mi">100</span>  <span class="c1"># get as percent</span>

            <span class="k">if</span> <span class="n">difference_percent</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>

                <span class="n">difference_text</span> <span class="o">=</span> <span class="s2">&quot;longer&quot;</span>

            <span class="k">else</span><span class="p">:</span>

                <span class="n">difference_text</span> <span class="o">=</span> <span class="s2">&quot;shorter&quot;</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">incorrect_length_msg</span> <span class="o">=</span> <span class="p">(</span>

                <span class="s2">&quot;The consensus is </span><span class="si">%d%%</span><span class="s2"> (</span><span class="si">%d</span><span class="s2"> bp) </span><span class="si">%s</span><span class="s2"> than the reference.&quot;</span>

                <span class="o">%</span> <span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="nb">abs</span><span class="p">(</span><span class="n">difference_percent</span><span class="p">)),</span> <span class="n">difference_bases</span><span class="p">,</span> <span class="n">difference_text</span><span class="p">)</span>

            <span class="p">)</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">geneblocks_done</span> <span class="o">=</span> <span class="kc">False</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">is_diffblocks_reverse</span> <span class="o">=</span> <span class="kc">None</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">is_comparison_successful</span> <span class="o">=</span> <span class="kc">False</span>  <span class="c1"># for PDF reporting</span>

            <span class="k">return</span>

        <span class="c1"># To find out the orientation of the assembly, we compare using Levenshtein</span>

        <span class="n">lev_distance</span> <span class="o">=</span> <span class="n">lev</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">assembly</span><span class="o">.</span><span class="n">seq</span><span class="p">),</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">))</span>

        <span class="n">lev_cutoff</span> <span class="o">=</span> <span class="mi">50</span>  <span class="c1"># more than this is too much to plot. Also for most plasmids,</span>

        <span class="c1"># this represents ~1% difference, which is too much error.</span>

        <span class="n">assembly_for_diffblocks</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">assembly</span>

        <span class="k">if</span> <span class="n">lev_distance</span> <span class="o">&gt;</span> <span class="n">lev_cutoff</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">perform_geneblocks</span> <span class="o">=</span> <span class="kc">False</span>

        <span class="k">else</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">perform_geneblocks</span> <span class="o">=</span> <span class="kc">True</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">perform_geneblocks</span><span class="p">:</span>

            <span class="k">try</span><span class="p">:</span>

                <span class="n">diff_blocks</span> <span class="o">=</span> <span class="n">geneblocks</span><span class="o">.</span><span class="n">DiffBlocks</span><span class="o">.</span><span class="n">from_sequences</span><span class="p">(</span>

                    <span class="n">assembly_for_diffblocks</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">record</span>

                <span class="p">)</span>

            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>

                <span class="bp">self</span><span class="o">.</span><span class="n">geneblocks_done</span> <span class="o">=</span> <span class="kc">False</span>

                <span class="bp">self</span><span class="o">.</span><span class="n">geneblocks_outcome</span> <span class="o">=</span> <span class="s2">&quot;geneblocks_error&quot;</span>

            <span class="k">else</span><span class="p">:</span>  <span class="c1"># diffblocks had no error</span>

                <span class="bp">self</span><span class="o">.</span><span class="n">geneblocks_done</span> <span class="o">=</span> <span class="kc">True</span>

                <span class="n">ax1</span><span class="p">,</span> <span class="n">ax2</span> <span class="o">=</span> <span class="n">diff_blocks</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">figure_width</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>

                <span class="bp">self</span><span class="o">.</span><span class="n">is_diffblocks_reverse</span> <span class="o">=</span> <span class="kc">False</span>

                <span class="bp">self</span><span class="o">.</span><span class="n">is_comparison_successful</span> <span class="o">=</span> <span class="kc">True</span>

                <span class="bp">self</span><span class="o">.</span><span class="n">geneblocks_outcome</span> <span class="o">=</span> <span class="s2">&quot;all_good&quot;</span>

                <span class="k">return</span> <span class="n">ax1</span>

            <span class="c1"># We try again as due to a bug in geneblocks, the reverse order may work:</span>

            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">geneblocks_outcome</span> <span class="o">==</span> <span class="s2">&quot;geneblocks_error&quot;</span><span class="p">:</span>

                <span class="k">try</span><span class="p">:</span>

                    <span class="n">diff_blocks</span> <span class="o">=</span> <span class="n">geneblocks</span><span class="o">.</span><span class="n">DiffBlocks</span><span class="o">.</span><span class="n">from_sequences</span><span class="p">(</span>

                        <span class="bp">self</span><span class="o">.</span><span class="n">record</span><span class="p">,</span> <span class="n">assembly_for_diffblocks</span>

                    <span class="p">)</span>

                <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>

                    <span class="k">return</span>

                <span class="k">else</span><span class="p">:</span>

                    <span class="bp">self</span><span class="o">.</span><span class="n">geneblocks_done</span> <span class="o">=</span> <span class="kc">True</span>  <span class="c1"># overwrite</span>

                    <span class="n">ax1</span><span class="p">,</span> <span class="n">ax2</span> <span class="o">=</span> <span class="n">diff_blocks</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">figure_width</span><span class="o">=</span><span class="mi">7</span><span class="p">)</span>

                    <span class="bp">self</span><span class="o">.</span><span class="n">is_diffblocks_reverse</span> <span class="o">=</span> <span class="kc">True</span>

                    <span class="bp">self</span><span class="o">.</span><span class="n">is_comparison_successful</span> <span class="o">=</span> <span class="kc">True</span>

                    <span class="bp">self</span><span class="o">.</span><span class="n">geneblocks_outcome</span> <span class="o">=</span> <span class="s2">&quot;swapped_diffblocks&quot;</span>

                    <span class="k">return</span> <span class="n">ax1</span>

        <span class="k">else</span><span class="p">:</span>  <span class="c1"># too many differences</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">geneblocks_done</span> <span class="o">=</span> <span class="kc">False</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">geneblocks_outcome</span> <span class="o">=</span> <span class="s2">&quot;too_many_differences&quot;</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">is_comparison_successful</span> <span class="o">=</span> <span class="kc">False</span>

    <span class="k">def</span> <span class="nf">filter_fastq</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fastq_path</span><span class="p">,</span> <span class="n">target</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>

        <span class="sd">&quot;&quot;&quot;Filter original FASTQ file for reads that best map to the reference.</span>

<span class="sd">        This function is used to obtain a FASTQ file that can be used for Canu assembly.</span>

<span class="sd">        **Parameters**</span>

<span class="sd">        **fastq_path**</span>

<span class="sd">        &gt; Path to FASTQ file (`str`).</span>

<span class="sd">        **target**</span>

<span class="sd">        &gt; Path to save the filtered FASTQ file (`str`).</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">target</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>

            <span class="n">target</span> <span class="o">=</span> <span class="n">fastq_path</span> <span class="o">+</span> <span class="s2">&quot;_filtered.fastq&quot;</span>

        <span class="n">read_names</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">paf</span><span class="p">[</span><span class="s2">&quot;query_name&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">()</span>

        <span class="n">input_seq_iterator</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">fastq_path</span><span class="p">,</span> <span class="s2">&quot;fastq&quot;</span><span class="p">)</span>

        <span class="n">seq_iterator</span> <span class="o">=</span> <span class="p">(</span>

            <span class="n">record</span> <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">input_seq_iterator</span> <span class="k">if</span> <span class="n">record</span><span class="o">.</span><span class="n">name</span> <span class="ow">in</span> <span class="n">read_names</span>

        <span class="p">)</span>

        <span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">seq_iterator</span><span class="p">,</span> <span class="n">target</span><span class="p">,</span> <span class="s2">&quot;fastq&quot;</span><span class="p">)</span>
</code></pre></div>

</details>
<h2 id="variables">Variables</h2>
<div class="codehilite"><pre><span></span><code><span class="n">result_keywords</span>
</code></pre></div>

<div class="codehilite"><pre><span></span><code><span class="n">vcf_selected_info_keytable</span>
</code></pre></div>

<h2 id="classes">Classes</h2>
<h3 id="comparator">Comparator</h3>
<div class="codehilite"><pre><span></span><code><span class="k">class</span> <span class="nc">Comparator</span><span class="p">(</span>
    <span class="n">reference</span><span class="p">,</span>
    <span class="n">alignment</span>
<span class="p">)</span>
</code></pre></div>

<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code><span class="k">class</span> <span class="n">Comparator</span><span class="p">:</span>

    <span class="sd">&quot;&quot;&quot;Analyse alignments to a reference.</span>

<span class="sd">    **Parameters**</span>

<span class="sd">    **reference**</span>

<span class="sd">    &gt; Dictionary of one name: one Biopython record.</span>

<span class="sd">    **alignment**</span>

<span class="sd">    &gt; Dictionary of pandas dataframes: {&quot;paf&quot;: df_paf, &quot;tsv&quot;: df_tsv}. The dataframes</span>

<span class="sd">    contain entries only for the reference.</span>

<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">def</span> <span class="n">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">reference</span><span class="p">,</span> <span class="n">alignment</span><span class="p">):</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">name</span> <span class="o">=</span> <span class="n">next</span><span class="p">(</span><span class="n">iter</span><span class="p">(</span><span class="n">reference</span><span class="p">))</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">record</span> <span class="o">=</span> <span class="n">next</span><span class="p">(</span><span class="n">iter</span><span class="p">(</span><span class="n">reference</span><span class="o">.</span><span class="n">values</span><span class="p">()))</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">paf</span> <span class="o">=</span> <span class="n">alignment</span><span class="p">[</span><span class="s2">&quot;paf&quot;</span><span class="p">]</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">tsv</span> <span class="o">=</span> <span class="n">alignment</span><span class="p">[</span><span class="s2">&quot;tsv&quot;</span><span class="p">]</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">coverage_cutoff</span> <span class="o">=</span> <span class="mf">0.5</span>  <span class="c1"># fraction of median coverage</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">coverage_cutoff_pct</span> <span class="o">=</span> <span class="nb nb-Type">int</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">coverage_cutoff</span> <span class="o">*</span> <span class="mi">100</span><span class="p">)</span>  <span class="c1"># for PDF report</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">reference_length</span> <span class="o">=</span> <span class="n">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">record</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">has_warnings</span> <span class="o">=</span> <span class="n">False</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">has_errors</span> <span class="o">=</span> <span class="n">False</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">geneblocks_outcome</span> <span class="o">=</span> <span class="s2">&quot;none&quot;</span>  <span class="c1"># stores outcome, used in PDF report making</span>

        <span class="c1"># Set True if references are DNA Cauldron-simulated files:</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">dnacauldron</span> <span class="o">=</span> <span class="n">True</span>  <span class="c1"># for plotting</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">has_consensus</span> <span class="o">=</span> <span class="n">False</span>  <span class="c1"># for consensus sequence comparison</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">is_uncertain</span> <span class="o">=</span> <span class="n">True</span>  <span class="c1"># for sequencing data quality, used in report</span>

    <span class="n">def</span> <span class="n">perform_comparison</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">assembly_path</span><span class="o">=</span><span class="n">None</span><span class="p">,</span> <span class="n">vcf_path</span><span class="o">=</span><span class="n">None</span><span class="p">):</span>

        <span class="sd">&quot;&quot;&quot;Plot coverage and compare reference with a consensus sequence.</span>

<span class="sd">        **Parameters**</span>

<span class="sd">        **assembly_path**</span>

<span class="sd">        &gt; Optional. Path to a consensus FASTA file (`str`).</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">fig</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">plot_coverage</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">assembly_path</span> <span class="k">is</span> <span class="ow">not</span> <span class="n">None</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">comparison_figure</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">compare_with_assembly</span><span class="p">(</span>

                <span class="n">assembly_path</span><span class="o">=</span><span class="n">assembly_path</span>

            <span class="p">)</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">has_consensus</span> <span class="o">=</span> <span class="n">True</span>

        <span class="k">if</span> <span class="n">vcf_path</span> <span class="k">is</span> <span class="ow">not</span> <span class="n">None</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">vcf_table</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">create_vcf_table</span><span class="p">(</span><span class="n">vcf_path</span><span class="o">=</span><span class="n">vcf_path</span><span class="p">)</span>

        <span class="c1"># plt.close(&quot;all&quot;)</span>

    <span class="n">def</span> <span class="n">create_vcf_table</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">vcf_path</span><span class="p">):</span>

        <span class="n">vcf_dict</span> <span class="o">=</span> <span class="p">{</span><span class="n">key</span><span class="p">:</span> <span class="p">[]</span> <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">vcf_selected_info_keytable</span><span class="o">.</span><span class="n">keys</span><span class="p">()}</span>

        <span class="n">vcf_dict</span><span class="p">[</span><span class="s2">&quot;REF&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="n">vcf_dict</span><span class="p">[</span><span class="s2">&quot;ALT&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="n">vcf_dict</span><span class="p">[</span><span class="s2">&quot;LOC&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">variant</span> <span class="ow">in</span> <span class="n">cyvcf2</span><span class="o">.</span><span class="n">VCF</span><span class="p">(</span><span class="n">vcf_path</span><span class="p">):</span>

            <span class="n">vcf_dict</span><span class="p">[</span><span class="s2">&quot;REF&quot;</span><span class="p">]</span> <span class="o">+=</span> <span class="p">[</span><span class="n">variant</span><span class="o">.</span><span class="n">REF</span><span class="p">]</span>

            <span class="n">vcf_dict</span><span class="p">[</span><span class="s2">&quot;ALT&quot;</span><span class="p">]</span> <span class="o">+=</span> <span class="p">[</span><span class="n">variant</span><span class="o">.</span><span class="n">ALT</span><span class="p">]</span>

            <span class="n">vcf_dict</span><span class="p">[</span><span class="s2">&quot;LOC&quot;</span><span class="p">]</span> <span class="o">+=</span> <span class="p">[</span><span class="n">variant</span><span class="o">.</span><span class="n">start</span><span class="p">]</span>

            <span class="k">for</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">vcf_selected_info_keytable</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>

                <span class="n">vcf_dict</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">+=</span> <span class="p">[</span><span class="n">variant</span><span class="o">.</span><span class="n">INFO</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">key</span><span class="p">)]</span>

        <span class="n">vcf_table</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>

            <span class="n">vcf_dict</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;LOC&quot;</span><span class="p">,</span> <span class="s2">&quot;REF&quot;</span><span class="p">,</span> <span class="s2">&quot;ALT&quot;</span><span class="p">,</span> <span class="s2">&quot;TYPE&quot;</span><span class="p">,</span> <span class="s2">&quot;DP&quot;</span><span class="p">,</span> <span class="s2">&quot;RO&quot;</span><span class="p">,</span> <span class="s2">&quot;AO&quot;</span><span class="p">]</span>

        <span class="p">)</span>  <span class="c1"># also set the order of the columns</span>

        <span class="k">return</span> <span class="n">vcf_table</span>

    <span class="n">def</span> <span class="n">find_big_inserts</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">threshold</span><span class="o">=</span><span class="mi">100</span><span class="p">):</span>

        <span class="sd">&quot;&quot;&quot;Calculate % of reads with big inserts.</span>

<span class="sd">        **Parameters**</span>

<span class="sd">        **threshold**</span>

<span class="sd">        &gt; Size of insert in bp to be considered big (`int`).&quot;&quot;&quot;</span>

        <span class="n">read_insert_dict</span> <span class="o">=</span> <span class="p">{</span>

            <span class="n">read_id</span><span class="p">:</span> <span class="n">False</span> <span class="k">for</span> <span class="n">read_id</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">paf</span><span class="p">[</span><span class="s2">&quot;query_name&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>

        <span class="p">}</span>  <span class="c1"># Initialize. Value True when read has insert.</span>

        <span class="k">for</span> <span class="n">index</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">paf</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>

            <span class="n">read_id</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;query_name&quot;</span><span class="p">]</span>

            <span class="n">cigar</span> <span class="o">=</span> <span class="n">row</span><span class="o">.</span><span class="n">tail</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">item</span><span class="p">()</span>  <span class="c1"># CIGAR string column is the last</span>

            <span class="k">for</span> <span class="n">size</span><span class="p">,</span> <span class="n">mutation</span> <span class="ow">in</span> <span class="n">re</span><span class="o">.</span><span class="n">findall</span><span class="p">(</span><span class="sa">r</span><span class="s2">&quot;(\d+)([A-Z]{1})&quot;</span><span class="p">,</span> <span class="n">cigar</span><span class="p">):</span>

                <span class="k">if</span> <span class="nb nb-Type">int</span><span class="p">(</span><span class="n">size</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">threshold</span><span class="p">:</span>

                    <span class="k">if</span> <span class="n">mutation</span> <span class="o">==</span> <span class="s2">&quot;I&quot;</span><span class="p">:</span>  <span class="c1"># CIGAR code for insert</span>

                        <span class="n">read_insert_dict</span><span class="p">[</span><span class="n">read_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">True</span>

                        <span class="k">break</span>

        <span class="n">fraction_of_big_inserts</span> <span class="o">=</span> <span class="n">sum</span><span class="p">(</span><span class="n">read_insert_dict</span><span class="o">.</span><span class="n">values</span><span class="p">())</span> <span class="o">/</span> <span class="n">len</span><span class="p">(</span>

            <span class="n">read_insert_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>

        <span class="p">)</span>

        <span class="c1"># imprecise rounding, but ok for our purposes:</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">pct_big_insert</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="n">fraction_of_big_inserts</span> <span class="o">*</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>  <span class="c1"># pct multiplier</span>

    <span class="n">def</span> <span class="n">calculate_stats</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

        <span class="sd">&quot;&quot;&quot;Calculate statistics for the coverage plot, used in plot_coverage().&quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">xx</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">))</span>  <span class="c1"># for the plot x axis</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">yy</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">tsv</span><span class="p">[</span><span class="s2">&quot;depth&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">()</span>  <span class="c1"># for plotting coverage</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">median_yy</span> <span class="o">=</span> <span class="n">statistics</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">yy</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">median_yy</span> <span class="o">&lt;</span> <span class="mi">30</span><span class="p">:</span>  <span class="c1"># a good cutoff for low-depth sequencing</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">has_low_coverage</span> <span class="o">=</span> <span class="n">True</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">has_warnings</span> <span class="o">=</span> <span class="n">True</span>

        <span class="k">else</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">is_uncertain</span> <span class="o">=</span> <span class="n">False</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">has_low_coverage</span> <span class="o">=</span> <span class="n">False</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">find_big_inserts</span><span class="p">()</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">pct_big_insert</span> <span class="o">&gt;=</span> <span class="mi">50</span><span class="p">:</span>  <span class="c1"># at least half of reads have big insert</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">has_errors</span> <span class="o">=</span> <span class="n">True</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">has_big_insert</span> <span class="o">=</span> <span class="n">True</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">pct_big_insert</span> <span class="o">&gt;=</span> <span class="mi">5</span><span class="p">:</span>  <span class="c1"># five pct has big insert</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">has_warnings</span> <span class="o">=</span> <span class="n">True</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">has_reads_with_insert</span> <span class="o">=</span> <span class="n">True</span>

        <span class="k">else</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">has_big_insert</span> <span class="o">=</span> <span class="n">False</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">has_reads_with_insert</span> <span class="o">=</span> <span class="n">False</span>

        <span class="c1"># This section creates a list of low coverage position to be reported</span>

        <span class="n">indices</span> <span class="o">=</span> <span class="p">[</span>

            <span class="n">i</span>

            <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">yy</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">value</span> <span class="o">&lt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">median_yy</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">coverage_cutoff</span>

        <span class="p">]</span>

        <span class="n">G</span> <span class="o">=</span> <span class="p">(</span>

            <span class="n">list</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>

            <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">itertools</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span>

                <span class="n">indices</span><span class="p">,</span> <span class="n">lambda</span> <span class="n">x</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="n">itertools</span><span class="o">.</span><span class="n">count</span><span class="p">():</span> <span class="n">next</span><span class="p">(</span><span class="n">c</span><span class="p">)</span> <span class="o">-</span> <span class="n">x</span>

            <span class="p">)</span>

        <span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">low_coverage_positions_string</span> <span class="o">=</span> <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>

            <span class="s2">&quot;-&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">map</span><span class="p">(</span><span class="nb">str</span><span class="p">,</span> <span class="p">(</span><span class="n">g</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">g</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])[:</span> <span class="n">len</span><span class="p">(</span><span class="n">g</span><span class="p">)]))</span> <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">G</span>

        <span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">low_coverage_positions_string</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">low_coverage_positions_string</span> <span class="o">=</span> <span class="s2">&quot;-&quot;</span>  <span class="c1"># looks better in the pdf report</span>

        <span class="k">else</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">has_warnings</span> <span class="o">=</span> <span class="n">True</span>

    <span class="n">def</span> <span class="n">plot_coverage</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

        <span class="sd">&quot;&quot;&quot;Plot the reference with the coverage and weak reads.&quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">hasattr</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="s2">&quot;xx&quot;</span><span class="p">):</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">calculate_stats</span><span class="p">()</span>  <span class="c1"># also calculates yy and zz</span>

        <span class="c1"># Plot</span>

        <span class="c1">######</span>

        <span class="n">fig</span><span class="p">,</span> <span class="p">(</span><span class="n">ax1</span><span class="p">,</span> <span class="n">ax2</span><span class="p">)</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span>

            <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">7</span><span class="p">,</span> <span class="mi">3</span><span class="p">),</span> <span class="n">sharex</span><span class="o">=</span><span class="n">True</span><span class="p">,</span> <span class="n">gridspec_kw</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;height_ratios&quot;</span><span class="p">:</span> <span class="p">[</span><span class="mi">4</span><span class="p">,</span> <span class="mi">1</span><span class="p">]}</span>

        <span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">dnacauldron</span><span class="p">:</span>

            <span class="n">graphic_record</span> <span class="o">=</span> <span class="n">CustomTranslator</span><span class="p">()</span><span class="o">.</span><span class="n">translate_record</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">record</span><span class="p">)</span>

        <span class="k">else</span><span class="p">:</span>

            <span class="n">graphic_record</span> <span class="o">=</span> <span class="n">dna_features_viewer</span><span class="o">.</span><span class="n">BiopythonTranslator</span><span class="p">()</span><span class="o">.</span><span class="n">translate_record</span><span class="p">(</span>

                <span class="bp">self</span><span class="o">.</span><span class="n">record</span>

            <span class="p">)</span>

        <span class="n">graphic_record</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">ax</span><span class="o">=</span><span class="n">ax1</span><span class="p">,</span> <span class="n">with_ruler</span><span class="o">=</span><span class="n">False</span><span class="p">,</span> <span class="n">strand_in_label_threshold</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span>

        <span class="c1"># Plot coverage</span>

        <span class="n">ax2</span><span class="o">.</span><span class="n">fill_between</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">xx</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">yy</span><span class="p">,</span> <span class="n">alpha</span><span class="o">=</span><span class="mf">0.8</span><span class="p">)</span>

        <span class="n">stdev_yy</span> <span class="o">=</span> <span class="n">statistics</span><span class="o">.</span><span class="n">stdev</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">yy</span><span class="p">)</span>

        <span class="n">ax2</span><span class="o">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="n">bottom</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

        <span class="n">ax2</span><span class="o">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Depth&quot;</span><span class="p">)</span>

        <span class="n">ax2</span><span class="o">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Base [bp]&quot;</span><span class="p">)</span>

        <span class="n">ax2</span><span class="o">.</span><span class="n">axhline</span><span class="p">(</span>

            <span class="n">y</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">median_yy</span><span class="p">,</span> <span class="n">xmin</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">xmax</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;grey&quot;</span><span class="p">,</span> <span class="n">linestyle</span><span class="o">=</span><span class="s2">&quot;-&quot;</span><span class="p">,</span> <span class="n">linewidth</span><span class="o">=</span><span class="mf">0.8</span>

        <span class="p">)</span>

        <span class="n">ax2</span><span class="o">.</span><span class="n">axhline</span><span class="p">(</span>

            <span class="n">y</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">median_yy</span> <span class="o">-</span> <span class="n">stdev_yy</span><span class="p">,</span>

            <span class="n">xmin</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>

            <span class="n">xmax</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>

            <span class="n">color</span><span class="o">=</span><span class="s2">&quot;grey&quot;</span><span class="p">,</span>

            <span class="n">linestyle</span><span class="o">=</span><span class="s2">&quot;--&quot;</span><span class="p">,</span>

            <span class="n">linewidth</span><span class="o">=</span><span class="mf">0.8</span><span class="p">,</span>

        <span class="p">)</span>

        <span class="n">ax2</span><span class="o">.</span><span class="n">axhline</span><span class="p">(</span>

            <span class="n">y</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">median_yy</span> <span class="o">+</span> <span class="n">stdev_yy</span><span class="p">,</span>

            <span class="n">xmin</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>

            <span class="n">xmax</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span>

            <span class="n">color</span><span class="o">=</span><span class="s2">&quot;grey&quot;</span><span class="p">,</span>

            <span class="n">linestyle</span><span class="o">=</span><span class="s2">&quot;--&quot;</span><span class="p">,</span>

            <span class="n">linewidth</span><span class="o">=</span><span class="mf">0.8</span><span class="p">,</span>

        <span class="p">)</span>

        <span class="k">return</span> <span class="n">fig</span>

    <span class="n">def</span> <span class="n">get_weak_read_histogram_data</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">cutoff</span><span class="p">):</span>

        <span class="sd">&quot;&quot;&quot;Create data for the plotting function.</span>

<span class="sd">        **Parameters**</span>

<span class="sd">        **cutoff**</span>

<span class="sd">        &gt; Cutoff for proportion of bases that map (`float`). Reads below the cutoff are</span>

<span class="sd">        viewed as weak reads. Recommended value: 0.8 (80%).</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># For each read, calculate the proportion of bases that map, then create an array</span>

        <span class="c1"># for each read that is below the cutoff. The sum of these arrays will give the</span>

        <span class="c1"># depth of aligning bad reads for each base of the reference.</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">proportion_maps</span> <span class="o">=</span> <span class="p">(</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">paf</span><span class="p">[</span><span class="s2">&quot;mapping_matches&quot;</span><span class="p">]</span> <span class="o">/</span> <span class="bp">self</span><span class="o">.</span><span class="n">paf</span><span class="p">[</span><span class="s2">&quot;query_length&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">()</span>

        <span class="p">)</span>

        <span class="n">filtered_paf</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">paf</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">proportion_maps</span> <span class="o">&lt;</span> <span class="n">cutoff</span><span class="p">]</span>

        <span class="n">array_length</span> <span class="o">=</span> <span class="n">filtered_paf</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;target_length&quot;</span><span class="p">]</span>

        <span class="n">histogram_array</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="k">for</span> <span class="n">index</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">filtered_paf</span><span class="o">.</span><span class="n">iterrows</span><span class="p">():</span>

            <span class="n">read_array</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="n">array_length</span>

            <span class="n">start</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;target_start&quot;</span><span class="p">]</span>

            <span class="n">end</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;target_end&quot;</span><span class="p">]</span>

            <span class="n">value</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">/</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;multialign&quot;</span><span class="p">]</span>

            <span class="n">values</span> <span class="o">=</span> <span class="p">[</span><span class="n">value</span><span class="p">]</span> <span class="o">*</span> <span class="p">(</span><span class="n">end</span> <span class="o">-</span> <span class="n">start</span><span class="p">)</span>

            <span class="n">read_array</span><span class="p">[</span><span class="n">start</span><span class="p">:</span><span class="n">end</span><span class="p">]</span> <span class="o">=</span> <span class="n">values</span>

            <span class="n">histogram_array</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">read_array</span><span class="p">)</span>

        <span class="n">histogram_numpy_array</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">histogram_array</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">zz</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">histogram_numpy_array</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

    <span class="n">def</span> <span class="n">compare_with_assembly</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">assembly_path</span><span class="p">):</span>

        <span class="sd">&quot;&quot;&quot;Compare the reference with a consensus sequence.</span>

<span class="sd">        Check length, orientation (sense or reverse complement), then use GeneBlocks to</span>

<span class="sd">        highlight differences between reference and the plasmid assembly.</span>

<span class="sd">        **Parameters**</span>

<span class="sd">        **assembly_path**</span>

<span class="sd">        &gt; Path to a consensus FASTA file (`str`).</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="c1"># Note that in this context &#39;assembly&#39; means a consensus of variant calls</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">assembly</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="n">handle</span><span class="o">=</span><span class="n">assembly_path</span><span class="p">,</span> <span class="n">format</span><span class="o">=</span><span class="s2">&quot;fasta&quot;</span><span class="p">)</span>

        <span class="n">length_ratio</span> <span class="o">=</span> <span class="n">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">assembly</span><span class="p">)</span> <span class="o">/</span> <span class="n">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">record</span><span class="p">)</span>

        <span class="k">if</span> <span class="mf">0.95</span> <span class="o">&lt;</span> <span class="n">length_ratio</span> <span class="o">&lt;</span> <span class="mf">1.05</span><span class="p">:</span>  <span class="c1"># length within reason</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">is_length_ok</span> <span class="o">=</span> <span class="n">True</span>  <span class="c1"># can compare with expected sequence</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">incorrect_length_msg</span> <span class="o">=</span> <span class="n">None</span>

        <span class="k">else</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">geneblocks_outcome</span> <span class="o">=</span> <span class="s2">&quot;incorrect_length&quot;</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">is_length_ok</span> <span class="o">=</span> <span class="n">False</span>  <span class="c1"># do not compare sequences</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">has_errors</span> <span class="o">=</span> <span class="n">True</span>

            <span class="n">difference_bases</span> <span class="o">=</span> <span class="n">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">assembly</span><span class="p">)</span> <span class="o">-</span> <span class="n">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">record</span><span class="p">)</span>

            <span class="n">difference_percent</span> <span class="o">=</span> <span class="p">(</span><span class="n">length_ratio</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="mi">100</span>  <span class="c1"># get as percent</span>

            <span class="k">if</span> <span class="n">difference_percent</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>

                <span class="n">difference_text</span> <span class="o">=</span> <span class="s2">&quot;longer&quot;</span>

            <span class="k">else</span><span class="p">:</span>

                <span class="n">difference_text</span> <span class="o">=</span> <span class="s2">&quot;shorter&quot;</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">incorrect_length_msg</span> <span class="o">=</span> <span class="p">(</span>

                <span class="s2">&quot;The consensus is </span><span class="si">%d%%</span><span class="s2"> (</span><span class="si">%d</span><span class="s2"> bp) </span><span class="si">%s</span><span class="s2"> than the reference.&quot;</span>

                <span class="o">%</span> <span class="p">(</span><span class="nb nb-Type">int</span><span class="p">(</span><span class="nb">abs</span><span class="p">(</span><span class="n">difference_percent</span><span class="p">)),</span> <span class="n">difference_bases</span><span class="p">,</span> <span class="n">difference_text</span><span class="p">)</span>

            <span class="p">)</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">geneblocks_done</span> <span class="o">=</span> <span class="n">False</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">is_diffblocks_reverse</span> <span class="o">=</span> <span class="n">None</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">is_comparison_successful</span> <span class="o">=</span> <span class="n">False</span>  <span class="c1"># for PDF reporting</span>

            <span class="k">return</span>

        <span class="c1"># To find out the orientation of the assembly, we compare using Levenshtein</span>

        <span class="n">lev_distance</span> <span class="o">=</span> <span class="n">lev</span><span class="p">(</span><span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">assembly</span><span class="o">.</span><span class="n">seq</span><span class="p">),</span> <span class="nb">str</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">))</span>

        <span class="n">lev_cutoff</span> <span class="o">=</span> <span class="mi">50</span>  <span class="c1"># more than this is too much to plot. Also for most plasmids,</span>

        <span class="c1"># this represents ~1% difference, which is too much error.</span>

        <span class="n">assembly_for_diffblocks</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">assembly</span>

        <span class="k">if</span> <span class="n">lev_distance</span> <span class="o">&gt;</span> <span class="n">lev_cutoff</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">perform_geneblocks</span> <span class="o">=</span> <span class="n">False</span>

        <span class="k">else</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">perform_geneblocks</span> <span class="o">=</span> <span class="n">True</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">perform_geneblocks</span><span class="p">:</span>

            <span class="n">try</span><span class="p">:</span>

                <span class="n">diff_blocks</span> <span class="o">=</span> <span class="n">geneblocks</span><span class="o">.</span><span class="n">DiffBlocks</span><span class="o">.</span><span class="n">from_sequences</span><span class="p">(</span>

                    <span class="n">assembly_for_diffblocks</span><span class="p">,</span> <span class="bp">self</span><span class="o">.</span><span class="n">record</span>

                <span class="p">)</span>

            <span class="n">except</span> <span class="n">KeyError</span><span class="p">:</span>

                <span class="bp">self</span><span class="o">.</span><span class="n">geneblocks_done</span> <span class="o">=</span> <span class="n">False</span>

                <span class="bp">self</span><span class="o">.</span><span class="n">geneblocks_outcome</span> <span class="o">=</span> <span class="s2">&quot;geneblocks_error&quot;</span>

            <span class="k">else</span><span class="p">:</span>  <span class="c1"># diffblocks had no error</span>

                <span class="bp">self</span><span class="o">.</span><span class="n">geneblocks_done</span> <span class="o">=</span> <span class="n">True</span>

                <span class="n">ax1</span><span class="p">,</span> <span class="n">ax2</span> <span class="o">=</span> <span class="n">diff_blocks</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">figure_width</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>

                <span class="bp">self</span><span class="o">.</span><span class="n">is_diffblocks_reverse</span> <span class="o">=</span> <span class="n">False</span>

                <span class="bp">self</span><span class="o">.</span><span class="n">is_comparison_successful</span> <span class="o">=</span> <span class="n">True</span>

                <span class="bp">self</span><span class="o">.</span><span class="n">geneblocks_outcome</span> <span class="o">=</span> <span class="s2">&quot;all_good&quot;</span>

                <span class="k">return</span> <span class="n">ax1</span>

            <span class="c1"># We try again as due to a bug in geneblocks, the reverse order may work:</span>

            <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">geneblocks_outcome</span> <span class="o">==</span> <span class="s2">&quot;geneblocks_error&quot;</span><span class="p">:</span>

                <span class="n">try</span><span class="p">:</span>

                    <span class="n">diff_blocks</span> <span class="o">=</span> <span class="n">geneblocks</span><span class="o">.</span><span class="n">DiffBlocks</span><span class="o">.</span><span class="n">from_sequences</span><span class="p">(</span>

                        <span class="bp">self</span><span class="o">.</span><span class="n">record</span><span class="p">,</span> <span class="n">assembly_for_diffblocks</span>

                    <span class="p">)</span>

                <span class="n">except</span> <span class="n">KeyError</span><span class="p">:</span>

                    <span class="k">return</span>

                <span class="k">else</span><span class="p">:</span>

                    <span class="bp">self</span><span class="o">.</span><span class="n">geneblocks_done</span> <span class="o">=</span> <span class="n">True</span>  <span class="c1"># overwrite</span>

                    <span class="n">ax1</span><span class="p">,</span> <span class="n">ax2</span> <span class="o">=</span> <span class="n">diff_blocks</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">figure_width</span><span class="o">=</span><span class="mi">7</span><span class="p">)</span>

                    <span class="bp">self</span><span class="o">.</span><span class="n">is_diffblocks_reverse</span> <span class="o">=</span> <span class="n">True</span>

                    <span class="bp">self</span><span class="o">.</span><span class="n">is_comparison_successful</span> <span class="o">=</span> <span class="n">True</span>

                    <span class="bp">self</span><span class="o">.</span><span class="n">geneblocks_outcome</span> <span class="o">=</span> <span class="s2">&quot;swapped_diffblocks&quot;</span>

                    <span class="k">return</span> <span class="n">ax1</span>

        <span class="k">else</span><span class="p">:</span>  <span class="c1"># too many differences</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">geneblocks_done</span> <span class="o">=</span> <span class="n">False</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">geneblocks_outcome</span> <span class="o">=</span> <span class="s2">&quot;too_many_differences&quot;</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">is_comparison_successful</span> <span class="o">=</span> <span class="n">False</span>

    <span class="n">def</span> <span class="n">filter_fastq</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">fastq_path</span><span class="p">,</span> <span class="n">target</span><span class="o">=</span><span class="n">None</span><span class="p">):</span>

        <span class="sd">&quot;&quot;&quot;Filter original FASTQ file for reads that best map to the reference.</span>

<span class="sd">        This function is used to obtain a FASTQ file that can be used for Canu assembly.</span>

<span class="sd">        **Parameters**</span>

<span class="sd">        **fastq_path**</span>

<span class="sd">        &gt; Path to FASTQ file (`str`).</span>

<span class="sd">        **target**</span>

<span class="sd">        &gt; Path to save the filtered FASTQ file (`str`).</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="n">target</span> <span class="k">is</span> <span class="n">None</span><span class="p">:</span>

            <span class="n">target</span> <span class="o">=</span> <span class="n">fastq_path</span> <span class="o">+</span> <span class="s2">&quot;_filtered.fastq&quot;</span>

        <span class="n">read_names</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">paf</span><span class="p">[</span><span class="s2">&quot;query_name&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">()</span>

        <span class="n">input_seq_iterator</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">fastq_path</span><span class="p">,</span> <span class="s2">&quot;fastq&quot;</span><span class="p">)</span>

        <span class="n">seq_iterator</span> <span class="o">=</span> <span class="p">(</span>

            <span class="n">record</span> <span class="k">for</span> <span class="n">record</span> <span class="ow">in</span> <span class="n">input_seq_iterator</span> <span class="k">if</span> <span class="n">record</span><span class="o">.</span><span class="n">name</span> <span class="ow">in</span> <span class="n">read_names</span>

        <span class="p">)</span>

        <span class="n">SeqIO</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">seq_iterator</span><span class="p">,</span> <span class="n">target</span><span class="p">,</span> <span class="s2">&quot;fastq&quot;</span><span class="p">)</span>
</code></pre></div>

</details>
<hr />
<h4 id="methods">Methods</h4>
<h4 id="calculate_stats">calculate_stats</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">calculate_stats</span><span class="p">(</span>
    <span class="bp">self</span>
<span class="p">)</span>
</code></pre></div>

<p>Calculate statistics for the coverage plot, used in plot_coverage().</p>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code>    <span class="n">def</span> <span class="n">calculate_stats</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

        <span class="sd">&quot;&quot;&quot;Calculate statistics for the coverage plot, used in plot_coverage().&quot;&quot;&quot;</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">xx</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">len</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">record</span><span class="o">.</span><span class="n">seq</span><span class="p">))</span>  <span class="c1"># for the plot x axis</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">yy</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">tsv</span><span class="p">[</span><span class="s2">&quot;depth&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">to_list</span><span class="p">()</span>  <span class="c1"># for plotting coverage</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">median_yy</span> <span class="o">=</span> <span class="n">statistics</span><span class="o">.</span><span class="n">median</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">yy</span><span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">median_yy</span> <span class="o">&lt;</span> <span class="mi">30</span><span class="p">:</span>  <span class="c1"># a good cutoff for low-depth sequencing</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">has_low_coverage</span> <span class="o">=</span> <span class="n">True</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">has_warnings</span> <span class="o">=</span> <span class="n">True</span>

        <span class="k">else</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">is_uncertain</span> <span class="o">=</span> <span class="n">False</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">has_low_coverage</span> <span class="o">=</span> <span class="n">False</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">find_big_inserts</span><span class="p">()</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">pct_big_insert</span> <span class="o">&gt;=</span> <span class="mi">50</span><span class="p">:</span>  <span class="c1"># at least half of reads have big insert</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">has_errors</span> <span class="o">=</span> <span class="n">True</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">has_big_insert</span> <span class="o">=</span> <span class="n">True</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">pct_big_insert</span> <span class="o">&gt;=</span> <span class="mi">5</span><span class="p">:</span>  <span class="c1"># five pct has big insert</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">has_warnings</span> <span class="o">=</span> <span class="n">True</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">has_reads_with_insert</span> <span class="o">=</span> <span class="n">True</span>

        <span class="k">else</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">has_big_insert</span> <span class="o">=</span> <span class="n">False</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">has_reads_with_insert</span> <span class="o">=</span> <span class="n">False</span>

        <span class="c1"># This section creates a list of low coverage position to be reported</span>

        <span class="n">indices</span> <span class="o">=</span> <span class="p">[</span>

            <span class="n">i</span>

            <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">enumerate</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">yy</span><span class="p">)</span>

            <span class="k">if</span> <span class="n">value</span> <span class="o">&lt;</span> <span class="bp">self</span><span class="o">.</span><span class="n">median_yy</span> <span class="o">*</span> <span class="bp">self</span><span class="o">.</span><span class="n">coverage_cutoff</span>

        <span class="p">]</span>

        <span class="n">G</span> <span class="o">=</span> <span class="p">(</span>

            <span class="n">list</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>

            <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">itertools</span><span class="o">.</span><span class="n">groupby</span><span class="p">(</span>

                <span class="n">indices</span><span class="p">,</span> <span class="n">lambda</span> <span class="n">x</span><span class="p">,</span> <span class="n">c</span><span class="o">=</span><span class="n">itertools</span><span class="o">.</span><span class="n">count</span><span class="p">():</span> <span class="n">next</span><span class="p">(</span><span class="n">c</span><span class="p">)</span> <span class="o">-</span> <span class="n">x</span>

            <span class="p">)</span>

        <span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">low_coverage_positions_string</span> <span class="o">=</span> <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>

            <span class="s2">&quot;-&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">map</span><span class="p">(</span><span class="nb">str</span><span class="p">,</span> <span class="p">(</span><span class="n">g</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">g</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])[:</span> <span class="n">len</span><span class="p">(</span><span class="n">g</span><span class="p">)]))</span> <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">G</span>

        <span class="p">)</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">low_coverage_positions_string</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">low_coverage_positions_string</span> <span class="o">=</span> <span class="s2">&quot;-&quot;</span>  <span class="c1"># looks better in the pdf report</span>

        <span class="k">else</span><span class="p">:</span>

            <span class="bp">self</span><span class="o">.</span><span class="n">has_warnings</span> <span class="o">=</span> <span class="n">True</span>
</code></pre></div>

</details>
<h4 id="compare_with_assembly">compare_with_assembly</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">compare_with_assembly</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">assembly_path</span>
<span class="p">)</span>
</code></pre></div>

<p>Compare the reference with a consensus sequence.</p>
<p>Check length, orientation (sense or reverse complement), then use GeneBlocks to
highlight differences between reference and the plasmid assembly.</p>
<p><strong>Parameters</strong></p>
<p><strong>assembly_path</strong></p>
<blockquote>
<p>Path to a consensus FASTA file (<code>str</code>).</p>
</blockquote>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code>    <span class="n">def</span> <span class="n">compare_with_assembly</span><span class="p">(</span><span class="n">self</span><span class="p">,</span> <span class="n">assembly_path</span><span class="p">)</span><span class="o">:</span>

        <span class="s2">&quot;</span><span class="se">&quot;&quot;</span><span class="s2">Compare the reference with a consensus sequence.</span>

<span class="s2">        Check length, orientation (sense or reverse complement), then use GeneBlocks to</span>

<span class="s2">        highlight differences between reference and the plasmid assembly.</span>

<span class="s2">        **Parameters**</span>

<span class="s2">        **assembly_path**</span>

<span class="s2">        &gt; Path to a consensus FASTA file (`str`).</span>

<span class="s2">        </span><span class="se">&quot;&quot;</span><span class="s2">&quot;</span>

        <span class="c1"># Note that in this context &#39;assembly&#39; means a consensus of variant calls</span>

        <span class="n">self</span><span class="p">.</span><span class="n">assembly</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="p">.</span><span class="k">read</span><span class="p">(</span><span class="n">handle</span><span class="o">=</span><span class="n">assembly_path</span><span class="p">,</span> <span class="k">format</span><span class="o">=</span><span class="s2">&quot;fasta&quot;</span><span class="p">)</span>

        <span class="n">length_ratio</span> <span class="o">=</span> <span class="n">len</span><span class="p">(</span><span class="n">self</span><span class="p">.</span><span class="n">assembly</span><span class="p">)</span> <span class="o">/</span> <span class="n">len</span><span class="p">(</span><span class="n">self</span><span class="p">.</span><span class="n">record</span><span class="p">)</span>

        <span class="k">if</span> <span class="mf">0.95</span> <span class="o">&lt;</span> <span class="n">length_ratio</span> <span class="o">&lt;</span> <span class="mf">1.05</span><span class="o">:</span>  <span class="c1"># length within reason</span>

            <span class="n">self</span><span class="p">.</span><span class="n">is_length_ok</span> <span class="o">=</span> <span class="no">True</span>  <span class="c1"># can compare with expected sequence</span>

            <span class="n">self</span><span class="p">.</span><span class="n">incorrect_length_msg</span> <span class="o">=</span> <span class="k">None</span>

        <span class="k">else</span><span class="o">:</span>

            <span class="n">self</span><span class="p">.</span><span class="n">geneblocks_outcome</span> <span class="o">=</span> <span class="s2">&quot;incorrect_length&quot;</span>

            <span class="n">self</span><span class="p">.</span><span class="n">is_length_ok</span> <span class="o">=</span> <span class="no">False</span>  <span class="c1"># do not compare sequences</span>

            <span class="n">self</span><span class="p">.</span><span class="n">has_errors</span> <span class="o">=</span> <span class="no">True</span>

            <span class="n">difference_bases</span> <span class="o">=</span> <span class="n">len</span><span class="p">(</span><span class="n">self</span><span class="p">.</span><span class="n">assembly</span><span class="p">)</span> <span class="o">-</span> <span class="n">len</span><span class="p">(</span><span class="n">self</span><span class="p">.</span><span class="n">record</span><span class="p">)</span>

            <span class="n">difference_percent</span> <span class="o">=</span> <span class="p">(</span><span class="n">length_ratio</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="o">*</span> <span class="mi">100</span>  <span class="c1"># get as percent</span>

            <span class="k">if</span> <span class="n">difference_percent</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="o">:</span>

                <span class="n">difference_text</span> <span class="o">=</span> <span class="s2">&quot;longer&quot;</span>

            <span class="k">else</span><span class="o">:</span>

                <span class="n">difference_text</span> <span class="o">=</span> <span class="s2">&quot;shorter&quot;</span>

            <span class="n">self</span><span class="p">.</span><span class="n">incorrect_length_msg</span> <span class="o">=</span> <span class="p">(</span>

                <span class="s2">&quot;The consensus is %d%% (%d bp) %s than the reference.&quot;</span>

                <span class="o">%</span> <span class="p">(</span><span class="kt">int</span><span class="p">(</span><span class="nf">abs</span><span class="p">(</span><span class="n">difference_percent</span><span class="p">)),</span> <span class="n">difference_bases</span><span class="p">,</span> <span class="n">difference_text</span><span class="p">)</span>

            <span class="p">)</span>

            <span class="n">self</span><span class="p">.</span><span class="n">geneblocks_done</span> <span class="o">=</span> <span class="no">False</span>

            <span class="n">self</span><span class="p">.</span><span class="n">is_diffblocks_reverse</span> <span class="o">=</span> <span class="k">None</span>

            <span class="n">self</span><span class="p">.</span><span class="n">is_comparison_successful</span> <span class="o">=</span> <span class="no">False</span>  <span class="c1"># for PDF reporting</span>

            <span class="k">return</span>

        <span class="c1"># To find out the orientation of the assembly, we compare using Levenshtein</span>

        <span class="n">lev_distance</span> <span class="o">=</span> <span class="n">lev</span><span class="p">(</span><span class="n">str</span><span class="p">(</span><span class="n">self</span><span class="p">.</span><span class="n">assembly</span><span class="p">.</span><span class="n">seq</span><span class="p">),</span> <span class="n">str</span><span class="p">(</span><span class="n">self</span><span class="p">.</span><span class="n">record</span><span class="p">.</span><span class="n">seq</span><span class="p">))</span>

        <span class="n">lev_cutoff</span> <span class="o">=</span> <span class="mi">50</span>  <span class="c1"># more than this is too much to plot. Also for most plasmids,</span>

        <span class="c1"># this represents ~1% difference, which is too much error.</span>

        <span class="n">assembly_for_diffblocks</span> <span class="o">=</span> <span class="n">self</span><span class="p">.</span><span class="n">assembly</span>

        <span class="k">if</span> <span class="n">lev_distance</span> <span class="o">&gt;</span> <span class="n">lev_cutoff</span><span class="o">:</span>

            <span class="n">self</span><span class="p">.</span><span class="n">perform_geneblocks</span> <span class="o">=</span> <span class="no">False</span>

        <span class="k">else</span><span class="o">:</span>

            <span class="n">self</span><span class="p">.</span><span class="n">perform_geneblocks</span> <span class="o">=</span> <span class="no">True</span>

        <span class="k">if</span> <span class="n">self</span><span class="p">.</span><span class="n">perform_geneblocks</span><span class="o">:</span>

            <span class="n">try</span><span class="o">:</span>

                <span class="n">diff_blocks</span> <span class="o">=</span> <span class="n">geneblocks</span><span class="p">.</span><span class="n">DiffBlocks</span><span class="p">.</span><span class="n">from_sequences</span><span class="p">(</span>

                    <span class="n">assembly_for_diffblocks</span><span class="p">,</span> <span class="n">self</span><span class="p">.</span><span class="n">record</span>

                <span class="p">)</span>

            <span class="k">except</span> <span class="n">KeyError</span><span class="o">:</span>

                <span class="n">self</span><span class="p">.</span><span class="n">geneblocks_done</span> <span class="o">=</span> <span class="no">False</span>

                <span class="n">self</span><span class="p">.</span><span class="n">geneblocks_outcome</span> <span class="o">=</span> <span class="s2">&quot;geneblocks_error&quot;</span>

            <span class="k">else</span><span class="o">:</span>  <span class="c1"># diffblocks had no error</span>

                <span class="n">self</span><span class="p">.</span><span class="n">geneblocks_done</span> <span class="o">=</span> <span class="no">True</span>

                <span class="n">ax1</span><span class="p">,</span> <span class="n">ax2</span> <span class="o">=</span> <span class="n">diff_blocks</span><span class="p">.</span><span class="n">plot</span><span class="p">(</span><span class="n">figure_width</span><span class="o">=</span><span class="mi">5</span><span class="p">)</span>

                <span class="n">self</span><span class="p">.</span><span class="n">is_diffblocks_reverse</span> <span class="o">=</span> <span class="no">False</span>

                <span class="n">self</span><span class="p">.</span><span class="n">is_comparison_successful</span> <span class="o">=</span> <span class="no">True</span>

                <span class="n">self</span><span class="p">.</span><span class="n">geneblocks_outcome</span> <span class="o">=</span> <span class="s2">&quot;all_good&quot;</span>

                <span class="k">return</span> <span class="n">ax1</span>

            <span class="c1"># We try again as due to a bug in geneblocks, the reverse order may work:</span>

            <span class="k">if</span> <span class="n">self</span><span class="p">.</span><span class="n">geneblocks_outcome</span> <span class="o">==</span> <span class="s2">&quot;geneblocks_error&quot;</span><span class="o">:</span>

                <span class="n">try</span><span class="o">:</span>

                    <span class="n">diff_blocks</span> <span class="o">=</span> <span class="n">geneblocks</span><span class="p">.</span><span class="n">DiffBlocks</span><span class="p">.</span><span class="n">from_sequences</span><span class="p">(</span>

                        <span class="n">self</span><span class="p">.</span><span class="n">record</span><span class="p">,</span> <span class="n">assembly_for_diffblocks</span>

                    <span class="p">)</span>

                <span class="k">except</span> <span class="n">KeyError</span><span class="o">:</span>

                    <span class="k">return</span>

                <span class="k">else</span><span class="o">:</span>

                    <span class="n">self</span><span class="p">.</span><span class="n">geneblocks_done</span> <span class="o">=</span> <span class="no">True</span>  <span class="c1"># overwrite</span>

                    <span class="n">ax1</span><span class="p">,</span> <span class="n">ax2</span> <span class="o">=</span> <span class="n">diff_blocks</span><span class="p">.</span><span class="n">plot</span><span class="p">(</span><span class="n">figure_width</span><span class="o">=</span><span class="mi">7</span><span class="p">)</span>

                    <span class="n">self</span><span class="p">.</span><span class="n">is_diffblocks_reverse</span> <span class="o">=</span> <span class="no">True</span>

                    <span class="n">self</span><span class="p">.</span><span class="n">is_comparison_successful</span> <span class="o">=</span> <span class="no">True</span>

                    <span class="n">self</span><span class="p">.</span><span class="n">geneblocks_outcome</span> <span class="o">=</span> <span class="s2">&quot;swapped_diffblocks&quot;</span>

                    <span class="k">return</span> <span class="n">ax1</span>

        <span class="k">else</span><span class="o">:</span>  <span class="c1"># too many differences</span>

            <span class="n">self</span><span class="p">.</span><span class="n">geneblocks_done</span> <span class="o">=</span> <span class="no">False</span>

            <span class="n">self</span><span class="p">.</span><span class="n">geneblocks_outcome</span> <span class="o">=</span> <span class="s2">&quot;too_many_differences&quot;</span>

            <span class="n">self</span><span class="p">.</span><span class="n">is_comparison_successful</span> <span class="o">=</span> <span class="no">False</span>
</code></pre></div>

</details>
<h4 id="create_vcf_table">create_vcf_table</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">create_vcf_table</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">vcf_path</span>
<span class="p">)</span>
</code></pre></div>

<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code><span class="w">    </span><span class="n">def</span><span class="w"> </span><span class="n">create_vcf_table</span><span class="p">(</span><span class="n">self</span><span class="p">,</span><span class="w"> </span><span class="n">vcf_path</span><span class="p">)</span><span class="err">:</span><span class="w"></span>

<span class="w">        </span><span class="n">vcf_dict</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="err">{</span><span class="k">key</span><span class="err">:</span><span class="w"> </span><span class="err">[]</span><span class="w"> </span><span class="k">for</span><span class="w"> </span><span class="k">key</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">vcf_selected_info_keytable</span><span class="p">.</span><span class="n">keys</span><span class="p">()</span><span class="err">}</span><span class="w"></span>

<span class="w">        </span><span class="n">vcf_dict</span><span class="o">[</span><span class="n">&quot;REF&quot;</span><span class="o">]</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="err">[]</span><span class="w"></span>

<span class="w">        </span><span class="n">vcf_dict</span><span class="o">[</span><span class="n">&quot;ALT&quot;</span><span class="o">]</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="err">[]</span><span class="w"></span>

<span class="w">        </span><span class="n">vcf_dict</span><span class="o">[</span><span class="n">&quot;LOC&quot;</span><span class="o">]</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="err">[]</span><span class="w"></span>

<span class="w">        </span><span class="k">for</span><span class="w"> </span><span class="n">variant</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">cyvcf2</span><span class="p">.</span><span class="n">VCF</span><span class="p">(</span><span class="n">vcf_path</span><span class="p">)</span><span class="err">:</span><span class="w"></span>

<span class="w">            </span><span class="n">vcf_dict</span><span class="o">[</span><span class="n">&quot;REF&quot;</span><span class="o">]</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="o">[</span><span class="n">variant.REF</span><span class="o">]</span><span class="w"></span>

<span class="w">            </span><span class="n">vcf_dict</span><span class="o">[</span><span class="n">&quot;ALT&quot;</span><span class="o">]</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="o">[</span><span class="n">variant.ALT</span><span class="o">]</span><span class="w"></span>

<span class="w">            </span><span class="n">vcf_dict</span><span class="o">[</span><span class="n">&quot;LOC&quot;</span><span class="o">]</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="o">[</span><span class="n">variant.start</span><span class="o">]</span><span class="w"></span>

<span class="w">            </span><span class="k">for</span><span class="w"> </span><span class="k">key</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">vcf_selected_info_keytable</span><span class="p">.</span><span class="n">keys</span><span class="p">()</span><span class="err">:</span><span class="w"></span>

<span class="w">                </span><span class="n">vcf_dict</span><span class="o">[</span><span class="n">key</span><span class="o">]</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="o">[</span><span class="n">variant.INFO.get(key)</span><span class="o">]</span><span class="w"></span>

<span class="w">        </span><span class="n">vcf_table</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pandas</span><span class="p">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="w"></span>

<span class="w">            </span><span class="n">vcf_dict</span><span class="p">,</span><span class="w"> </span><span class="n">columns</span><span class="o">=[</span><span class="n">&quot;LOC&quot;, &quot;REF&quot;, &quot;ALT&quot;, &quot;TYPE&quot;, &quot;DP&quot;, &quot;RO&quot;, &quot;AO&quot;</span><span class="o">]</span><span class="w"></span>

<span class="w">        </span><span class="p">)</span><span class="w">  </span><span class="err">#</span><span class="w"> </span><span class="n">also</span><span class="w"> </span><span class="k">set</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="k">order</span><span class="w"> </span><span class="k">of</span><span class="w"> </span><span class="n">the</span><span class="w"> </span><span class="n">columns</span><span class="w"></span>

<span class="w">        </span><span class="k">return</span><span class="w"> </span><span class="n">vcf_table</span><span class="w"></span>
</code></pre></div>

</details>
<h4 id="filter_fastq">filter_fastq</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">filter_fastq</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">fastq_path</span><span class="p">,</span>
    <span class="n">target</span><span class="o">=</span><span class="kc">None</span>
<span class="p">)</span>
</code></pre></div>

<p>Filter original FASTQ file for reads that best map to the reference.</p>
<p>This function is used to obtain a FASTQ file that can be used for Canu assembly.</p>
<p><strong>Parameters</strong></p>
<p><strong>fastq_path</strong></p>
<blockquote>
<p>Path to FASTQ file (<code>str</code>).</p>
</blockquote>
<p><strong>target</strong></p>
<blockquote>
<p>Path to save the filtered FASTQ file (<code>str</code>).</p>
</blockquote>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code>    <span class="n">def</span> <span class="n">filter_fastq</span><span class="p">(</span><span class="n">self</span><span class="p">,</span> <span class="n">fastq_path</span><span class="p">,</span> <span class="n">target</span><span class="o">=</span><span class="k">None</span><span class="p">)</span><span class="o">:</span>

        <span class="s2">&quot;</span><span class="se">&quot;&quot;</span><span class="s2">Filter original FASTQ file for reads that best map to the reference.</span>

<span class="s2">        This function is used to obtain a FASTQ file that can be used for Canu assembly.</span>

<span class="s2">        **Parameters**</span>

<span class="s2">        **fastq_path**</span>

<span class="s2">        &gt; Path to FASTQ file (`str`).</span>

<span class="s2">        **target**</span>

<span class="s2">        &gt; Path to save the filtered FASTQ file (`str`).</span>

<span class="s2">        </span><span class="se">&quot;&quot;</span><span class="s2">&quot;</span>

        <span class="k">if</span> <span class="n">target</span> <span class="k">is</span> <span class="k">None</span><span class="o">:</span>

            <span class="n">target</span> <span class="o">=</span> <span class="n">fastq_path</span> <span class="o">+</span> <span class="s2">&quot;_filtered.fastq&quot;</span>

        <span class="n">read_names</span> <span class="o">=</span> <span class="n">self</span><span class="p">.</span><span class="n">paf</span><span class="err">[</span><span class="s2">&quot;query_name&quot;</span><span class="err">]</span><span class="p">.</span><span class="n">to_list</span><span class="p">()</span>

        <span class="n">input_seq_iterator</span> <span class="o">=</span> <span class="n">SeqIO</span><span class="p">.</span><span class="n">parse</span><span class="p">(</span><span class="n">fastq_path</span><span class="p">,</span> <span class="s2">&quot;fastq&quot;</span><span class="p">)</span>

        <span class="n">seq_iterator</span> <span class="o">=</span> <span class="p">(</span>

            <span class="n">record</span> <span class="k">for</span> <span class="n">record</span> <span class="k">in</span> <span class="n">input_seq_iterator</span> <span class="k">if</span> <span class="n">record</span><span class="p">.</span><span class="k">name</span> <span class="k">in</span> <span class="n">read_names</span>

        <span class="p">)</span>

        <span class="n">SeqIO</span><span class="p">.</span><span class="k">write</span><span class="p">(</span><span class="n">seq_iterator</span><span class="p">,</span> <span class="n">target</span><span class="p">,</span> <span class="s2">&quot;fastq&quot;</span><span class="p">)</span>
</code></pre></div>

</details>
<h4 id="find_big_inserts">find_big_inserts</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">find_big_inserts</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">threshold</span><span class="o">=</span><span class="mi">100</span>
<span class="p">)</span>
</code></pre></div>

<p>Calculate % of reads with big inserts.</p>
<p><strong>Parameters</strong></p>
<p><strong>threshold</strong></p>
<blockquote>
<p>Size of insert in bp to be considered big (<code>int</code>).</p>
</blockquote>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code>    <span class="n">def</span> <span class="n">find_big_inserts</span><span class="p">(</span><span class="n">self</span><span class="p">,</span> <span class="n">threshold</span><span class="o">=</span><span class="mi">100</span><span class="p">)</span><span class="o">:</span>

        <span class="s2">&quot;</span><span class="se">&quot;&quot;</span><span class="s2">Calculate % of reads with big inserts.</span>

<span class="s2">        **Parameters**</span>

<span class="s2">        **threshold**</span>

<span class="s2">        &gt; Size of insert in bp to be considered big (`int`).</span><span class="se">&quot;&quot;</span><span class="s2">&quot;</span>

        <span class="n">read_insert_dict</span> <span class="o">=</span> <span class="err">{</span>

            <span class="n">read_id</span><span class="o">:</span> <span class="no">False</span> <span class="k">for</span> <span class="n">read_id</span> <span class="k">in</span> <span class="n">self</span><span class="p">.</span><span class="n">paf</span><span class="err">[</span><span class="s2">&quot;query_name&quot;</span><span class="err">]</span><span class="p">.</span><span class="k">unique</span><span class="p">()</span>

        <span class="err">}</span>  <span class="c1"># Initialize. Value True when read has insert.</span>

        <span class="k">for</span> <span class="k">index</span><span class="p">,</span> <span class="k">row</span> <span class="k">in</span> <span class="n">self</span><span class="p">.</span><span class="n">paf</span><span class="p">.</span><span class="n">iterrows</span><span class="p">()</span><span class="o">:</span>

            <span class="n">read_id</span> <span class="o">=</span> <span class="k">row</span><span class="err">[</span><span class="s2">&quot;query_name&quot;</span><span class="err">]</span>

            <span class="n">cigar</span> <span class="o">=</span> <span class="k">row</span><span class="p">.</span><span class="n">tail</span><span class="p">(</span><span class="mi">1</span><span class="p">).</span><span class="n">item</span><span class="p">()</span>  <span class="c1"># CIGAR string column is the last</span>

            <span class="k">for</span> <span class="n">size</span><span class="p">,</span> <span class="n">mutation</span> <span class="k">in</span> <span class="n">re</span><span class="p">.</span><span class="n">findall</span><span class="p">(</span><span class="n">r</span><span class="s2">&quot;(</span><span class="err">\</span><span class="s2">d+)([A-Z]{1})&quot;</span><span class="p">,</span> <span class="n">cigar</span><span class="p">)</span><span class="o">:</span>

                <span class="k">if</span> <span class="kt">int</span><span class="p">(</span><span class="n">size</span><span class="p">)</span> <span class="o">&gt;=</span> <span class="n">threshold</span><span class="o">:</span>

                    <span class="k">if</span> <span class="n">mutation</span> <span class="o">==</span> <span class="s2">&quot;I&quot;</span><span class="o">:</span>  <span class="c1"># CIGAR code for insert</span>

                        <span class="n">read_insert_dict</span><span class="err">[</span><span class="n">read_id</span><span class="err">]</span> <span class="o">=</span> <span class="no">True</span>

                        <span class="n">break</span>

        <span class="n">fraction_of_big_inserts</span> <span class="o">=</span> <span class="nf">sum</span><span class="p">(</span><span class="n">read_insert_dict</span><span class="p">.</span><span class="k">values</span><span class="p">())</span> <span class="o">/</span> <span class="n">len</span><span class="p">(</span>

            <span class="n">read_insert_dict</span><span class="p">.</span><span class="k">keys</span><span class="p">()</span>

        <span class="p">)</span>

        <span class="c1"># imprecise rounding, but ok for our purposes:</span>

        <span class="n">self</span><span class="p">.</span><span class="n">pct_big_insert</span> <span class="o">=</span> <span class="nf">round</span><span class="p">(</span><span class="n">fraction_of_big_inserts</span> <span class="o">*</span> <span class="mi">100</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>  <span class="c1"># pct multiplier</span>
</code></pre></div>

</details>
<h4 id="get_weak_read_histogram_data">get_weak_read_histogram_data</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">get_weak_read_histogram_data</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">cutoff</span>
<span class="p">)</span>
</code></pre></div>

<p>Create data for the plotting function.</p>
<p><strong>Parameters</strong></p>
<p><strong>cutoff</strong></p>
<blockquote>
<p>Cutoff for proportion of bases that map (<code>float</code>). Reads below the cutoff are
viewed as weak reads. Recommended value: 0.8 (80%).</p>
</blockquote>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code>    <span class="n">def</span> <span class="n">get_weak_read_histogram_data</span><span class="p">(</span><span class="n">self</span><span class="p">,</span> <span class="n">cutoff</span><span class="p">)</span><span class="o">:</span>

        <span class="s2">&quot;</span><span class="se">&quot;&quot;</span><span class="s2">Create data for the plotting function.</span>

<span class="s2">        **Parameters**</span>

<span class="s2">        **cutoff**</span>

<span class="s2">        &gt; Cutoff for proportion of bases that map (`float`). Reads below the cutoff are</span>

<span class="s2">        viewed as weak reads. Recommended value: 0.8 (80%).</span>

<span class="s2">        </span><span class="se">&quot;&quot;</span><span class="s2">&quot;</span>

        <span class="c1"># For each read, calculate the proportion of bases that map, then create an array</span>

        <span class="c1"># for each read that is below the cutoff. The sum of these arrays will give the</span>

        <span class="c1"># depth of aligning bad reads for each base of the reference.</span>

        <span class="n">self</span><span class="p">.</span><span class="n">proportion_maps</span> <span class="o">=</span> <span class="p">(</span>

            <span class="n">self</span><span class="p">.</span><span class="n">paf</span><span class="err">[</span><span class="s2">&quot;mapping_matches&quot;</span><span class="err">]</span> <span class="o">/</span> <span class="n">self</span><span class="p">.</span><span class="n">paf</span><span class="err">[</span><span class="s2">&quot;query_length&quot;</span><span class="err">]</span><span class="p">.</span><span class="n">to_list</span><span class="p">()</span>

        <span class="p">)</span>

        <span class="n">filtered_paf</span> <span class="o">=</span> <span class="n">self</span><span class="p">.</span><span class="n">paf</span><span class="err">[</span><span class="n">self</span><span class="p">.</span><span class="n">proportion_maps</span> <span class="o">&lt;</span> <span class="n">cutoff</span><span class="err">]</span>

        <span class="n">array_length</span> <span class="o">=</span> <span class="n">filtered_paf</span><span class="p">.</span><span class="n">iloc</span><span class="err">[</span><span class="mi">0</span><span class="err">][</span><span class="s2">&quot;target_length&quot;</span><span class="err">]</span>

        <span class="n">histogram_array</span> <span class="o">=</span> <span class="err">[]</span>

        <span class="k">for</span> <span class="k">index</span><span class="p">,</span> <span class="k">row</span> <span class="k">in</span> <span class="n">filtered_paf</span><span class="p">.</span><span class="n">iterrows</span><span class="p">()</span><span class="o">:</span>

            <span class="n">read_array</span> <span class="o">=</span> <span class="err">[</span><span class="mi">0</span><span class="err">]</span> <span class="o">*</span> <span class="n">array_length</span>

            <span class="k">start</span> <span class="o">=</span> <span class="k">row</span><span class="err">[</span><span class="s2">&quot;target_start&quot;</span><span class="err">]</span>

            <span class="k">end</span> <span class="o">=</span> <span class="k">row</span><span class="err">[</span><span class="s2">&quot;target_end&quot;</span><span class="err">]</span>

            <span class="k">value</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">/</span> <span class="k">row</span><span class="err">[</span><span class="s2">&quot;multialign&quot;</span><span class="err">]</span>

            <span class="k">values</span> <span class="o">=</span> <span class="err">[</span><span class="k">value</span><span class="err">]</span> <span class="o">*</span> <span class="p">(</span><span class="k">end</span> <span class="o">-</span> <span class="k">start</span><span class="p">)</span>

            <span class="n">read_array</span><span class="err">[</span><span class="k">start</span><span class="o">:</span><span class="k">end</span><span class="err">]</span> <span class="o">=</span> <span class="k">values</span>

            <span class="n">histogram_array</span><span class="p">.</span><span class="n">append</span><span class="p">(</span><span class="n">read_array</span><span class="p">)</span>

        <span class="n">histogram_numpy_array</span> <span class="o">=</span> <span class="n">numpy</span><span class="p">.</span><span class="k">array</span><span class="p">(</span><span class="n">histogram_array</span><span class="p">)</span>

        <span class="n">self</span><span class="p">.</span><span class="n">zz</span> <span class="o">=</span> <span class="n">numpy</span><span class="p">.</span><span class="nf">sum</span><span class="p">(</span><span class="n">histogram_numpy_array</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
</code></pre></div>

</details>
<h4 id="perform_comparison">perform_comparison</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">perform_comparison</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">assembly_path</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">vcf_path</span><span class="o">=</span><span class="kc">None</span>
<span class="p">)</span>
</code></pre></div>

<p>Plot coverage and compare reference with a consensus sequence.</p>
<p><strong>Parameters</strong></p>
<p><strong>assembly_path</strong></p>
<blockquote>
<p>Optional. Path to a consensus FASTA file (<code>str</code>).</p>
</blockquote>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code>    <span class="n">def</span> <span class="n">perform_comparison</span><span class="p">(</span><span class="n">self</span><span class="p">,</span> <span class="n">assembly_path</span><span class="o">=</span><span class="k">None</span><span class="p">,</span> <span class="n">vcf_path</span><span class="o">=</span><span class="k">None</span><span class="p">)</span><span class="o">:</span>

        <span class="s2">&quot;</span><span class="se">&quot;&quot;</span><span class="s2">Plot coverage and compare reference with a consensus sequence.</span>

<span class="s2">        **Parameters**</span>

<span class="s2">        **assembly_path**</span>

<span class="s2">        &gt; Optional. Path to a consensus FASTA file (`str`).</span>

<span class="s2">        </span><span class="se">&quot;&quot;</span><span class="s2">&quot;</span>

        <span class="n">self</span><span class="p">.</span><span class="n">fig</span> <span class="o">=</span> <span class="n">self</span><span class="p">.</span><span class="n">plot_coverage</span><span class="p">()</span>

        <span class="k">if</span> <span class="n">assembly_path</span> <span class="k">is</span> <span class="k">not</span> <span class="k">None</span><span class="o">:</span>

            <span class="n">self</span><span class="p">.</span><span class="n">comparison_figure</span> <span class="o">=</span> <span class="n">self</span><span class="p">.</span><span class="n">compare_with_assembly</span><span class="p">(</span>

                <span class="n">assembly_path</span><span class="o">=</span><span class="n">assembly_path</span>

            <span class="p">)</span>

            <span class="n">self</span><span class="p">.</span><span class="n">has_consensus</span> <span class="o">=</span> <span class="no">True</span>

        <span class="k">if</span> <span class="n">vcf_path</span> <span class="k">is</span> <span class="k">not</span> <span class="k">None</span><span class="o">:</span>

            <span class="n">self</span><span class="p">.</span><span class="n">vcf_table</span> <span class="o">=</span> <span class="n">self</span><span class="p">.</span><span class="n">create_vcf_table</span><span class="p">(</span><span class="n">vcf_path</span><span class="o">=</span><span class="n">vcf_path</span><span class="p">)</span>
</code></pre></div>

</details>
<h4 id="plot_coverage">plot_coverage</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">plot_coverage</span><span class="p">(</span>
    <span class="bp">self</span>
<span class="p">)</span>
</code></pre></div>

<p>Plot the reference with the coverage and weak reads.</p>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code><span class="w">    </span><span class="n">def</span><span class="w"> </span><span class="n">plot_coverage</span><span class="p">(</span><span class="n">self</span><span class="p">)</span><span class="err">:</span><span class="w"></span>

<span class="w">        </span><span class="ss">&quot;&quot;&quot;Plot the reference with the coverage and weak reads.&quot;&quot;&quot;</span><span class="w"></span>

<span class="w">        </span><span class="k">if</span><span class="w"> </span><span class="ow">not</span><span class="w"> </span><span class="n">hasattr</span><span class="p">(</span><span class="n">self</span><span class="p">,</span><span class="w"> </span><span class="ss">&quot;xx&quot;</span><span class="p">)</span><span class="err">:</span><span class="w"></span>

<span class="w">            </span><span class="n">self</span><span class="p">.</span><span class="n">calculate_stats</span><span class="p">()</span><span class="w">  </span><span class="err">#</span><span class="w"> </span><span class="n">also</span><span class="w"> </span><span class="n">calculates</span><span class="w"> </span><span class="n">yy</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="n">zz</span><span class="w"></span>

<span class="w">        </span><span class="err">#</span><span class="w"> </span><span class="n">Plot</span><span class="w"></span>

<span class="w">        </span><span class="err">######</span><span class="w"></span>

<span class="w">        </span><span class="n">fig</span><span class="p">,</span><span class="w"> </span><span class="p">(</span><span class="n">ax1</span><span class="p">,</span><span class="w"> </span><span class="n">ax2</span><span class="p">)</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">plt</span><span class="p">.</span><span class="n">subplots</span><span class="p">(</span><span class="w"></span>

<span class="w">            </span><span class="mi">2</span><span class="p">,</span><span class="w"> </span><span class="mi">1</span><span class="p">,</span><span class="w"> </span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">7</span><span class="p">,</span><span class="w"> </span><span class="mi">3</span><span class="p">),</span><span class="w"> </span><span class="n">sharex</span><span class="o">=</span><span class="k">True</span><span class="p">,</span><span class="w"> </span><span class="n">gridspec_kw</span><span class="o">=</span><span class="err">{</span><span class="ss">&quot;height_ratios&quot;</span><span class="err">:</span><span class="w"> </span><span class="o">[</span><span class="n">4, 1</span><span class="o">]</span><span class="err">}</span><span class="w"></span>

<span class="w">        </span><span class="p">)</span><span class="w"></span>

<span class="w">        </span><span class="k">if</span><span class="w"> </span><span class="n">self</span><span class="p">.</span><span class="nl">dnacauldron</span><span class="p">:</span><span class="w"></span>

<span class="w">            </span><span class="n">graphic_record</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">CustomTranslator</span><span class="p">().</span><span class="n">translate_record</span><span class="p">(</span><span class="n">self</span><span class="p">.</span><span class="n">record</span><span class="p">)</span><span class="w"></span>

<span class="w">        </span><span class="k">else</span><span class="err">:</span><span class="w"></span>

<span class="w">            </span><span class="n">graphic_record</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">dna_features_viewer</span><span class="p">.</span><span class="n">BiopythonTranslator</span><span class="p">().</span><span class="n">translate_record</span><span class="p">(</span><span class="w"></span>

<span class="w">                </span><span class="n">self</span><span class="p">.</span><span class="n">record</span><span class="w"></span>

<span class="w">            </span><span class="p">)</span><span class="w"></span>

<span class="w">        </span><span class="n">graphic_record</span><span class="p">.</span><span class="n">plot</span><span class="p">(</span><span class="n">ax</span><span class="o">=</span><span class="n">ax1</span><span class="p">,</span><span class="w"> </span><span class="n">with_ruler</span><span class="o">=</span><span class="k">False</span><span class="p">,</span><span class="w"> </span><span class="n">strand_in_label_threshold</span><span class="o">=</span><span class="mi">4</span><span class="p">)</span><span class="w"></span>

<span class="w">        </span><span class="err">#</span><span class="w"> </span><span class="n">Plot</span><span class="w"> </span><span class="n">coverage</span><span class="w"></span>

<span class="w">        </span><span class="n">ax2</span><span class="p">.</span><span class="n">fill_between</span><span class="p">(</span><span class="n">self</span><span class="p">.</span><span class="n">xx</span><span class="p">,</span><span class="w"> </span><span class="n">self</span><span class="p">.</span><span class="n">yy</span><span class="p">,</span><span class="w"> </span><span class="n">alpha</span><span class="o">=</span><span class="mf">0.8</span><span class="p">)</span><span class="w"></span>

<span class="w">        </span><span class="n">stdev_yy</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="k">statistics</span><span class="p">.</span><span class="nf">stdev</span><span class="p">(</span><span class="n">self</span><span class="p">.</span><span class="n">yy</span><span class="p">)</span><span class="w"></span>

<span class="w">        </span><span class="n">ax2</span><span class="p">.</span><span class="n">set_ylim</span><span class="p">(</span><span class="n">bottom</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span><span class="w"></span>

<span class="w">        </span><span class="n">ax2</span><span class="p">.</span><span class="n">set_ylabel</span><span class="p">(</span><span class="ss">&quot;Depth&quot;</span><span class="p">)</span><span class="w"></span>

<span class="w">        </span><span class="n">ax2</span><span class="p">.</span><span class="n">set_xlabel</span><span class="p">(</span><span class="ss">&quot;Base [bp]&quot;</span><span class="p">)</span><span class="w"></span>

<span class="w">        </span><span class="n">ax2</span><span class="p">.</span><span class="n">axhline</span><span class="p">(</span><span class="w"></span>

<span class="w">            </span><span class="n">y</span><span class="o">=</span><span class="n">self</span><span class="p">.</span><span class="n">median_yy</span><span class="p">,</span><span class="w"> </span><span class="n">xmin</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span><span class="w"> </span><span class="n">xmax</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span><span class="w"> </span><span class="n">color</span><span class="o">=</span><span class="ss">&quot;grey&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">linestyle</span><span class="o">=</span><span class="ss">&quot;-&quot;</span><span class="p">,</span><span class="w"> </span><span class="n">linewidth</span><span class="o">=</span><span class="mf">0.8</span><span class="w"></span>

<span class="w">        </span><span class="p">)</span><span class="w"></span>

<span class="w">        </span><span class="n">ax2</span><span class="p">.</span><span class="n">axhline</span><span class="p">(</span><span class="w"></span>

<span class="w">            </span><span class="n">y</span><span class="o">=</span><span class="n">self</span><span class="p">.</span><span class="n">median_yy</span><span class="w"> </span><span class="o">-</span><span class="w"> </span><span class="n">stdev_yy</span><span class="p">,</span><span class="w"></span>

<span class="w">            </span><span class="n">xmin</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span><span class="w"></span>

<span class="w">            </span><span class="n">xmax</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span><span class="w"></span>

<span class="w">            </span><span class="n">color</span><span class="o">=</span><span class="ss">&quot;grey&quot;</span><span class="p">,</span><span class="w"></span>

<span class="w">            </span><span class="n">linestyle</span><span class="o">=</span><span class="ss">&quot;--&quot;</span><span class="p">,</span><span class="w"></span>

<span class="w">            </span><span class="n">linewidth</span><span class="o">=</span><span class="mf">0.8</span><span class="p">,</span><span class="w"></span>

<span class="w">        </span><span class="p">)</span><span class="w"></span>

<span class="w">        </span><span class="n">ax2</span><span class="p">.</span><span class="n">axhline</span><span class="p">(</span><span class="w"></span>

<span class="w">            </span><span class="n">y</span><span class="o">=</span><span class="n">self</span><span class="p">.</span><span class="n">median_yy</span><span class="w"> </span><span class="o">+</span><span class="w"> </span><span class="n">stdev_yy</span><span class="p">,</span><span class="w"></span>

<span class="w">            </span><span class="n">xmin</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span><span class="w"></span>

<span class="w">            </span><span class="n">xmax</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span><span class="w"></span>

<span class="w">            </span><span class="n">color</span><span class="o">=</span><span class="ss">&quot;grey&quot;</span><span class="p">,</span><span class="w"></span>

<span class="w">            </span><span class="n">linestyle</span><span class="o">=</span><span class="ss">&quot;--&quot;</span><span class="p">,</span><span class="w"></span>

<span class="w">            </span><span class="n">linewidth</span><span class="o">=</span><span class="mf">0.8</span><span class="p">,</span><span class="w"></span>

<span class="w">        </span><span class="p">)</span><span class="w"></span>

<span class="w">        </span><span class="k">return</span><span class="w"> </span><span class="n">fig</span><span class="w"></span>
</code></pre></div>

</details>
<h3 id="comparatorgroup">ComparatorGroup</h3>
<div class="codehilite"><pre><span></span><code><span class="k">class</span> <span class="nc">ComparatorGroup</span><span class="p">(</span>
    <span class="n">references</span><span class="p">,</span>
    <span class="n">alignments</span><span class="p">,</span>
    <span class="n">barcode</span><span class="o">=</span><span class="s1">&#39;barcode&#39;</span><span class="p">,</span>
    <span class="n">assembly_paths</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">vcf_paths</span><span class="o">=</span><span class="kc">None</span>
<span class="p">)</span>
</code></pre></div>

<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code><span class="s s-Atom">class</span> <span class="nv">ComparatorGroup</span><span class="s s-Atom">:</span>

    <span class="s2">&quot;&quot;&quot;Analyse alignments to a set of references.</span>

<span class="s2">    **Parameters**</span>

<span class="s2">    **references**</span>

<span class="s2">    &gt; Dictionary of name: Biopython record.</span>

<span class="s2">    **alignments**</span>

<span class="s2">    &gt; Dictionary of pandas dataframes: {&quot;</span><span class="s s-Atom">paf</span><span class="s2">&quot;: df_paf, &quot;</span><span class="s s-Atom">tsv</span><span class="s2">&quot;: df_tsv}. The PAF dataframe</span>

<span class="s2">    was created from the output of `paftools.js sam2paf` and the TSV dataframe was</span>

<span class="s2">    created from the output of `samtools depth -aa`.</span>

<span class="s2">    **barcode**</span>

<span class="s2">    &gt; The barcode number or ID (`str`).</span>

<span class="s2">    **assembly_paths**</span>

<span class="s2">    &gt; Dictionary of sample: consensus sequence path (`dict`).</span>

<span class="s2">    **vcf_paths**</span>

<span class="s2">    &gt; Dictionary of sample: filtered VCF filepath (`dict`).</span>

<span class="s2">    &quot;&quot;&quot;</span>

    <span class="s s-Atom">def</span> <span class="k">__</span><span class="nf">init__</span><span class="p">(</span>

        <span class="s s-Atom">self</span><span class="p">,</span>

        <span class="s s-Atom">references</span><span class="p">,</span>

        <span class="s s-Atom">alignments</span><span class="p">,</span>

        <span class="s s-Atom">barcode=</span><span class="s2">&quot;barcode&quot;</span><span class="p">,</span>

        <span class="s s-Atom">assembly_paths</span><span class="o">=</span><span class="nv">None</span><span class="p">,</span>

        <span class="s s-Atom">vcf_paths</span><span class="o">=</span><span class="nv">None</span><span class="p">,</span>

    <span class="p">)</span><span class="s s-Atom">:</span>

        <span class="s s-Atom">self</span><span class="p">.</span><span class="s s-Atom">references</span> <span class="o">=</span> <span class="s s-Atom">references</span>

        <span class="s s-Atom">self</span><span class="p">.</span><span class="s s-Atom">paf</span> <span class="o">=</span> <span class="s s-Atom">alignments</span><span class="p">[</span><span class="s2">&quot;paf&quot;</span><span class="p">]</span>

        <span class="s s-Atom">self</span><span class="p">.</span><span class="s s-Atom">tsv</span> <span class="o">=</span> <span class="s s-Atom">alignments</span><span class="p">[</span><span class="s2">&quot;tsv&quot;</span><span class="p">]</span>

        <span class="s s-Atom">self</span><span class="p">.</span><span class="s s-Atom">comparators</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="s s-Atom">self</span><span class="p">.</span><span class="s s-Atom">barcode</span> <span class="o">=</span> <span class="s s-Atom">barcode</span>

        <span class="s s-Atom">self</span><span class="p">.</span><span class="s s-Atom">assembly_paths</span> <span class="o">=</span> <span class="s s-Atom">assembly_paths</span>

        <span class="s s-Atom">self</span><span class="p">.</span><span class="s s-Atom">vcf_paths</span> <span class="o">=</span> <span class="s s-Atom">vcf_paths</span>

    <span class="s s-Atom">def</span> <span class="nf">create_comparator</span><span class="p">(</span><span class="s s-Atom">self</span><span class="p">,</span> <span class="s s-Atom">reference_name</span><span class="p">)</span><span class="s s-Atom">:</span>

        <span class="s2">&quot;&quot;&quot;Create a Comparator instance from one of the loaded references.</span>

<span class="s2">        **Parameters**</span>

<span class="s2">        **reference_name**</span>

<span class="s2">        &gt; The name of the reference (`str`).</span>

<span class="s2">        &quot;&quot;&quot;</span>

        <span class="s s-Atom">subset_paf</span> <span class="o">=</span> <span class="s s-Atom">self</span><span class="p">.</span><span class="nf">subset_paf</span><span class="p">(</span><span class="s s-Atom">reference_name</span><span class="p">)</span>

        <span class="s s-Atom">subset_tsv</span> <span class="o">=</span> <span class="s s-Atom">self</span><span class="p">.</span><span class="s s-Atom">tsv</span><span class="p">[</span><span class="s s-Atom">self</span><span class="p">.</span><span class="s s-Atom">tsv</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s s-Atom">reference_name</span><span class="p">]</span>

        <span class="s s-Atom">alignment</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;paf&quot;</span><span class="s s-Atom">:</span> <span class="s s-Atom">subset_paf</span><span class="p">,</span> <span class="s2">&quot;tsv&quot;</span><span class="s s-Atom">:</span> <span class="s s-Atom">subset_tsv</span><span class="p">}</span>

        <span class="s s-Atom">comparator</span> <span class="o">=</span> <span class="nv">Comparator</span><span class="p">(</span>

            <span class="p">{</span><span class="s s-Atom">reference_name:</span> <span class="s s-Atom">self</span><span class="p">.</span><span class="s s-Atom">references</span><span class="p">[</span><span class="s s-Atom">reference_name</span><span class="p">]},</span> <span class="s s-Atom">alignment</span>

        <span class="p">)</span>

        <span class="s s-Atom">return</span> <span class="s s-Atom">comparator</span>

    <span class="s s-Atom">def</span> <span class="nf">add_comparator</span><span class="p">(</span><span class="s s-Atom">self</span><span class="p">,</span> <span class="s s-Atom">reference_name</span><span class="p">)</span><span class="s s-Atom">:</span>

        <span class="s2">&quot;&quot;&quot;Create comparator and add to the group. See create_comparator().&quot;&quot;&quot;</span>

        <span class="s s-Atom">self</span><span class="p">.</span><span class="s s-Atom">comparators</span> <span class="s s-Atom">+=</span> <span class="p">[</span><span class="s s-Atom">self</span><span class="p">.</span><span class="nf">create_comparator</span><span class="p">(</span><span class="s s-Atom">reference_name</span><span class="p">)]</span>

    <span class="s s-Atom">def</span> <span class="nf">subset_paf</span><span class="p">(</span><span class="s s-Atom">self</span><span class="p">,</span> <span class="s s-Atom">reference_name</span><span class="p">)</span><span class="s s-Atom">:</span>

        <span class="s2">&quot;&quot;&quot;Subset PAF dataframe for the reference.</span>

<span class="s2">        Select reads that align to the reference and also align *best* to the reference.</span>

<span class="s2">        **Parameters**</span>

<span class="s2">        **reference_name**</span>

<span class="s2">        &gt; Name of the reference sequence to filter for (`str`).</span>

<span class="s2">        &quot;&quot;&quot;</span>

        <span class="s s-Atom">#</span> <span class="nv">Work</span> <span class="s s-Atom">only</span> <span class="s s-Atom">with</span> <span class="s s-Atom">the</span> <span class="s s-Atom">values</span> <span class="s s-Atom">used</span> <span class="s s-Atom">for</span> <span class="nn">filtering</span><span class="p">:</span>

        <span class="s s-Atom">paf3</span> <span class="o">=</span> <span class="s s-Atom">self</span><span class="p">.</span><span class="s s-Atom">paf</span><span class="p">[[</span><span class="s2">&quot;query_name&quot;</span><span class="p">,</span> <span class="s2">&quot;target_name&quot;</span><span class="p">,</span> <span class="s2">&quot;mapping_matches&quot;</span><span class="p">]]</span>

        <span class="s s-Atom">grouped</span> <span class="o">=</span> <span class="s s-Atom">paf3</span><span class="p">.</span><span class="nf">groupby</span><span class="p">(</span><span class="s2">&quot;query_name&quot;</span><span class="p">)</span>

        <span class="s s-Atom">selected_reads</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="s s-Atom">for</span> <span class="s s-Atom">name</span><span class="p">,</span> <span class="s s-Atom">group</span> <span class="s s-Atom">in</span> <span class="nn">grouped</span><span class="p">:</span>

            <span class="s s-Atom">#</span> <span class="o">not</span> <span class="s s-Atom">all</span> <span class="s s-Atom">reads</span> <span class="s s-Atom">align</span> <span class="s s-Atom">to</span> <span class="s s-Atom">the</span> <span class="nn">reference</span><span class="p">:</span>

            <span class="s s-Atom">if</span> <span class="s s-Atom">reference_name</span> <span class="s s-Atom">in</span> <span class="s s-Atom">group</span><span class="p">[</span><span class="s2">&quot;target_name&quot;</span><span class="p">].</span><span class="nn">values</span><span class="p">:</span>

                <span class="s s-Atom">max_matches</span> <span class="o">=</span> <span class="s s-Atom">group</span><span class="p">[</span><span class="s2">&quot;mapping_matches&quot;</span><span class="p">].</span><span class="nf">max</span><span class="p">()</span>

                <span class="s s-Atom">#</span> <span class="s s-Atom">filter</span> <span class="s s-Atom">that</span> <span class="s s-Atom">the</span> <span class="s s-Atom">read</span> <span class="s s-Atom">aligns</span> <span class="s s-Atom">best</span> <span class="s s-Atom">to</span> <span class="s s-Atom">this</span> <span class="nn">reference</span><span class="p">:</span>

                <span class="nf">if</span> <span class="p">(</span>

                    <span class="nf">max</span><span class="p">(</span>

                        <span class="s s-Atom">group</span><span class="p">[</span><span class="s s-Atom">group</span><span class="p">[</span><span class="s2">&quot;target_name&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s s-Atom">reference_name</span><span class="p">][</span>

                            <span class="s2">&quot;mapping_matches&quot;</span>

                        <span class="p">].</span><span class="s s-Atom">values</span>

                    <span class="p">)</span>

                    <span class="o">==</span> <span class="s s-Atom">max_matches</span>

                <span class="p">)</span><span class="s s-Atom">:</span>

                    <span class="s s-Atom">selected_reads</span> <span class="s s-Atom">+=</span> <span class="p">[</span><span class="s s-Atom">name</span><span class="p">]</span>

        <span class="s s-Atom">#</span> <span class="nv">Filter</span> <span class="s s-Atom">for</span> <span class="s s-Atom">selected</span> <span class="s s-Atom">reads</span> <span class="p">..</span>

        <span class="s s-Atom">subset_paf_filtered</span> <span class="o">=</span> <span class="s s-Atom">self</span><span class="p">.</span><span class="s s-Atom">paf</span><span class="p">[</span><span class="s s-Atom">self</span><span class="p">.</span><span class="s s-Atom">paf</span><span class="p">[</span><span class="s2">&quot;query_name&quot;</span><span class="p">].</span><span class="nf">isin</span><span class="p">(</span><span class="s s-Atom">selected_reads</span><span class="p">)]</span>

        <span class="s s-Atom">#</span> <span class="p">..</span><span class="s s-Atom">then</span> <span class="s s-Atom">get</span> <span class="s s-Atom">only</span> <span class="s s-Atom">the</span> <span class="s s-Atom">reads</span> <span class="s s-Atom">aligning</span> <span class="s s-Atom">to</span> <span class="nn">reference</span><span class="p">:</span>

        <span class="s s-Atom">subset_paf_final</span> <span class="o">=</span> <span class="s s-Atom">subset_paf_filtered</span><span class="p">[</span>

            <span class="s s-Atom">subset_paf_filtered</span><span class="p">[</span><span class="s2">&quot;target_name&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s s-Atom">reference_name</span>

        <span class="p">]</span>

        <span class="s s-Atom">return</span> <span class="s s-Atom">subset_paf_final</span>

    <span class="s s-Atom">def</span> <span class="nf">perform_all_comparisons</span><span class="p">(</span><span class="s s-Atom">self</span><span class="p">)</span><span class="s s-Atom">:</span>

        <span class="s2">&quot;&quot;&quot;Call perform_comparison() on each Comparator.</span>

<span class="s2">        **Parameters**</span>

<span class="s2">        **assembly_paths**</span>

<span class="s2">        &gt; Dictionary of construct name: consensus sequence file.</span>

<span class="s2">        For example: `{&quot;</span><span class="nv">Construct_1</span><span class="s2">&quot;: &quot;</span><span class="o">/</span><span class="s s-Atom">path</span><span class="o">/</span><span class="s s-Atom">to</span><span class="o">/</span><span class="s s-Atom">assembly</span><span class="p">.</span><span class="s s-Atom">fa</span><span class="s2">&quot;}`.</span>

<span class="s2">        &quot;&quot;&quot;</span>

        <span class="s s-Atom">if</span> <span class="s s-Atom">self</span><span class="p">.</span><span class="s s-Atom">assembly_paths</span> <span class="o">is</span> <span class="nv">None</span><span class="s s-Atom">:</span>

            <span class="s s-Atom">assembly_paths</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="nn">else</span><span class="p">:</span>

            <span class="s s-Atom">assembly_paths</span> <span class="o">=</span> <span class="s s-Atom">self</span><span class="p">.</span><span class="s s-Atom">assembly_paths</span>

        <span class="s s-Atom">if</span> <span class="s s-Atom">self</span><span class="p">.</span><span class="s s-Atom">vcf_paths</span> <span class="o">is</span> <span class="nv">None</span><span class="s s-Atom">:</span>

            <span class="s s-Atom">vcf_paths</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="nn">else</span><span class="p">:</span>

            <span class="s s-Atom">vcf_paths</span> <span class="o">=</span> <span class="s s-Atom">self</span><span class="p">.</span><span class="s s-Atom">vcf_paths</span>

        <span class="s s-Atom">for</span> <span class="s s-Atom">comparator</span> <span class="s s-Atom">in</span> <span class="s s-Atom">self</span><span class="p">.</span><span class="nn">comparators</span><span class="p">:</span>

            <span class="nn">try</span><span class="p">:</span>

                <span class="s s-Atom">assembly_path</span> <span class="o">=</span> <span class="s s-Atom">assembly_paths</span><span class="p">[</span><span class="s s-Atom">comparator</span><span class="p">.</span><span class="s s-Atom">name</span><span class="p">]</span>

            <span class="s s-Atom">except</span> <span class="nv">Exception</span><span class="s s-Atom">:</span>

                <span class="s s-Atom">assembly_path</span> <span class="o">=</span> <span class="nv">None</span>

            <span class="nn">try</span><span class="p">:</span>

                <span class="s s-Atom">vcf_path</span> <span class="o">=</span> <span class="s s-Atom">vcf_paths</span><span class="p">[</span><span class="s s-Atom">comparator</span><span class="p">.</span><span class="s s-Atom">name</span><span class="p">]</span>

            <span class="s s-Atom">except</span> <span class="nv">Exception</span><span class="s s-Atom">:</span>

                <span class="s s-Atom">vcf_path</span> <span class="o">=</span> <span class="nv">None</span>

            <span class="s s-Atom">comparator</span><span class="p">.</span><span class="nf">perform_comparison</span><span class="p">(</span>

                <span class="s s-Atom">assembly_path</span><span class="o">=</span><span class="s s-Atom">assembly_path</span><span class="p">,</span> <span class="s s-Atom">vcf_path</span><span class="o">=</span><span class="s s-Atom">vcf_path</span>

            <span class="p">)</span>

        <span class="s s-Atom">self</span><span class="p">.</span><span class="s s-Atom">comparisons_performed</span> <span class="o">=</span> <span class="nv">True</span>

        <span class="s s-Atom">self</span><span class="p">.</span><span class="nf">summarise_analysis</span><span class="p">()</span>

    <span class="s s-Atom">def</span> <span class="nf">summarise_analysis</span><span class="p">(</span><span class="s s-Atom">self</span><span class="p">)</span><span class="s s-Atom">:</span>

        <span class="s2">&quot;&quot;&quot;Create variables for the PDF report summary.&quot;&quot;&quot;</span>

        <span class="s s-Atom">self</span><span class="p">.</span><span class="s s-Atom">number_of_constructs</span> <span class="o">=</span> <span class="nf">len</span><span class="p">(</span><span class="s s-Atom">self</span><span class="p">.</span><span class="s s-Atom">comparators</span><span class="p">)</span>

        <span class="s s-Atom">self</span><span class="p">.</span><span class="s s-Atom">result_good</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="s s-Atom">self</span><span class="p">.</span><span class="s s-Atom">result_warning</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="s s-Atom">self</span><span class="p">.</span><span class="s s-Atom">result_error</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="s s-Atom">self</span><span class="p">.</span><span class="s s-Atom">result_uncertain</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="s s-Atom">names</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="s s-Atom">reference_lengths</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="s s-Atom">results</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="s s-Atom">number_of_reads_aligning</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="s s-Atom">median_coverages</span> <span class="o">=</span> <span class="p">[]</span>

        <span class="s s-Atom">for</span> <span class="s s-Atom">comparator</span> <span class="s s-Atom">in</span> <span class="s s-Atom">self</span><span class="p">.</span><span class="nn">comparators</span><span class="p">:</span>

            <span class="s s-Atom">names</span> <span class="s s-Atom">+=</span> <span class="p">[</span><span class="s s-Atom">comparator</span><span class="p">.</span><span class="s s-Atom">name</span><span class="p">]</span>

            <span class="s s-Atom">reference_lengths</span> <span class="s s-Atom">+=</span> <span class="p">[</span><span class="nf">str</span><span class="p">(</span><span class="s s-Atom">comparator</span><span class="p">.</span><span class="s s-Atom">reference_length</span><span class="p">)]</span>

            <span class="s s-Atom">number_of_reads_aligning</span> <span class="s s-Atom">+=</span> <span class="p">[</span>

                <span class="nf">str</span><span class="p">(</span><span class="nf">len</span><span class="p">(</span><span class="s s-Atom">comparator</span><span class="p">.</span><span class="s s-Atom">paf</span><span class="p">[</span><span class="s2">&quot;query_name&quot;</span><span class="p">].</span><span class="nf">unique</span><span class="p">()))</span>

            <span class="p">]</span>

            <span class="s s-Atom">median_coverages</span> <span class="s s-Atom">+=</span> <span class="p">[</span><span class="nf">str</span><span class="p">(</span><span class="nf">int</span><span class="p">(</span><span class="s s-Atom">comparator</span><span class="p">.</span><span class="s s-Atom">median_yy</span><span class="p">))]</span>

            <span class="s s-Atom">if</span> <span class="s s-Atom">comparator</span><span class="p">.</span><span class="s s-Atom">is_uncertain:</span>

                <span class="s s-Atom">self</span><span class="p">.</span><span class="s s-Atom">result_uncertain</span> <span class="s s-Atom">+=</span> <span class="mi">1</span>

                <span class="s s-Atom">comparator</span><span class="p">.</span><span class="s s-Atom">result</span> <span class="o">=</span> <span class="s s-Atom">result_keywords</span><span class="p">[</span><span class="s2">&quot;uncertain&quot;</span><span class="p">]</span>

            <span class="nn">else</span><span class="p">:</span>

                <span class="s s-Atom">if</span> <span class="s s-Atom">comparator</span><span class="p">.</span><span class="s s-Atom">has_errors:</span>

                    <span class="s s-Atom">self</span><span class="p">.</span><span class="s s-Atom">result_error</span> <span class="s s-Atom">+=</span> <span class="mi">1</span>

                    <span class="s s-Atom">comparator</span><span class="p">.</span><span class="s s-Atom">result</span> <span class="o">=</span> <span class="s s-Atom">result_keywords</span><span class="p">[</span><span class="s2">&quot;error&quot;</span><span class="p">]</span>

                <span class="nn">else</span><span class="p">:</span>

                    <span class="s s-Atom">if</span> <span class="s s-Atom">comparator</span><span class="p">.</span><span class="s s-Atom">has_warnings:</span>

                        <span class="s s-Atom">self</span><span class="p">.</span><span class="s s-Atom">result_warning</span> <span class="s s-Atom">+=</span> <span class="mi">1</span>

                        <span class="s s-Atom">comparator</span><span class="p">.</span><span class="s s-Atom">result</span> <span class="o">=</span> <span class="s s-Atom">result_keywords</span><span class="p">[</span><span class="s2">&quot;warning&quot;</span><span class="p">]</span>

                    <span class="nn">else</span><span class="p">:</span>

                        <span class="s s-Atom">self</span><span class="p">.</span><span class="s s-Atom">result_good</span> <span class="s s-Atom">+=</span> <span class="mi">1</span>

                        <span class="s s-Atom">comparator</span><span class="p">.</span><span class="s s-Atom">result</span> <span class="o">=</span> <span class="s s-Atom">result_keywords</span><span class="p">[</span><span class="s2">&quot;good&quot;</span><span class="p">]</span>

            <span class="s s-Atom">results</span> <span class="s s-Atom">+=</span> <span class="p">[</span><span class="s s-Atom">comparator</span><span class="p">.</span><span class="s s-Atom">result</span><span class="p">]</span>

        <span class="s s-Atom">self</span><span class="p">.</span><span class="s s-Atom">n_fastq_reads</span> <span class="o">=</span> <span class="nf">len</span><span class="p">(</span><span class="nf">set</span><span class="p">(</span><span class="s s-Atom">self</span><span class="p">.</span><span class="s s-Atom">paf</span><span class="p">.</span><span class="s s-Atom">query_name</span><span class="p">))</span>

        <span class="s s-Atom">self</span><span class="p">.</span><span class="s s-Atom">fastq_plot</span> <span class="o">=</span> <span class="s s-Atom">self</span><span class="p">.</span><span class="nf">plot_fastq_histogram</span><span class="p">()</span>

        <span class="s s-Atom">#</span> <span class="nv">Summary</span> <span class="s s-Atom">table</span>

        <span class="s s-Atom">d</span> <span class="o">=</span> <span class="p">{</span>

            <span class="s2">&quot;Name&quot;</span><span class="s s-Atom">:</span> <span class="s s-Atom">pandas</span><span class="p">.</span><span class="nv">Series</span><span class="p">(</span><span class="s s-Atom">names</span><span class="p">),</span>

            <span class="s2">&quot;Result&quot;</span><span class="s s-Atom">:</span> <span class="s s-Atom">pandas</span><span class="p">.</span><span class="nv">Series</span><span class="p">(</span><span class="s s-Atom">results</span><span class="p">),</span>

            <span class="s2">&quot;Length [bp]&quot;</span><span class="s s-Atom">:</span> <span class="s s-Atom">pandas</span><span class="p">.</span><span class="nv">Series</span><span class="p">(</span><span class="s s-Atom">reference_lengths</span><span class="p">),</span>

            <span class="s2">&quot;FASTQ reads&quot;</span><span class="s s-Atom">:</span> <span class="s s-Atom">pandas</span><span class="p">.</span><span class="nv">Series</span><span class="p">(</span><span class="s s-Atom">number_of_reads_aligning</span><span class="p">),</span>

            <span class="s2">&quot;Coverage [x]&quot;</span><span class="s s-Atom">:</span> <span class="s s-Atom">pandas</span><span class="p">.</span><span class="nv">Series</span><span class="p">(</span><span class="s s-Atom">median_coverages</span><span class="p">),</span>

        <span class="p">}</span>

        <span class="s s-Atom">self</span><span class="p">.</span><span class="s s-Atom">summary_table</span> <span class="o">=</span> <span class="s s-Atom">pandas</span><span class="p">.</span><span class="nv">DataFrame</span><span class="p">(</span><span class="s s-Atom">d</span><span class="p">)</span>

        <span class="s s-Atom">#</span> <span class="s s-Atom">plt</span><span class="p">.</span><span class="nf">close</span><span class="p">(</span><span class="s2">&quot;all&quot;</span><span class="p">)</span>

    <span class="s s-Atom">def</span> <span class="nf">plot_fastq_histogram</span><span class="p">(</span><span class="s s-Atom">self</span><span class="p">,</span> <span class="s s-Atom">n_bins</span><span class="o">=</span><span class="mi">50</span><span class="p">)</span><span class="s s-Atom">:</span>

        <span class="s2">&quot;&quot;&quot;Plot a histogram of the FASTQ reads.</span>

<span class="s2">        **Parameters**</span>

<span class="s2">        **n_bins**</span>

<span class="s2">        &gt; Number of bins in the histogram (`int`).</span>

<span class="s2">        &quot;&quot;&quot;</span>

        <span class="s s-Atom">fig</span><span class="p">,</span> <span class="s s-Atom">ax</span> <span class="o">=</span> <span class="s s-Atom">plt</span><span class="p">.</span><span class="nf">subplots</span><span class="p">()</span>

        <span class="s s-Atom">plt</span><span class="p">.</span><span class="nf">hist</span><span class="p">(</span>

            <span class="s s-Atom">self</span><span class="p">.</span><span class="s s-Atom">paf</span><span class="p">[</span><span class="s2">&quot;query_length&quot;</span><span class="p">],</span>

            <span class="s s-Atom">bins</span><span class="o">=</span><span class="nf">int</span><span class="p">(</span><span class="s s-Atom">self</span><span class="p">.</span><span class="s s-Atom">paf</span><span class="p">.</span><span class="s s-Atom">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="s2">&quot;target_length&quot;</span><span class="p">]</span> <span class="o">/</span> <span class="s s-Atom">n_bins</span><span class="p">),</span>

            <span class="s s-Atom">alpha</span><span class="o">=</span><span class="mf">0.8</span><span class="p">,</span>

        <span class="p">)</span>

        <span class="s s-Atom">ax</span><span class="p">.</span><span class="nf">set_ylabel</span><span class="p">(</span><span class="s2">&quot;Number of reads&quot;</span><span class="p">)</span>

        <span class="s s-Atom">ax</span><span class="p">.</span><span class="nf">set_xlabel</span><span class="p">(</span><span class="s2">&quot;Read length [bp]&quot;</span><span class="p">)</span>

        <span class="s s-Atom">#</span> <span class="nv">Mark</span> <span class="s s-Atom">expected</span> <span class="s s-Atom">lengths</span> <span class="s s-Atom">for</span> <span class="s s-Atom">interpreting</span> <span class="s s-Atom">the</span> <span class="nn">plot</span><span class="p">:</span>

        <span class="s s-Atom">for</span> <span class="s s-Atom">comparator</span> <span class="s s-Atom">in</span> <span class="s s-Atom">self</span><span class="p">.</span><span class="nn">comparators</span><span class="p">:</span>

            <span class="s s-Atom">plt</span><span class="p">.</span><span class="nf">axvline</span><span class="p">(</span><span class="s s-Atom">x</span><span class="o">=</span><span class="nf">len</span><span class="p">(</span><span class="s s-Atom">comparator</span><span class="p">.</span><span class="s s-Atom">record</span><span class="p">),</span> <span class="s s-Atom">color=</span><span class="s2">&quot;#fd5a31&quot;</span><span class="p">)</span>

        <span class="s s-Atom">return</span> <span class="s s-Atom">fig</span>

    <span class="s s-Atom">@staticmethod</span>

    <span class="s s-Atom">def</span> <span class="nf">load_paf</span><span class="p">(</span><span class="s s-Atom">paf_path</span><span class="p">)</span><span class="s s-Atom">:</span>

        <span class="s2">&quot;&quot;&quot;Create a dataframe from a PAF file of alignments.</span>

<span class="s2">        **Parameters**</span>

<span class="s2">        **paf_path**</span>

<span class="s2">        &gt; Path (`str`) to PAF file: `paftools.js sam2paf aln.sam &gt; aln.paf`.</span>

<span class="s2">        &quot;&quot;&quot;</span>

        <span class="nn">try</span><span class="p">:</span>

            <span class="s s-Atom">paf</span> <span class="o">=</span> <span class="s s-Atom">pandas</span><span class="p">.</span><span class="nf">read_csv</span><span class="p">(</span><span class="s s-Atom">paf_path</span><span class="p">,</span> <span class="s s-Atom">sep=</span><span class="s2">&quot;\t&quot;</span><span class="p">,</span> <span class="s s-Atom">header</span><span class="o">=</span><span class="nv">None</span><span class="p">)</span>

        <span class="s s-Atom">except</span> <span class="nv">Exception</span><span class="s s-Atom">:</span>  <span class="s s-Atom">#</span> <span class="s s-Atom">unequal</span> <span class="s s-Atom">number</span> <span class="s s-Atom">of</span> <span class="s s-Atom">columns</span><span class="p">,</span> <span class="s s-Atom">the</span> <span class="s s-Atom">first</span> <span class="mi">12</span> <span class="o">is</span> <span class="s s-Atom">assumed</span> <span class="s s-Atom">to</span> <span class="s s-Atom">be</span> <span class="nv">OK</span>

            <span class="s s-Atom">#</span> <span class="nv">The</span> <span class="s s-Atom">last</span> <span class="s s-Atom">column</span> <span class="o">is</span> <span class="nv">CIGAR</span><span class="p">,</span> <span class="s s-Atom">so</span> <span class="s s-Atom">we</span> <span class="s s-Atom">need</span> <span class="s s-Atom">to</span> <span class="s s-Atom">read</span> <span class="s s-Atom">lines</span> <span class="s s-Atom">one</span> <span class="s s-Atom">by</span> <span class="nn">one</span><span class="p">:</span>

            <span class="s s-Atom">rows</span> <span class="o">=</span> <span class="p">[]</span>

            <span class="s s-Atom">with</span> <span class="nf">open</span><span class="p">(</span><span class="s s-Atom">paf_path</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="s s-Atom">as</span> <span class="s s-Atom">f_input:</span>

                <span class="s s-Atom">paf_lines</span> <span class="o">=</span> <span class="s s-Atom">f_input</span><span class="p">.</span><span class="nf">read</span><span class="p">().</span><span class="nf">splitlines</span><span class="p">()</span>

                <span class="s s-Atom">for</span> <span class="s s-Atom">paf_line</span> <span class="s s-Atom">in</span> <span class="s s-Atom">paf_lines:</span>

                    <span class="s s-Atom">paf_line</span> <span class="o">=</span> <span class="s s-Atom">paf_line</span><span class="p">.</span><span class="nf">split</span><span class="p">(</span><span class="s2">&quot;\t&quot;</span><span class="p">)</span>  <span class="s s-Atom">#</span> <span class="nv">PAF</span> <span class="o">is</span> <span class="s s-Atom">tab</span> <span class="s s-Atom">separated</span>

                    <span class="s s-Atom">row</span> <span class="o">=</span> <span class="s s-Atom">paf_line</span><span class="p">[</span><span class="mi">0</span><span class="s s-Atom">:</span><span class="mi">12</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="s2">&quot;;&quot;</span><span class="p">.</span><span class="nf">join</span><span class="p">(</span><span class="s s-Atom">paf_line</span><span class="p">[</span><span class="mi">12</span><span class="p">:-</span><span class="mi">1</span><span class="p">])]</span> <span class="o">+</span> <span class="p">[</span><span class="s s-Atom">paf_line</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]]</span>

                    <span class="s s-Atom">rows</span> <span class="s s-Atom">+=</span> <span class="p">[</span><span class="s s-Atom">row</span><span class="p">]</span>

            <span class="s s-Atom">paf</span> <span class="o">=</span> <span class="s s-Atom">pandas</span><span class="p">.</span><span class="nv">DataFrame</span><span class="p">(</span><span class="s s-Atom">rows</span><span class="p">)</span>

            <span class="s s-Atom">#</span> <span class="nv">Set</span> <span class="s s-Atom">numeric</span> <span class="nn">columns</span><span class="p">:</span>

            <span class="s s-Atom">numeric_columns</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">6</span><span class="p">,</span> <span class="mi">7</span><span class="p">,</span> <span class="mi">8</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">11</span><span class="p">]</span>  <span class="s s-Atom">#</span> <span class="s s-Atom">see</span> <span class="s s-Atom">list</span> <span class="s s-Atom">below</span>

            <span class="s s-Atom">paf</span><span class="p">[</span><span class="s s-Atom">numeric_columns</span><span class="p">]</span> <span class="o">=</span> <span class="s s-Atom">paf</span><span class="p">[</span><span class="s s-Atom">numeric_columns</span><span class="p">].</span><span class="nf">apply</span><span class="p">(</span><span class="s s-Atom">pandas</span><span class="p">.</span><span class="s s-Atom">to_numeric</span><span class="p">)</span>

        <span class="s s-Atom">#</span> <span class="nv">First</span> <span class="mi">12</span> <span class="s s-Atom">columns</span> <span class="s s-Atom">are</span> <span class="s s-Atom">defined</span> <span class="s s-Atom">by</span> <span class="s s-Atom">the</span> <span class="nn">format</span><span class="p">:</span>

        <span class="s s-Atom">columns_12</span> <span class="o">=</span> <span class="p">[</span>

            <span class="s2">&quot;query_name&quot;</span><span class="p">,</span>

            <span class="s2">&quot;query_length&quot;</span><span class="p">,</span>

            <span class="s2">&quot;query_start&quot;</span><span class="p">,</span>

            <span class="s2">&quot;query_end&quot;</span><span class="p">,</span>

            <span class="s2">&quot;strand&quot;</span><span class="p">,</span>

            <span class="s2">&quot;target_name&quot;</span><span class="p">,</span>

            <span class="s2">&quot;target_length&quot;</span><span class="p">,</span>

            <span class="s2">&quot;target_start&quot;</span><span class="p">,</span>

            <span class="s2">&quot;target_end&quot;</span><span class="p">,</span>

            <span class="s2">&quot;mapping_matches&quot;</span><span class="p">,</span>

            <span class="s2">&quot;mapping_size&quot;</span><span class="p">,</span>

            <span class="s2">&quot;mapping_quality&quot;</span><span class="p">,</span>

        <span class="p">]</span>

        <span class="s s-Atom">#</span> <span class="nv">For</span> <span class="s s-Atom">additional</span> <span class="s s-Atom">columns</span> <span class="s s-Atom">optionally</span> <span class="s s-Atom">created</span> <span class="s s-Atom">during</span> <span class="nn">alignment</span><span class="p">:</span>

        <span class="s s-Atom">columns_13plus</span> <span class="o">=</span> <span class="p">[</span><span class="nf">str</span><span class="p">(</span><span class="s s-Atom">index</span><span class="p">)</span> <span class="s s-Atom">for</span> <span class="s s-Atom">index</span> <span class="s s-Atom">in</span> <span class="nf">range</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="nf">len</span><span class="p">(</span><span class="s s-Atom">paf</span><span class="p">.</span><span class="s s-Atom">columns</span><span class="p">))]</span>

        <span class="s s-Atom">paf</span><span class="p">.</span><span class="s s-Atom">columns</span> <span class="o">=</span> <span class="s s-Atom">columns_12</span> <span class="o">+</span> <span class="s s-Atom">columns_13plus</span>

        <span class="s s-Atom">return</span> <span class="s s-Atom">paf</span>

    <span class="s s-Atom">@staticmethod</span>

    <span class="s s-Atom">def</span> <span class="nf">load_tsv</span><span class="p">(</span><span class="s s-Atom">tsv_path</span><span class="p">)</span><span class="s s-Atom">:</span>

        <span class="s2">&quot;&quot;&quot;Create a dataframe from a TSV file of depth counts.</span>

<span class="s2">        **Parameters**</span>

<span class="s2">        **tsv_path**</span>

<span class="s2">        &gt; Path to output of `samtools depth -aa` (`str`).</span>

<span class="s2">        &quot;&quot;&quot;</span>

        <span class="s s-Atom">tsv</span> <span class="o">=</span> <span class="s s-Atom">pandas</span><span class="p">.</span><span class="nf">read_csv</span><span class="p">(</span><span class="s s-Atom">tsv_path</span><span class="p">,</span> <span class="s s-Atom">sep=</span><span class="s2">&quot;\t&quot;</span><span class="p">,</span> <span class="s s-Atom">header</span><span class="o">=</span><span class="nv">None</span><span class="p">)</span>

        <span class="s s-Atom">tsv</span><span class="p">.</span><span class="s s-Atom">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">,</span> <span class="s2">&quot;position&quot;</span><span class="p">,</span> <span class="s2">&quot;depth&quot;</span><span class="p">]</span>

        <span class="s s-Atom">return</span> <span class="s s-Atom">tsv</span>
</code></pre></div>

</details>
<hr />
<h4 id="static-methods">Static methods</h4>
<h4 id="load_paf">load_paf</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">load_paf</span><span class="p">(</span>
    <span class="n">paf_path</span>
<span class="p">)</span>
</code></pre></div>

<p>Create a dataframe from a PAF file of alignments.</p>
<p><strong>Parameters</strong></p>
<p><strong>paf_path</strong></p>
<blockquote>
<p>Path (<code>str</code>) to PAF file: <code>paftools.js sam2paf aln.sam &gt; aln.paf</code>.</p>
</blockquote>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code>    <span class="s s-Atom">@staticmethod</span>

    <span class="s s-Atom">def</span> <span class="nf">load_paf</span><span class="p">(</span><span class="s s-Atom">paf_path</span><span class="p">)</span><span class="s s-Atom">:</span>

        <span class="s2">&quot;&quot;&quot;Create a dataframe from a PAF file of alignments.</span>

<span class="s2">        **Parameters**</span>

<span class="s2">        **paf_path**</span>

<span class="s2">        &gt; Path (`str`) to PAF file: `paftools.js sam2paf aln.sam &gt; aln.paf`.</span>

<span class="s2">        &quot;&quot;&quot;</span>

        <span class="nn">try</span><span class="p">:</span>

            <span class="s s-Atom">paf</span> <span class="o">=</span> <span class="s s-Atom">pandas</span><span class="p">.</span><span class="nf">read_csv</span><span class="p">(</span><span class="s s-Atom">paf_path</span><span class="p">,</span> <span class="s s-Atom">sep=</span><span class="s2">&quot;\t&quot;</span><span class="p">,</span> <span class="s s-Atom">header</span><span class="o">=</span><span class="nv">None</span><span class="p">)</span>

        <span class="s s-Atom">except</span> <span class="nv">Exception</span><span class="s s-Atom">:</span>  <span class="s s-Atom">#</span> <span class="s s-Atom">unequal</span> <span class="s s-Atom">number</span> <span class="s s-Atom">of</span> <span class="s s-Atom">columns</span><span class="p">,</span> <span class="s s-Atom">the</span> <span class="s s-Atom">first</span> <span class="mi">12</span> <span class="o">is</span> <span class="s s-Atom">assumed</span> <span class="s s-Atom">to</span> <span class="s s-Atom">be</span> <span class="nv">OK</span>

            <span class="s s-Atom">#</span> <span class="nv">The</span> <span class="s s-Atom">last</span> <span class="s s-Atom">column</span> <span class="o">is</span> <span class="nv">CIGAR</span><span class="p">,</span> <span class="s s-Atom">so</span> <span class="s s-Atom">we</span> <span class="s s-Atom">need</span> <span class="s s-Atom">to</span> <span class="s s-Atom">read</span> <span class="s s-Atom">lines</span> <span class="s s-Atom">one</span> <span class="s s-Atom">by</span> <span class="nn">one</span><span class="p">:</span>

            <span class="s s-Atom">rows</span> <span class="o">=</span> <span class="p">[]</span>

            <span class="s s-Atom">with</span> <span class="nf">open</span><span class="p">(</span><span class="s s-Atom">paf_path</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="s s-Atom">as</span> <span class="s s-Atom">f_input:</span>

                <span class="s s-Atom">paf_lines</span> <span class="o">=</span> <span class="s s-Atom">f_input</span><span class="p">.</span><span class="nf">read</span><span class="p">().</span><span class="nf">splitlines</span><span class="p">()</span>

                <span class="s s-Atom">for</span> <span class="s s-Atom">paf_line</span> <span class="s s-Atom">in</span> <span class="s s-Atom">paf_lines:</span>

                    <span class="s s-Atom">paf_line</span> <span class="o">=</span> <span class="s s-Atom">paf_line</span><span class="p">.</span><span class="nf">split</span><span class="p">(</span><span class="s2">&quot;\t&quot;</span><span class="p">)</span>  <span class="s s-Atom">#</span> <span class="nv">PAF</span> <span class="o">is</span> <span class="s s-Atom">tab</span> <span class="s s-Atom">separated</span>

                    <span class="s s-Atom">row</span> <span class="o">=</span> <span class="s s-Atom">paf_line</span><span class="p">[</span><span class="mi">0</span><span class="s s-Atom">:</span><span class="mi">12</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="s2">&quot;;&quot;</span><span class="p">.</span><span class="nf">join</span><span class="p">(</span><span class="s s-Atom">paf_line</span><span class="p">[</span><span class="mi">12</span><span class="p">:-</span><span class="mi">1</span><span class="p">])]</span> <span class="o">+</span> <span class="p">[</span><span class="s s-Atom">paf_line</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">]]</span>

                    <span class="s s-Atom">rows</span> <span class="s s-Atom">+=</span> <span class="p">[</span><span class="s s-Atom">row</span><span class="p">]</span>

            <span class="s s-Atom">paf</span> <span class="o">=</span> <span class="s s-Atom">pandas</span><span class="p">.</span><span class="nv">DataFrame</span><span class="p">(</span><span class="s s-Atom">rows</span><span class="p">)</span>

            <span class="s s-Atom">#</span> <span class="nv">Set</span> <span class="s s-Atom">numeric</span> <span class="nn">columns</span><span class="p">:</span>

            <span class="s s-Atom">numeric_columns</span> <span class="o">=</span> <span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">6</span><span class="p">,</span> <span class="mi">7</span><span class="p">,</span> <span class="mi">8</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">10</span><span class="p">,</span> <span class="mi">11</span><span class="p">]</span>  <span class="s s-Atom">#</span> <span class="s s-Atom">see</span> <span class="s s-Atom">list</span> <span class="s s-Atom">below</span>

            <span class="s s-Atom">paf</span><span class="p">[</span><span class="s s-Atom">numeric_columns</span><span class="p">]</span> <span class="o">=</span> <span class="s s-Atom">paf</span><span class="p">[</span><span class="s s-Atom">numeric_columns</span><span class="p">].</span><span class="nf">apply</span><span class="p">(</span><span class="s s-Atom">pandas</span><span class="p">.</span><span class="s s-Atom">to_numeric</span><span class="p">)</span>

        <span class="s s-Atom">#</span> <span class="nv">First</span> <span class="mi">12</span> <span class="s s-Atom">columns</span> <span class="s s-Atom">are</span> <span class="s s-Atom">defined</span> <span class="s s-Atom">by</span> <span class="s s-Atom">the</span> <span class="nn">format</span><span class="p">:</span>

        <span class="s s-Atom">columns_12</span> <span class="o">=</span> <span class="p">[</span>

            <span class="s2">&quot;query_name&quot;</span><span class="p">,</span>

            <span class="s2">&quot;query_length&quot;</span><span class="p">,</span>

            <span class="s2">&quot;query_start&quot;</span><span class="p">,</span>

            <span class="s2">&quot;query_end&quot;</span><span class="p">,</span>

            <span class="s2">&quot;strand&quot;</span><span class="p">,</span>

            <span class="s2">&quot;target_name&quot;</span><span class="p">,</span>

            <span class="s2">&quot;target_length&quot;</span><span class="p">,</span>

            <span class="s2">&quot;target_start&quot;</span><span class="p">,</span>

            <span class="s2">&quot;target_end&quot;</span><span class="p">,</span>

            <span class="s2">&quot;mapping_matches&quot;</span><span class="p">,</span>

            <span class="s2">&quot;mapping_size&quot;</span><span class="p">,</span>

            <span class="s2">&quot;mapping_quality&quot;</span><span class="p">,</span>

        <span class="p">]</span>

        <span class="s s-Atom">#</span> <span class="nv">For</span> <span class="s s-Atom">additional</span> <span class="s s-Atom">columns</span> <span class="s s-Atom">optionally</span> <span class="s s-Atom">created</span> <span class="s s-Atom">during</span> <span class="nn">alignment</span><span class="p">:</span>

        <span class="s s-Atom">columns_13plus</span> <span class="o">=</span> <span class="p">[</span><span class="nf">str</span><span class="p">(</span><span class="s s-Atom">index</span><span class="p">)</span> <span class="s s-Atom">for</span> <span class="s s-Atom">index</span> <span class="s s-Atom">in</span> <span class="nf">range</span><span class="p">(</span><span class="mi">12</span><span class="p">,</span> <span class="nf">len</span><span class="p">(</span><span class="s s-Atom">paf</span><span class="p">.</span><span class="s s-Atom">columns</span><span class="p">))]</span>

        <span class="s s-Atom">paf</span><span class="p">.</span><span class="s s-Atom">columns</span> <span class="o">=</span> <span class="s s-Atom">columns_12</span> <span class="o">+</span> <span class="s s-Atom">columns_13plus</span>

        <span class="s s-Atom">return</span> <span class="s s-Atom">paf</span>
</code></pre></div>

</details>
<h4 id="load_tsv">load_tsv</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">load_tsv</span><span class="p">(</span>
    <span class="n">tsv_path</span>
<span class="p">)</span>
</code></pre></div>

<p>Create a dataframe from a TSV file of depth counts.</p>
<p><strong>Parameters</strong></p>
<p><strong>tsv_path</strong></p>
<blockquote>
<p>Path to output of <code>samtools depth -aa</code> (<code>str</code>).</p>
</blockquote>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code>    <span class="err">@</span><span class="n">staticmethod</span>

    <span class="n">def</span> <span class="n">load_tsv</span><span class="p">(</span><span class="n">tsv_path</span><span class="p">):</span>

        <span class="sd">&quot;&quot;&quot;Create a dataframe from a TSV file of depth counts.</span>

<span class="sd">        **Parameters**</span>

<span class="sd">        **tsv_path**</span>

<span class="sd">        &gt; Path to output of `samtools depth -aa` (`str`).</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">tsv</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">tsv_path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="n">None</span><span class="p">)</span>

        <span class="n">tsv</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">,</span> <span class="s2">&quot;position&quot;</span><span class="p">,</span> <span class="s2">&quot;depth&quot;</span><span class="p">]</span>

        <span class="k">return</span> <span class="n">tsv</span>
</code></pre></div>

</details>
<h4 id="methods_1">Methods</h4>
<h4 id="add_comparator">add_comparator</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">add_comparator</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">reference_name</span>
<span class="p">)</span>
</code></pre></div>

<p>Create comparator and add to the group. See create_comparator().</p>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code>    def add_comparator(self, reference_name):

        &quot;&quot;&quot;Create comparator and add to the group. See create_comparator().&quot;&quot;&quot;

        self.comparators += [self.create_comparator(reference_name)]
</code></pre></div>

</details>
<h4 id="create_comparator">create_comparator</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">create_comparator</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">reference_name</span>
<span class="p">)</span>
</code></pre></div>

<p>Create a Comparator instance from one of the loaded references.</p>
<p><strong>Parameters</strong></p>
<p><strong>reference_name</strong></p>
<blockquote>
<p>The name of the reference (<code>str</code>).</p>
</blockquote>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code>    <span class="n">def</span> <span class="n">create_comparator</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">reference_name</span><span class="p">):</span>

        <span class="sd">&quot;&quot;&quot;Create a Comparator instance from one of the loaded references.</span>

<span class="sd">        **Parameters**</span>

<span class="sd">        **reference_name**</span>

<span class="sd">        &gt; The name of the reference (`str`).</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">subset_paf</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">subset_paf</span><span class="p">(</span><span class="n">reference_name</span><span class="p">)</span>

        <span class="n">subset_tsv</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">tsv</span><span class="p">[</span><span class="bp">self</span><span class="o">.</span><span class="n">tsv</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">reference_name</span><span class="p">]</span>

        <span class="n">alignment</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;paf&quot;</span><span class="p">:</span> <span class="n">subset_paf</span><span class="p">,</span> <span class="s2">&quot;tsv&quot;</span><span class="p">:</span> <span class="n">subset_tsv</span><span class="p">}</span>

        <span class="n">comparator</span> <span class="o">=</span> <span class="n">Comparator</span><span class="p">(</span>

            <span class="p">{</span><span class="n">reference_name</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">references</span><span class="p">[</span><span class="n">reference_name</span><span class="p">]},</span> <span class="n">alignment</span>

        <span class="p">)</span>

        <span class="k">return</span> <span class="n">comparator</span>
</code></pre></div>

</details>
<h4 id="perform_all_comparisons">perform_all_comparisons</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">perform_all_comparisons</span><span class="p">(</span>
    <span class="bp">self</span>
<span class="p">)</span>
</code></pre></div>

<p>Call perform_comparison() on each Comparator.</p>
<p><strong>Parameters</strong></p>
<p><strong>assembly_paths</strong></p>
<blockquote>
<p>Dictionary of construct name: consensus sequence file.
For example: <code>{"Construct_1": "/path/to/assembly.fa"}</code>.</p>
</blockquote>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code>    <span class="n">def</span> <span class="n">perform_all_comparisons</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>

        <span class="sd">&quot;&quot;&quot;Call perform_comparison() on each Comparator.</span>

<span class="sd">        **Parameters**</span>

<span class="sd">        **assembly_paths**</span>

<span class="sd">        &gt; Dictionary of construct name: consensus sequence file.</span>

<span class="sd">        For example: `{&quot;Construct_1&quot;: &quot;/path/to/assembly.fa&quot;}`.</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">assembly_paths</span> <span class="k">is</span> <span class="n">None</span><span class="p">:</span>

            <span class="n">assembly_paths</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="k">else</span><span class="p">:</span>

            <span class="n">assembly_paths</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">assembly_paths</span>

        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">vcf_paths</span> <span class="k">is</span> <span class="n">None</span><span class="p">:</span>

            <span class="n">vcf_paths</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="k">else</span><span class="p">:</span>

            <span class="n">vcf_paths</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">vcf_paths</span>

        <span class="k">for</span> <span class="n">comparator</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">comparators</span><span class="p">:</span>

            <span class="n">try</span><span class="p">:</span>

                <span class="n">assembly_path</span> <span class="o">=</span> <span class="n">assembly_paths</span><span class="p">[</span><span class="n">comparator</span><span class="o">.</span><span class="n">name</span><span class="p">]</span>

            <span class="n">except</span> <span class="n">Exception</span><span class="p">:</span>

                <span class="n">assembly_path</span> <span class="o">=</span> <span class="n">None</span>

            <span class="n">try</span><span class="p">:</span>

                <span class="n">vcf_path</span> <span class="o">=</span> <span class="n">vcf_paths</span><span class="p">[</span><span class="n">comparator</span><span class="o">.</span><span class="n">name</span><span class="p">]</span>

            <span class="n">except</span> <span class="n">Exception</span><span class="p">:</span>

                <span class="n">vcf_path</span> <span class="o">=</span> <span class="n">None</span>

            <span class="n">comparator</span><span class="o">.</span><span class="n">perform_comparison</span><span class="p">(</span>

                <span class="n">assembly_path</span><span class="o">=</span><span class="n">assembly_path</span><span class="p">,</span> <span class="n">vcf_path</span><span class="o">=</span><span class="n">vcf_path</span>

            <span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">comparisons_performed</span> <span class="o">=</span> <span class="n">True</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">summarise_analysis</span><span class="p">()</span>
</code></pre></div>

</details>
<h4 id="plot_fastq_histogram">plot_fastq_histogram</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">plot_fastq_histogram</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">n_bins</span><span class="o">=</span><span class="mi">50</span>
<span class="p">)</span>
</code></pre></div>

<p>Plot a histogram of the FASTQ reads.</p>
<p><strong>Parameters</strong></p>
<p><strong>n_bins</strong></p>
<blockquote>
<p>Number of bins in the histogram (<code>int</code>).</p>
</blockquote>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code>    <span class="n">def</span> <span class="n">plot_fastq_histogram</span><span class="p">(</span><span class="n">self</span><span class="p">,</span> <span class="n">n_bins</span><span class="o">=</span><span class="mi">50</span><span class="p">)</span><span class="o">:</span>

        <span class="s2">&quot;</span><span class="se">&quot;&quot;</span><span class="s2">Plot a histogram of the FASTQ reads.</span>

<span class="s2">        **Parameters**</span>

<span class="s2">        **n_bins**</span>

<span class="s2">        &gt; Number of bins in the histogram (`int`).</span>

<span class="s2">        </span><span class="se">&quot;&quot;</span><span class="s2">&quot;</span>

        <span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="p">.</span><span class="n">subplots</span><span class="p">()</span>

        <span class="n">plt</span><span class="p">.</span><span class="n">hist</span><span class="p">(</span>

            <span class="n">self</span><span class="p">.</span><span class="n">paf</span><span class="err">[</span><span class="s2">&quot;query_length&quot;</span><span class="err">]</span><span class="p">,</span>

            <span class="n">bins</span><span class="o">=</span><span class="kt">int</span><span class="p">(</span><span class="n">self</span><span class="p">.</span><span class="n">paf</span><span class="p">.</span><span class="n">iloc</span><span class="err">[</span><span class="mi">0</span><span class="err">][</span><span class="s2">&quot;target_length&quot;</span><span class="err">]</span> <span class="o">/</span> <span class="n">n_bins</span><span class="p">),</span>

            <span class="n">alpha</span><span class="o">=</span><span class="mf">0.8</span><span class="p">,</span>

        <span class="p">)</span>

        <span class="n">ax</span><span class="p">.</span><span class="k">set</span><span class="n">_ylabel</span><span class="p">(</span><span class="s2">&quot;Number of reads&quot;</span><span class="p">)</span>

        <span class="n">ax</span><span class="p">.</span><span class="k">set</span><span class="n">_xlabel</span><span class="p">(</span><span class="s2">&quot;Read length [bp]&quot;</span><span class="p">)</span>

        <span class="c1"># Mark expected lengths for interpreting the plot:</span>

        <span class="k">for</span> <span class="n">comparator</span> <span class="k">in</span> <span class="n">self</span><span class="p">.</span><span class="n">comparators</span><span class="o">:</span>

            <span class="n">plt</span><span class="p">.</span><span class="n">axvline</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">len</span><span class="p">(</span><span class="n">comparator</span><span class="p">.</span><span class="n">record</span><span class="p">),</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;#fd5a31&quot;</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">fig</span>
</code></pre></div>

</details>
<h4 id="subset_paf">subset_paf</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">subset_paf</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">reference_name</span>
<span class="p">)</span>
</code></pre></div>

<p>Subset PAF dataframe for the reference.</p>
<p>Select reads that align to the reference and also align <em>best</em> to the reference.</p>
<p><strong>Parameters</strong></p>
<p><strong>reference_name</strong></p>
<blockquote>
<p>Name of the reference sequence to filter for (<code>str</code>).</p>
</blockquote>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code>    <span class="n">def</span> <span class="n">subset_paf</span><span class="p">(</span><span class="n">self</span><span class="p">,</span> <span class="n">reference_name</span><span class="p">)</span><span class="o">:</span>

        <span class="s2">&quot;</span><span class="se">&quot;&quot;</span><span class="s2">Subset PAF dataframe for the reference.</span>

<span class="s2">        Select reads that align to the reference and also align *best* to the reference.</span>

<span class="s2">        **Parameters**</span>

<span class="s2">        **reference_name**</span>

<span class="s2">        &gt; Name of the reference sequence to filter for (`str`).</span>

<span class="s2">        </span><span class="se">&quot;&quot;</span><span class="s2">&quot;</span>

        <span class="c1"># Work only with the values used for filtering:</span>

        <span class="n">paf3</span> <span class="o">=</span> <span class="n">self</span><span class="p">.</span><span class="n">paf</span><span class="err">[[</span><span class="s2">&quot;query_name&quot;</span><span class="p">,</span> <span class="s2">&quot;target_name&quot;</span><span class="p">,</span> <span class="s2">&quot;mapping_matches&quot;</span><span class="err">]]</span>

        <span class="n">grouped</span> <span class="o">=</span> <span class="n">paf3</span><span class="p">.</span><span class="n">groupby</span><span class="p">(</span><span class="s2">&quot;query_name&quot;</span><span class="p">)</span>

        <span class="n">selected_reads</span> <span class="o">=</span> <span class="err">[]</span>

        <span class="k">for</span> <span class="k">name</span><span class="p">,</span> <span class="k">group</span> <span class="k">in</span> <span class="n">grouped</span><span class="o">:</span>

            <span class="c1"># not all reads align to the reference:</span>

            <span class="k">if</span> <span class="n">reference_name</span> <span class="k">in</span> <span class="k">group</span><span class="err">[</span><span class="s2">&quot;target_name&quot;</span><span class="err">]</span><span class="p">.</span><span class="k">values</span><span class="o">:</span>

                <span class="n">max_matches</span> <span class="o">=</span> <span class="k">group</span><span class="err">[</span><span class="s2">&quot;mapping_matches&quot;</span><span class="err">]</span><span class="p">.</span><span class="nf">max</span><span class="p">()</span>

                <span class="c1"># filter that the read aligns best to this reference:</span>

                <span class="k">if</span> <span class="p">(</span>

                    <span class="nf">max</span><span class="p">(</span>

                        <span class="k">group</span><span class="err">[</span><span class="k">group</span><span class="err">[</span><span class="s2">&quot;target_name&quot;</span><span class="err">]</span> <span class="o">==</span> <span class="n">reference_name</span><span class="err">][</span>

                            <span class="s2">&quot;mapping_matches&quot;</span>

                        <span class="err">]</span><span class="p">.</span><span class="k">values</span>

                    <span class="p">)</span>

                    <span class="o">==</span> <span class="n">max_matches</span>

                <span class="p">)</span><span class="o">:</span>

                    <span class="n">selected_reads</span> <span class="o">+=</span> <span class="err">[</span><span class="k">name</span><span class="err">]</span>

        <span class="c1"># Filter for selected reads ..</span>

        <span class="n">subset_paf_filtered</span> <span class="o">=</span> <span class="n">self</span><span class="p">.</span><span class="n">paf</span><span class="err">[</span><span class="n">self</span><span class="p">.</span><span class="n">paf</span><span class="err">[</span><span class="s2">&quot;query_name&quot;</span><span class="err">]</span><span class="p">.</span><span class="n">isin</span><span class="p">(</span><span class="n">selected_reads</span><span class="p">)</span><span class="err">]</span>

        <span class="c1"># ..then get only the reads aligning to reference:</span>

        <span class="n">subset_paf_final</span> <span class="o">=</span> <span class="n">subset_paf_filtered</span><span class="err">[</span>

            <span class="n">subset_paf_filtered</span><span class="err">[</span><span class="s2">&quot;target_name&quot;</span><span class="err">]</span> <span class="o">==</span> <span class="n">reference_name</span>

        <span class="err">]</span>

        <span class="k">return</span> <span class="n">subset_paf_final</span>
</code></pre></div>

</details>
<h4 id="summarise_analysis">summarise_analysis</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">summarise_analysis</span><span class="p">(</span>
    <span class="bp">self</span>
<span class="p">)</span>
</code></pre></div>

<p>Create variables for the PDF report summary.</p>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code><span class="w">    </span><span class="n">def</span><span class="w"> </span><span class="n">summarise_analysis</span><span class="p">(</span><span class="n">self</span><span class="p">)</span><span class="err">:</span><span class="w"></span>

<span class="w">        </span><span class="ss">&quot;&quot;&quot;Create variables for the PDF report summary.&quot;&quot;&quot;</span><span class="w"></span>

<span class="w">        </span><span class="n">self</span><span class="p">.</span><span class="n">number_of_constructs</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">len</span><span class="p">(</span><span class="n">self</span><span class="p">.</span><span class="n">comparators</span><span class="p">)</span><span class="w"></span>

<span class="w">        </span><span class="n">self</span><span class="p">.</span><span class="n">result_good</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mi">0</span><span class="w"></span>

<span class="w">        </span><span class="n">self</span><span class="p">.</span><span class="n">result_warning</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mi">0</span><span class="w"></span>

<span class="w">        </span><span class="n">self</span><span class="p">.</span><span class="n">result_error</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mi">0</span><span class="w"></span>

<span class="w">        </span><span class="n">self</span><span class="p">.</span><span class="n">result_uncertain</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="mi">0</span><span class="w"></span>

<span class="w">        </span><span class="k">names</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="err">[]</span><span class="w"></span>

<span class="w">        </span><span class="n">reference_lengths</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="err">[]</span><span class="w"></span>

<span class="w">        </span><span class="n">results</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="err">[]</span><span class="w"></span>

<span class="w">        </span><span class="n">number_of_reads_aligning</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="err">[]</span><span class="w"></span>

<span class="w">        </span><span class="n">median_coverages</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="err">[]</span><span class="w"></span>

<span class="w">        </span><span class="k">for</span><span class="w"> </span><span class="n">comparator</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">self</span><span class="p">.</span><span class="nl">comparators</span><span class="p">:</span><span class="w"></span>

<span class="w">            </span><span class="k">names</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="o">[</span><span class="n">comparator.name</span><span class="o">]</span><span class="w"></span>

<span class="w">            </span><span class="n">reference_lengths</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="o">[</span><span class="n">str(comparator.reference_length)</span><span class="o">]</span><span class="w"></span>

<span class="w">            </span><span class="n">number_of_reads_aligning</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="o">[</span><span class="n"></span>

<span class="n">                str(len(comparator.paf[&quot;query_name&quot;</span><span class="o">]</span><span class="p">.</span><span class="k">unique</span><span class="p">()))</span><span class="w"></span>

<span class="w">            </span><span class="err">]</span><span class="w"></span>

<span class="w">            </span><span class="n">median_coverages</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="o">[</span><span class="n">str(int(comparator.median_yy))</span><span class="o">]</span><span class="w"></span>

<span class="w">            </span><span class="k">if</span><span class="w"> </span><span class="n">comparator</span><span class="p">.</span><span class="nl">is_uncertain</span><span class="p">:</span><span class="w"></span>

<span class="w">                </span><span class="n">self</span><span class="p">.</span><span class="n">result_uncertain</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="mi">1</span><span class="w"></span>

<span class="w">                </span><span class="n">comparator</span><span class="p">.</span><span class="k">result</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">result_keywords</span><span class="o">[</span><span class="n">&quot;uncertain&quot;</span><span class="o">]</span><span class="w"></span>

<span class="w">            </span><span class="k">else</span><span class="err">:</span><span class="w"></span>

<span class="w">                </span><span class="k">if</span><span class="w"> </span><span class="n">comparator</span><span class="p">.</span><span class="nl">has_errors</span><span class="p">:</span><span class="w"></span>

<span class="w">                    </span><span class="n">self</span><span class="p">.</span><span class="n">result_error</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="mi">1</span><span class="w"></span>

<span class="w">                    </span><span class="n">comparator</span><span class="p">.</span><span class="k">result</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">result_keywords</span><span class="o">[</span><span class="n">&quot;error&quot;</span><span class="o">]</span><span class="w"></span>

<span class="w">                </span><span class="k">else</span><span class="err">:</span><span class="w"></span>

<span class="w">                    </span><span class="k">if</span><span class="w"> </span><span class="n">comparator</span><span class="p">.</span><span class="nl">has_warnings</span><span class="p">:</span><span class="w"></span>

<span class="w">                        </span><span class="n">self</span><span class="p">.</span><span class="n">result_warning</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="mi">1</span><span class="w"></span>

<span class="w">                        </span><span class="n">comparator</span><span class="p">.</span><span class="k">result</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">result_keywords</span><span class="o">[</span><span class="n">&quot;warning&quot;</span><span class="o">]</span><span class="w"></span>

<span class="w">                    </span><span class="k">else</span><span class="err">:</span><span class="w"></span>

<span class="w">                        </span><span class="n">self</span><span class="p">.</span><span class="n">result_good</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="mi">1</span><span class="w"></span>

<span class="w">                        </span><span class="n">comparator</span><span class="p">.</span><span class="k">result</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">result_keywords</span><span class="o">[</span><span class="n">&quot;good&quot;</span><span class="o">]</span><span class="w"></span>

<span class="w">            </span><span class="n">results</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="o">[</span><span class="n">comparator.result</span><span class="o">]</span><span class="w"></span>

<span class="w">        </span><span class="n">self</span><span class="p">.</span><span class="n">n_fastq_reads</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="nf">len</span><span class="p">(</span><span class="k">set</span><span class="p">(</span><span class="n">self</span><span class="p">.</span><span class="n">paf</span><span class="p">.</span><span class="n">query_name</span><span class="p">))</span><span class="w"></span>

<span class="w">        </span><span class="n">self</span><span class="p">.</span><span class="n">fastq_plot</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">self</span><span class="p">.</span><span class="n">plot_fastq_histogram</span><span class="p">()</span><span class="w"></span>

<span class="w">        </span><span class="err">#</span><span class="w"> </span><span class="n">Summary</span><span class="w"> </span><span class="nc">table</span><span class="w"></span>

<span class="w">        </span><span class="n">d</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="err">{</span><span class="w"></span>

<span class="w">            </span><span class="ss">&quot;Name&quot;</span><span class="err">:</span><span class="w"> </span><span class="n">pandas</span><span class="p">.</span><span class="n">Series</span><span class="p">(</span><span class="k">names</span><span class="p">),</span><span class="w"></span>

<span class="w">            </span><span class="ss">&quot;Result&quot;</span><span class="err">:</span><span class="w"> </span><span class="n">pandas</span><span class="p">.</span><span class="n">Series</span><span class="p">(</span><span class="n">results</span><span class="p">),</span><span class="w"></span>

<span class="w">            </span><span class="ss">&quot;Length [bp]&quot;</span><span class="err">:</span><span class="w"> </span><span class="n">pandas</span><span class="p">.</span><span class="n">Series</span><span class="p">(</span><span class="n">reference_lengths</span><span class="p">),</span><span class="w"></span>

<span class="w">            </span><span class="ss">&quot;FASTQ reads&quot;</span><span class="err">:</span><span class="w"> </span><span class="n">pandas</span><span class="p">.</span><span class="n">Series</span><span class="p">(</span><span class="n">number_of_reads_aligning</span><span class="p">),</span><span class="w"></span>

<span class="w">            </span><span class="ss">&quot;Coverage [x]&quot;</span><span class="err">:</span><span class="w"> </span><span class="n">pandas</span><span class="p">.</span><span class="n">Series</span><span class="p">(</span><span class="n">median_coverages</span><span class="p">),</span><span class="w"></span>

<span class="w">        </span><span class="err">}</span><span class="w"></span>

<span class="w">        </span><span class="n">self</span><span class="p">.</span><span class="n">summary_table</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">pandas</span><span class="p">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">d</span><span class="p">)</span><span class="w"></span>
</code></pre></div>

</details>
<h3 id="customtranslator">CustomTranslator</h3>
<div class="codehilite"><pre><span></span><code><span class="k">class</span> <span class="nc">CustomTranslator</span><span class="p">(</span>
    <span class="n">features_filters</span><span class="o">=</span><span class="p">(),</span>
    <span class="n">features_properties</span><span class="o">=</span><span class="kc">None</span>
<span class="p">)</span>
</code></pre></div>

<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code><span class="k">class</span><span class="w"> </span><span class="n">CustomTranslator</span><span class="p">(</span><span class="n">dna_features_viewer</span><span class="p">.</span><span class="n">BiopythonTranslator</span><span class="p">)</span><span class="err">:</span><span class="w"></span>

<span class="w">    </span><span class="ss">&quot;&quot;&quot;Custom translator.&quot;&quot;&quot;</span><span class="w"></span>

<span class="w">    </span><span class="n">def</span><span class="w"> </span><span class="n">compute_filtered_features</span><span class="p">(</span><span class="n">self</span><span class="p">,</span><span class="w"> </span><span class="n">features</span><span class="p">)</span><span class="err">:</span><span class="w"></span>

<span class="w">        </span><span class="ss">&quot;&quot;&quot;Display only &quot;</span><span class="k">From</span><span class="w"> </span><span class="ss">&quot; features and overhangs.&quot;&quot;&quot;</span><span class="w"></span>

<span class="w">        </span><span class="n">filtered_features</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="err">[]</span><span class="w"></span>

<span class="w">        </span><span class="k">for</span><span class="w"> </span><span class="n">feature</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="nl">features</span><span class="p">:</span><span class="w"></span>

<span class="w">            </span><span class="k">try</span><span class="err">:</span><span class="w">  </span><span class="err">#</span><span class="w"> </span><span class="n">may</span><span class="w"> </span><span class="ow">not</span><span class="w"> </span><span class="n">have</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="s1">&#39;label&#39;</span><span class="p">,</span><span class="w"> </span><span class="nf">count</span><span class="w"> </span><span class="mi">4</span><span class="w"> </span><span class="k">to</span><span class="w"> </span><span class="k">include</span><span class="w"> </span><span class="n">overhang</span><span class="w"> </span><span class="n">annotations</span><span class="w"></span>

<span class="w">                </span><span class="k">if</span><span class="w"> </span><span class="p">(</span><span class="nf">len</span><span class="p">(</span><span class="n">feature</span><span class="p">.</span><span class="n">qualifiers</span><span class="p">.</span><span class="k">get</span><span class="p">(</span><span class="ss">&quot;label&quot;</span><span class="p">,</span><span class="w"> </span><span class="ss">&quot;&quot;</span><span class="p">)</span><span class="o">[</span><span class="n">0</span><span class="o">]</span><span class="p">)</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="mi">4</span><span class="p">)</span><span class="w"> </span><span class="ow">or</span><span class="w"> </span><span class="p">(</span><span class="w"></span>

<span class="w">                    </span><span class="ss">&quot;From &quot;</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="nf">str</span><span class="p">(</span><span class="n">feature</span><span class="p">.</span><span class="n">qualifiers</span><span class="p">.</span><span class="k">get</span><span class="p">(</span><span class="ss">&quot;label&quot;</span><span class="p">,</span><span class="w"> </span><span class="ss">&quot;&quot;</span><span class="p">))</span><span class="w"></span>

<span class="w">                </span><span class="p">)</span><span class="err">:</span><span class="w"></span>

<span class="w">                    </span><span class="n">filtered_features</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="o">[</span><span class="n">feature</span><span class="o">]</span><span class="w"></span>

<span class="w">            </span><span class="ow">except</span><span class="err">:</span><span class="w"></span>

<span class="w">                </span><span class="n">pass</span><span class="w"></span>

<span class="w">        </span><span class="k">return</span><span class="w"> </span><span class="n">filtered_features</span><span class="w"></span>
</code></pre></div>

</details>
<hr />
<h4 id="ancestors-in-mro">Ancestors (in MRO)</h4>
<ul>
<li>dna_features_viewer.BiopythonTranslator.BiopythonTranslator.BiopythonTranslator</li>
<li>dna_features_viewer.BiopythonTranslator.BiopythonTranslatorBase.BiopythonTranslatorBase</li>
</ul>
<h4 id="class-variables">Class variables</h4>
<div class="codehilite"><pre><span></span><code><span class="n">default_feature_color</span>
</code></pre></div>

<div class="codehilite"><pre><span></span><code><span class="n">graphic_record_parameters</span>
</code></pre></div>

<div class="codehilite"><pre><span></span><code><span class="n">ignored_features_types</span>
</code></pre></div>

<div class="codehilite"><pre><span></span><code><span class="n">label_fields</span>
</code></pre></div>

<h4 id="static-methods_1">Static methods</h4>
<h4 id="quick_class_plot">quick_class_plot</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">quick_class_plot</span><span class="p">(</span>
    <span class="n">record</span><span class="p">,</span>
    <span class="n">figure_width</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span>
    <span class="o">**</span><span class="n">kwargs</span>
<span class="p">)</span>
</code></pre></div>

<p>Allows super quick and dirty plotting of Biopython records.</p>
<p>This is really meant for use in a Jupyter/Ipython notebook with
the "%matplotlib inline" setting.</p>
<blockquote>
<blockquote>
<blockquote>
<p>from dna_features_viewer import BiopythonTranslator
BiopythonTranslator.quick_plot(my_record)</p>
</blockquote>
</blockquote>
</blockquote>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code>    <span class="nd">@classmethod</span>

    <span class="k">def</span> <span class="nf">quick_class_plot</span><span class="p">(</span><span class="bp">cls</span><span class="p">,</span> <span class="n">record</span><span class="p">,</span> <span class="n">figure_width</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>

        <span class="sd">&quot;&quot;&quot;Allows super quick and dirty plotting of Biopython records.</span>

<span class="sd">        This is really meant for use in a Jupyter/Ipython notebook with</span>

<span class="sd">        the &quot;%matplotlib inline&quot; setting.</span>

<span class="sd">        &gt;&gt;&gt; from dna_features_viewer import BiopythonTranslator</span>

<span class="sd">        &gt;&gt;&gt; BiopythonTranslator.quick_plot(my_record)</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">graphic_record</span> <span class="o">=</span> <span class="bp">cls</span><span class="p">()</span><span class="o">.</span><span class="n">translate_record</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>

        <span class="n">ax</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">graphic_record</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">figure_width</span><span class="o">=</span><span class="n">figure_width</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">ax</span>
</code></pre></div>

</details>
<h4 id="methods_2">Methods</h4>
<h4 id="compute_feature_box_color">compute_feature_box_color</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">compute_feature_box_color</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">feature</span>
<span class="p">)</span>
</code></pre></div>

<p>Compute a box_color for this feature.</p>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code>    <span class="nv">def</span> <span class="nv">compute_feature_box_color</span><span class="ss">(</span><span class="nv">self</span>, <span class="nv">feature</span><span class="ss">)</span>:

        <span class="s2">&quot;&quot;&quot;</span><span class="s">Compute a box_color for this feature.</span><span class="s2">&quot;&quot;&quot;</span>

        <span class="k">return</span> <span class="s2">&quot;</span><span class="s">auto</span><span class="s2">&quot;</span>
</code></pre></div>

</details>
<h4 id="compute_feature_box_linewidth">compute_feature_box_linewidth</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">compute_feature_box_linewidth</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">feature</span>
<span class="p">)</span>
</code></pre></div>

<p>Compute a box_linewidth for this feature.</p>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code>    <span class="nv">def</span> <span class="nv">compute_feature_box_linewidth</span><span class="ss">(</span><span class="nv">self</span>, <span class="nv">feature</span><span class="ss">)</span>:

        <span class="s2">&quot;&quot;&quot;</span><span class="s">Compute a box_linewidth for this feature.</span><span class="s2">&quot;&quot;&quot;</span>

        <span class="k">return</span> <span class="mi">0</span>.<span class="mi">3</span>
</code></pre></div>

</details>
<h4 id="compute_feature_color">compute_feature_color</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">compute_feature_color</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">feature</span>
<span class="p">)</span>
</code></pre></div>

<p>Compute a color for this feature.</p>
<p>If the feature has a <code>color</code> qualifier it will be used. Otherwise,
the classe's <code>default_feature_color</code> is used.</p>
<p>To change the behaviour, create a subclass of <code>BiopythonTranslator</code>
and overwrite this method.</p>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code>    <span class="n">def</span> <span class="n">compute_feature_color</span><span class="p">(</span><span class="n">self</span><span class="p">,</span> <span class="n">feature</span><span class="p">)</span><span class="o">:</span>

        <span class="s2">&quot;</span><span class="se">&quot;&quot;</span><span class="s2">Compute a color for this feature.</span>

<span class="s2">        If the feature has a ``color`` qualifier it will be used. Otherwise,</span>

<span class="s2">        the classe&#39;s ``default_feature_color`` is used.</span>

<span class="s2">        To change the behaviour, create a subclass of ``BiopythonTranslator``</span>

<span class="s2">        and overwrite this method.</span>

<span class="s2">        </span><span class="se">&quot;&quot;</span><span class="s2">&quot;</span>

        <span class="k">if</span> <span class="s2">&quot;color&quot;</span> <span class="k">in</span> <span class="n">feature</span><span class="p">.</span><span class="n">qualifiers</span><span class="o">:</span>

            <span class="n">color</span> <span class="o">=</span> <span class="n">feature</span><span class="p">.</span><span class="n">qualifiers</span><span class="err">[</span><span class="s2">&quot;color&quot;</span><span class="err">]</span>

            <span class="k">if</span> <span class="n">isinstance</span><span class="p">(</span><span class="n">color</span><span class="err">[</span><span class="mi">0</span><span class="err">]</span><span class="p">,</span> <span class="n">str</span><span class="p">)</span><span class="o">:</span>

                <span class="k">return</span> <span class="s2">&quot;&quot;</span><span class="p">.</span><span class="k">join</span><span class="p">(</span><span class="n">feature</span><span class="p">.</span><span class="n">qualifiers</span><span class="err">[</span><span class="s2">&quot;color&quot;</span><span class="err">]</span><span class="p">)</span>

            <span class="k">else</span><span class="o">:</span>

                <span class="k">return</span> <span class="n">color</span>

        <span class="k">else</span><span class="o">:</span>

            <span class="k">return</span> <span class="n">self</span><span class="p">.</span><span class="n">default_feature_color</span>
</code></pre></div>

</details>
<h4 id="compute_feature_fontdict">compute_feature_fontdict</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">compute_feature_fontdict</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">feature</span>
<span class="p">)</span>
</code></pre></div>

<p>Compute a font dict for this feature.</p>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code>    <span class="nv">def</span> <span class="nv">compute_feature_fontdict</span><span class="ss">(</span><span class="nv">self</span>, <span class="nv">feature</span><span class="ss">)</span>:

        <span class="s2">&quot;&quot;&quot;</span><span class="s">Compute a font dict for this feature.</span><span class="s2">&quot;&quot;&quot;</span>

        <span class="k">return</span> <span class="nv">None</span>
</code></pre></div>

</details>
<h4 id="compute_feature_html">compute_feature_html</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">compute_feature_html</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">feature</span>
<span class="p">)</span>
</code></pre></div>

<p>Gets the 'label' of the feature.</p>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code>    <span class="nv">def</span> <span class="nv">compute_feature_html</span><span class="ss">(</span><span class="nv">self</span>, <span class="nv">feature</span><span class="ss">)</span>:

        <span class="s2">&quot;&quot;&quot;</span><span class="s">Gets the &#39;label&#39; of the feature.</span><span class="s2">&quot;&quot;&quot;</span>

        <span class="k">return</span> <span class="nv">self</span>.<span class="nv">compute_feature_label</span><span class="ss">(</span><span class="nv">feature</span><span class="ss">)</span>
</code></pre></div>

</details>
<h4 id="compute_feature_label">compute_feature_label</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">compute_feature_label</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">feature</span>
<span class="p">)</span>
</code></pre></div>

<p>Compute the label of the feature.</p>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code><span class="w">    </span><span class="n">def</span><span class="w"> </span><span class="n">compute_feature_label</span><span class="p">(</span><span class="n">self</span><span class="p">,</span><span class="w"> </span><span class="n">feature</span><span class="p">)</span><span class="err">:</span><span class="w"></span>

<span class="w">        </span><span class="ss">&quot;&quot;&quot;Compute the label of the feature.&quot;&quot;&quot;</span><span class="w"></span>

<span class="w">        </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">feature</span><span class="p">.</span><span class="n">type</span><span class="w"></span>

<span class="w">        </span><span class="k">for</span><span class="w"> </span><span class="k">key</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">self</span><span class="p">.</span><span class="nl">label_fields</span><span class="p">:</span><span class="w"></span>

<span class="w">            </span><span class="k">if</span><span class="w"> </span><span class="k">key</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="n">feature</span><span class="p">.</span><span class="n">qualifiers</span><span class="w"> </span><span class="ow">and</span><span class="w"> </span><span class="nf">len</span><span class="p">(</span><span class="n">feature</span><span class="p">.</span><span class="n">qualifiers</span><span class="o">[</span><span class="n">key</span><span class="o">]</span><span class="p">)</span><span class="err">:</span><span class="w"></span>

<span class="w">                </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">feature</span><span class="p">.</span><span class="n">qualifiers</span><span class="o">[</span><span class="n">key</span><span class="o">]</span><span class="w"></span>

<span class="w">                </span><span class="k">break</span><span class="w"></span>

<span class="w">        </span><span class="k">if</span><span class="w"> </span><span class="n">isinstance</span><span class="p">(</span><span class="n">label</span><span class="p">,</span><span class="w"> </span><span class="n">list</span><span class="p">)</span><span class="err">:</span><span class="w"></span>

<span class="w">            </span><span class="n">label</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="ss">&quot;|&quot;</span><span class="p">.</span><span class="k">join</span><span class="p">(</span><span class="n">label</span><span class="p">)</span><span class="w"></span>

<span class="w">        </span><span class="k">return</span><span class="w"> </span><span class="n">label</span><span class="w"></span>
</code></pre></div>

</details>
<h4 id="compute_feature_label_link_color">compute_feature_label_link_color</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">compute_feature_label_link_color</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">feature</span>
<span class="p">)</span>
</code></pre></div>

<p>Compute the color of the line linking the label to its feature.</p>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code>    <span class="nv">def</span> <span class="nv">compute_feature_label_link_color</span><span class="ss">(</span><span class="nv">self</span>, <span class="nv">feature</span><span class="ss">)</span>:

        <span class="s2">&quot;&quot;&quot;</span><span class="s">Compute the color of the line linking the label to its feature.</span><span class="s2">&quot;&quot;&quot;</span>

        <span class="k">return</span> <span class="s2">&quot;</span><span class="s">black</span><span class="s2">&quot;</span>
</code></pre></div>

</details>
<h4 id="compute_feature_legend_text">compute_feature_legend_text</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">compute_feature_legend_text</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">feature</span>
<span class="p">)</span>
</code></pre></div>

<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code>    <span class="nv">def</span> <span class="nv">compute_feature_legend_text</span><span class="ss">(</span><span class="nv">self</span>, <span class="nv">feature</span><span class="ss">)</span>:

        <span class="k">return</span> <span class="nv">None</span>
</code></pre></div>

</details>
<h4 id="compute_feature_linewidth">compute_feature_linewidth</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">compute_feature_linewidth</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">feature</span>
<span class="p">)</span>
</code></pre></div>

<p>Compute the edge width of the feature's arrow/rectangle.</p>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code>    <span class="nv">def</span> <span class="nv">compute_feature_linewidth</span><span class="ss">(</span><span class="nv">self</span>, <span class="nv">feature</span><span class="ss">)</span>:

        <span class="s2">&quot;&quot;&quot;</span><span class="s">Compute the edge width of the feature&#39;s arrow/rectangle.</span><span class="s2">&quot;&quot;&quot;</span>

        <span class="k">return</span> <span class="mi">1</span>.<span class="mi">0</span>
</code></pre></div>

</details>
<h4 id="compute_filtered_features">compute_filtered_features</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">compute_filtered_features</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">features</span>
<span class="p">)</span>
</code></pre></div>

<p>Display only "From " features and overhangs.</p>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code><span class="w">    </span><span class="n">def</span><span class="w"> </span><span class="n">compute_filtered_features</span><span class="p">(</span><span class="n">self</span><span class="p">,</span><span class="w"> </span><span class="n">features</span><span class="p">)</span><span class="err">:</span><span class="w"></span>

<span class="w">        </span><span class="ss">&quot;&quot;&quot;Display only &quot;</span><span class="k">From</span><span class="w"> </span><span class="ss">&quot; features and overhangs.&quot;&quot;&quot;</span><span class="w"></span>

<span class="w">        </span><span class="n">filtered_features</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="err">[]</span><span class="w"></span>

<span class="w">        </span><span class="k">for</span><span class="w"> </span><span class="n">feature</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="nl">features</span><span class="p">:</span><span class="w"></span>

<span class="w">            </span><span class="k">try</span><span class="err">:</span><span class="w">  </span><span class="err">#</span><span class="w"> </span><span class="n">may</span><span class="w"> </span><span class="ow">not</span><span class="w"> </span><span class="n">have</span><span class="w"> </span><span class="n">a</span><span class="w"> </span><span class="s1">&#39;label&#39;</span><span class="p">,</span><span class="w"> </span><span class="nf">count</span><span class="w"> </span><span class="mi">4</span><span class="w"> </span><span class="k">to</span><span class="w"> </span><span class="k">include</span><span class="w"> </span><span class="n">overhang</span><span class="w"> </span><span class="n">annotations</span><span class="w"></span>

<span class="w">                </span><span class="k">if</span><span class="w"> </span><span class="p">(</span><span class="nf">len</span><span class="p">(</span><span class="n">feature</span><span class="p">.</span><span class="n">qualifiers</span><span class="p">.</span><span class="k">get</span><span class="p">(</span><span class="ss">&quot;label&quot;</span><span class="p">,</span><span class="w"> </span><span class="ss">&quot;&quot;</span><span class="p">)</span><span class="o">[</span><span class="n">0</span><span class="o">]</span><span class="p">)</span><span class="w"> </span><span class="o">==</span><span class="w"> </span><span class="mi">4</span><span class="p">)</span><span class="w"> </span><span class="ow">or</span><span class="w"> </span><span class="p">(</span><span class="w"></span>

<span class="w">                    </span><span class="ss">&quot;From &quot;</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="nf">str</span><span class="p">(</span><span class="n">feature</span><span class="p">.</span><span class="n">qualifiers</span><span class="p">.</span><span class="k">get</span><span class="p">(</span><span class="ss">&quot;label&quot;</span><span class="p">,</span><span class="w"> </span><span class="ss">&quot;&quot;</span><span class="p">))</span><span class="w"></span>

<span class="w">                </span><span class="p">)</span><span class="err">:</span><span class="w"></span>

<span class="w">                    </span><span class="n">filtered_features</span><span class="w"> </span><span class="o">+=</span><span class="w"> </span><span class="o">[</span><span class="n">feature</span><span class="o">]</span><span class="w"></span>

<span class="w">            </span><span class="ow">except</span><span class="err">:</span><span class="w"></span>

<span class="w">                </span><span class="n">pass</span><span class="w"></span>

<span class="w">        </span><span class="k">return</span><span class="w"> </span><span class="n">filtered_features</span><span class="w"></span>
</code></pre></div>

</details>
<h4 id="quick_plot">quick_plot</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">quick_plot</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">record</span><span class="p">,</span>
    <span class="n">figure_width</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span>
    <span class="o">**</span><span class="n">kwargs</span>
<span class="p">)</span>
</code></pre></div>

<p>Allows super quick and dirty plotting of Biopython records.</p>
<p>This is really meant for use in a Jupyter/Ipython notebook with
the "%matplotlib inline" setting.</p>
<blockquote>
<blockquote>
<blockquote>
<p>from dna_features_viewer import BiopythonTranslator
BiopythonTranslator.quick_plot(my_record)</p>
</blockquote>
</blockquote>
</blockquote>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code>    <span class="k">def</span> <span class="nf">quick_plot</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">record</span><span class="p">,</span> <span class="n">figure_width</span><span class="o">=</span><span class="mi">12</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>

        <span class="sd">&quot;&quot;&quot;Allows super quick and dirty plotting of Biopython records.</span>

<span class="sd">        This is really meant for use in a Jupyter/Ipython notebook with</span>

<span class="sd">        the &quot;%matplotlib inline&quot; setting.</span>

<span class="sd">        &gt;&gt;&gt; from dna_features_viewer import BiopythonTranslator</span>

<span class="sd">        &gt;&gt;&gt; BiopythonTranslator.quick_plot(my_record)</span>

<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">graphic_record</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">translate_record</span><span class="p">(</span><span class="n">record</span><span class="p">)</span>

        <span class="n">ax</span><span class="p">,</span> <span class="n">_</span> <span class="o">=</span> <span class="n">graphic_record</span><span class="o">.</span><span class="n">plot</span><span class="p">(</span><span class="n">figure_width</span><span class="o">=</span><span class="n">figure_width</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">ax</span>
</code></pre></div>

</details>
<h4 id="translate_feature">translate_feature</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">translate_feature</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">feature</span>
<span class="p">)</span>
</code></pre></div>

<p>Translate a Biopython feature into a Dna Features Viewer feature.</p>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code>    <span class="nv">def</span> <span class="nv">translate_feature</span><span class="ss">(</span><span class="nv">self</span>, <span class="nv">feature</span><span class="ss">)</span>:

        <span class="s2">&quot;&quot;&quot;</span><span class="s">Translate a Biopython feature into a Dna Features Viewer feature.</span><span class="s2">&quot;&quot;&quot;</span>

        <span class="nv">properties</span> <span class="o">=</span> <span class="nv">dict</span><span class="ss">(</span>

            <span class="nv">label</span><span class="o">=</span><span class="nv">self</span>.<span class="nv">compute_feature_label</span><span class="ss">(</span><span class="nv">feature</span><span class="ss">)</span>,

            <span class="nv">color</span><span class="o">=</span><span class="nv">self</span>.<span class="nv">compute_feature_color</span><span class="ss">(</span><span class="nv">feature</span><span class="ss">)</span>,

            <span class="nv">html</span><span class="o">=</span><span class="nv">self</span>.<span class="nv">compute_feature_html</span><span class="ss">(</span><span class="nv">feature</span><span class="ss">)</span>,

            <span class="nv">fontdict</span><span class="o">=</span><span class="nv">self</span>.<span class="nv">compute_feature_fontdict</span><span class="ss">(</span><span class="nv">feature</span><span class="ss">)</span>,

            <span class="nv">box_linewidth</span><span class="o">=</span><span class="nv">self</span>.<span class="nv">compute_feature_box_linewidth</span><span class="ss">(</span><span class="nv">feature</span><span class="ss">)</span>,

            <span class="nv">box_color</span><span class="o">=</span><span class="nv">self</span>.<span class="nv">compute_feature_box_color</span><span class="ss">(</span><span class="nv">feature</span><span class="ss">)</span>,

            <span class="nv">linewidth</span><span class="o">=</span><span class="nv">self</span>.<span class="nv">compute_feature_linewidth</span><span class="ss">(</span><span class="nv">feature</span><span class="ss">)</span>,

            <span class="nv">label_link_color</span><span class="o">=</span><span class="nv">self</span>.<span class="nv">compute_feature_label_link_color</span><span class="ss">(</span><span class="nv">feature</span><span class="ss">)</span>,

            <span class="nv">legend_text</span><span class="o">=</span><span class="nv">self</span>.<span class="nv">compute_feature_legend_text</span><span class="ss">(</span><span class="nv">feature</span><span class="ss">)</span>,

        <span class="ss">)</span>

        <span class="k">if</span> <span class="nv">self</span>.<span class="nv">features_properties</span> <span class="nv">is</span> <span class="nv">not</span> <span class="nv">None</span>:

            <span class="nv">other_properties</span> <span class="o">=</span> <span class="nv">self</span>.<span class="nv">features_properties</span>

            <span class="k">if</span> <span class="nv">hasattr</span><span class="ss">(</span><span class="nv">other_properties</span>, <span class="s2">&quot;</span><span class="s">__call__</span><span class="s2">&quot;</span><span class="ss">)</span>:

                <span class="nv">other_properties</span> <span class="o">=</span> <span class="nv">other_properties</span><span class="ss">(</span><span class="nv">feature</span><span class="ss">)</span>

            <span class="nv">properties</span>.<span class="nv">update</span><span class="ss">(</span><span class="nv">other_properties</span><span class="ss">)</span>

        <span class="k">return</span> <span class="nv">GraphicFeature</span><span class="ss">(</span>

            <span class="nv">start</span><span class="o">=</span><span class="nv">feature</span>.<span class="nv">location</span>.<span class="nv">start</span>,

            <span class="k">end</span><span class="o">=</span><span class="nv">feature</span>.<span class="nv">location</span>.<span class="k">end</span>,

            <span class="nv">strand</span><span class="o">=</span><span class="nv">feature</span>.<span class="nv">location</span>.<span class="nv">strand</span>,

            <span class="o">**</span><span class="nv">properties</span>

        <span class="ss">)</span>
</code></pre></div>

</details>
<h4 id="translate_record">translate_record</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">translate_record</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">record</span><span class="p">,</span>
    <span class="n">record_class</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">filetype</span><span class="o">=</span><span class="kc">None</span>
<span class="p">)</span>
</code></pre></div>

<p>Create a new GraphicRecord from a BioPython Record object.</p>
<p><strong>Parameters:</strong></p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Type</th>
<th>Description</th>
<th>Default</th>
</tr>
</thead>
<tbody>
<tr>
<td>record</td>
<td>None</td>
<td>A BioPython Record object or the path to a Genbank or a GFF file.</td>
<td>None</td>
</tr>
<tr>
<td>record_class</td>
<td>None</td>
<td>The graphic record class to use, e.g. GraphicRecord (default) or</td>
<td></td>
</tr>
<tr>
<td>CircularGraphicRecord. Strings 'circular' and 'linear' can also be</td>
<td></td>
<td></td>
<td></td>
</tr>
<tr>
<td>provided.</td>
<td>None</td>
<td></td>
<td></td>
</tr>
<tr>
<td>filetype</td>
<td>None</td>
<td>Used only when a Genbank or a GFF file is provided; one of "genbank"</td>
<td></td>
</tr>
<tr>
<td>or "gff" to be used. Default None infers from file extension.</td>
<td>None</td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code><span class="w">    </span><span class="n">def</span><span class="w"> </span><span class="n">translate_record</span><span class="p">(</span><span class="n">self</span><span class="p">,</span><span class="w"> </span><span class="n">record</span><span class="p">,</span><span class="w"> </span><span class="n">record_class</span><span class="o">=</span><span class="k">None</span><span class="p">,</span><span class="w"> </span><span class="n">filetype</span><span class="o">=</span><span class="k">None</span><span class="p">)</span><span class="err">:</span><span class="w"></span>

<span class="w">        </span><span class="ss">&quot;&quot;&quot;Create a new GraphicRecord from a BioPython Record object.</span>

<span class="ss">        Parameters</span>

<span class="ss">        ----------</span>

<span class="ss">        record</span>

<span class="ss">          A BioPython Record object or the path to a Genbank or a GFF file.</span>

<span class="ss">        record_class</span>

<span class="ss">          The graphic record class to use, e.g. GraphicRecord (default) or</span>

<span class="ss">          CircularGraphicRecord. Strings &#39;circular&#39; and &#39;linear&#39; can also be</span>

<span class="ss">          provided.</span>

<span class="ss">        filetype</span>

<span class="ss">          Used only when a Genbank or a GFF file is provided; one of &quot;</span><span class="n">genbank</span><span class="ss">&quot;</span>

<span class="ss">          or &quot;</span><span class="n">gff</span><span class="ss">&quot; to be used. Default None infers from file extension.</span>

<span class="ss">        &quot;&quot;&quot;</span><span class="w"></span>

<span class="w">        </span><span class="n">classes</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="err">{</span><span class="w"></span>

<span class="w">            </span><span class="ss">&quot;linear&quot;</span><span class="err">:</span><span class="w"> </span><span class="n">GraphicRecord</span><span class="p">,</span><span class="w"></span>

<span class="w">            </span><span class="ss">&quot;circular&quot;</span><span class="err">:</span><span class="w"> </span><span class="n">CircularGraphicRecord</span><span class="p">,</span><span class="w"></span>

<span class="w">            </span><span class="k">None</span><span class="err">:</span><span class="w"> </span><span class="n">GraphicRecord</span><span class="p">,</span><span class="w"></span>

<span class="w">        </span><span class="err">}</span><span class="w"></span>

<span class="w">        </span><span class="k">if</span><span class="w"> </span><span class="n">record_class</span><span class="w"> </span><span class="ow">in</span><span class="w"> </span><span class="nl">classes</span><span class="p">:</span><span class="w"></span>

<span class="w">            </span><span class="n">record_class</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">classes</span><span class="o">[</span><span class="n">record_class</span><span class="o">]</span><span class="w"></span>

<span class="w">        </span><span class="k">if</span><span class="w"> </span><span class="n">isinstance</span><span class="p">(</span><span class="n">record</span><span class="p">,</span><span class="w"> </span><span class="nf">str</span><span class="p">)</span><span class="w"> </span><span class="ow">or</span><span class="w"> </span><span class="n">hasattr</span><span class="p">(</span><span class="n">record</span><span class="p">,</span><span class="w"> </span><span class="ss">&quot;read&quot;</span><span class="p">)</span><span class="err">:</span><span class="w"></span>

<span class="w">            </span><span class="n">record</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">load_record</span><span class="p">(</span><span class="n">record</span><span class="p">,</span><span class="w"> </span><span class="n">filetype</span><span class="o">=</span><span class="n">filetype</span><span class="p">)</span><span class="w"></span>

<span class="w">        </span><span class="n">filtered_features</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">self</span><span class="p">.</span><span class="n">compute_filtered_features</span><span class="p">(</span><span class="n">record</span><span class="p">.</span><span class="n">features</span><span class="p">)</span><span class="w"></span>

<span class="w">        </span><span class="k">return</span><span class="w"> </span><span class="n">record_class</span><span class="p">(</span><span class="w"></span>

<span class="w">            </span><span class="n">sequence_length</span><span class="o">=</span><span class="nf">len</span><span class="p">(</span><span class="n">record</span><span class="p">),</span><span class="w"></span>

<span class="w">            </span><span class="k">sequence</span><span class="o">=</span><span class="nf">str</span><span class="p">(</span><span class="n">record</span><span class="p">.</span><span class="n">seq</span><span class="p">),</span><span class="w"></span>

<span class="w">            </span><span class="n">features</span><span class="o">=[</span><span class="n"></span>

<span class="n">                self.translate_feature(feature)</span>

<span class="n">                for feature in filtered_features</span>

<span class="n">                if feature.location is not None</span>

<span class="n">            </span><span class="o">]</span><span class="p">,</span><span class="w"></span>

<span class="w">            </span><span class="o">**</span><span class="n">self</span><span class="p">.</span><span class="n">graphic_record_parameters</span><span class="w"></span>

<span class="w">        </span><span class="p">)</span><span class="w"></span>
</code></pre></div>

</details>
<h3 id="sequencinggroup">SequencingGroup</h3>
<div class="codehilite"><pre><span></span><code><span class="k">class</span> <span class="nc">SequencingGroup</span><span class="p">(</span>
    <span class="n">comparatorgroups</span><span class="p">,</span>
    <span class="n">name</span><span class="o">=</span><span class="s1">&#39;Unnamed&#39;</span>
<span class="p">)</span>
</code></pre></div>

<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code><span class="n">class</span> <span class="n">SequencingGroup</span><span class="o">:</span>

    <span class="s2">&quot;</span><span class="se">&quot;&quot;</span><span class="s2">Analyse alignments of a sequencing run.</span>

<span class="s2">    **Parameters**</span>

<span class="s2">    **comparatorgroups**</span>

<span class="s2">    &gt; List of `ComparatorGroup` instances.</span>

<span class="s2">    **name**</span>

<span class="s2">    &gt; Name of the sequencing project (`str`).</span>

<span class="s2">    </span><span class="se">&quot;&quot;</span><span class="s2">&quot;</span>

    <span class="n">def</span> <span class="n">__init__</span><span class="p">(</span><span class="n">self</span><span class="p">,</span> <span class="n">comparatorgroups</span><span class="p">,</span> <span class="k">name</span><span class="o">=</span><span class="s2">&quot;Unnamed&quot;</span><span class="p">)</span><span class="o">:</span>

        <span class="n">self</span><span class="p">.</span><span class="n">comparatorgroups</span> <span class="o">=</span> <span class="n">comparatorgroups</span>

        <span class="n">self</span><span class="p">.</span><span class="k">name</span> <span class="o">=</span> <span class="k">name</span>

        <span class="n">self</span><span class="p">.</span><span class="n">vcf_cutoff</span> <span class="o">=</span> <span class="mi">20</span>  <span class="c1"># max number of VCF entries to display in report</span>

        <span class="n">self</span><span class="p">.</span><span class="n">result_keywords</span> <span class="o">=</span> <span class="n">result_keywords</span>

    <span class="n">def</span> <span class="n">perform_all_comparisons_in_sequencinggroup</span><span class="p">(</span><span class="n">self</span><span class="p">)</span><span class="o">:</span>

        <span class="n">self</span><span class="p">.</span><span class="n">number_of_reads</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="k">for</span> <span class="n">comparatorgroup</span> <span class="k">in</span> <span class="n">self</span><span class="p">.</span><span class="n">comparatorgroups</span><span class="o">:</span>

            <span class="n">comparatorgroup</span><span class="p">.</span><span class="n">perform_all_comparisons</span><span class="p">()</span>

            <span class="n">self</span><span class="p">.</span><span class="n">number_of_reads</span> <span class="o">+=</span> <span class="n">comparatorgroup</span><span class="p">.</span><span class="n">n_fastq_reads</span>

        <span class="n">self</span><span class="p">.</span><span class="n">number_of_barcodes</span> <span class="o">=</span> <span class="n">len</span><span class="p">(</span><span class="n">self</span><span class="p">.</span><span class="n">comparatorgroups</span><span class="p">)</span>

        <span class="n">self</span><span class="p">.</span><span class="n">comparisons_performed</span> <span class="o">=</span> <span class="no">True</span>
</code></pre></div>

</details>
<hr />
<h4 id="methods_3">Methods</h4>
<h4 id="perform_all_comparisons_in_sequencinggroup">perform_all_comparisons_in_sequencinggroup</h4>
<div class="codehilite"><pre><span></span><code><span class="k">def</span> <span class="nf">perform_all_comparisons_in_sequencinggroup</span><span class="p">(</span>
    <span class="bp">self</span>
<span class="p">)</span>
</code></pre></div>

<details class="example">
<summary>View Source</summary>
<div class="codehilite"><pre><span></span><code>    <span class="nv">def</span> <span class="nv">perform_all_comparisons_in_sequencinggroup</span><span class="ss">(</span><span class="nv">self</span><span class="ss">)</span>:

        <span class="nv">self</span>.<span class="nv">number_of_reads</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="k">for</span> <span class="nv">comparatorgroup</span> <span class="nv">in</span> <span class="nv">self</span>.<span class="nv">comparatorgroups</span>:

            <span class="nv">comparatorgroup</span>.<span class="nv">perform_all_comparisons</span><span class="ss">()</span>

            <span class="nv">self</span>.<span class="nv">number_of_reads</span> <span class="o">+=</span> <span class="nv">comparatorgroup</span>.<span class="nv">n_fastq_reads</span>

        <span class="nv">self</span>.<span class="nv">number_of_barcodes</span> <span class="o">=</span> <span class="nv">len</span><span class="ss">(</span><span class="nv">self</span>.<span class="nv">comparatorgroups</span><span class="ss">)</span>

        <span class="nv">self</span>.<span class="nv">comparisons_performed</span> <span class="o">=</span> <span class="nv">True</span>
</code></pre></div>

</details>
                
              
              
                


              
            </article>
          </div>
        </div>
        
      </main>
      
        
<footer class="md-footer">
  
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid">
        
          <a href="../Assembly/" title="Assembly" class="md-flex md-footer-nav__link md-footer-nav__link--prev" rel="prev">
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-back md-footer-nav__button"></i>
            </div>
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Previous
                </span>
                Assembly
              </span>
            </div>
          </a>
        
        
          <a href="../" title="Index" class="md-flex md-footer-nav__link md-footer-nav__link--next" rel="next">
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Next
                </span>
                Index
              </span>
            </div>
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-forward md-footer-nav__button"></i>
            </div>
          </a>
        
      </nav>
    </div>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
        Powered by
        <a href="http://timothycrosley.github.io/portray">portray.</a>
        You too can
        <a href="http://timothycrosley.github.io/portray">
          portray</a>
        your Python project well using automatic documentation.
      </div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    <script id="__config" type="application/json">{"base": "../../..", "features": [], "translations": {"clipboard.copy": "Copy to clipboard", "clipboard.copied": "Copied to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.placeholder": "Type to start searching", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.term.missing": "Missing", "select.version.title": "Select version"}, "search": "../../../assets/javascripts/workers/search.f8263e09.min.js", "version": null}</script>
    
    
      <script src="../../../assets/javascripts/bundle.4fc53ad4.min.js"></script>
      
    
  </body>
</html>